{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "297d58d0-f2bd-4a6c-9feb-fb0836166cb6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "db0a7fcb-217e-44b9-971d-8b224457a88f",
   "metadata": {},
   "outputs": [],
   "source": [
    "clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "merged_df[['cluster', 'kingdom']].to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "da945e14-e934-46fe-aec8-e12d97548e45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3030"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_df = merged_df[merged_df['kingdom'] == 'Viruses']\n",
    "viral_df.to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/viral_clusters.csv', index=None)\n",
    "len(viral_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a82666f3-101e-4425-9e55-a9f4177291df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90908"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteria_df = merged_df[merged_df['kingdom'] == 'Bacteria']\n",
    "bacteria_df.to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Test_blast_SENZOR/LCA_results/bacteria_clusters.csv')\n",
    "len(bacteria_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "93ec16e2-82bf-4457-9244-1b1d66c535b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample of combined nr_df before dropping NaNs in 'taxid':\n",
      "                                               query  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459~NODE_10000_length_3...   \n",
      "1  AIAMA_GOT005_OS_S43_673459~NODE_10003_length_3...   \n",
      "2  AIAMA_GOT005_OS_S43_673459~NODE_10010_length_3...   \n",
      "3  AIAMA_GOT005_OS_S43_673459~NODE_10014_length_3...   \n",
      "4  AIAMA_GOT005_OS_S43_673459~NODE_10015_length_3...   \n",
      "\n",
      "                                             subject  identity  align_length  \\\n",
      "0  MCU8647632.1 hypothetical protein [Escherichia...     100.0            73   \n",
      "1                       QLI46611.1 VP4 [Rotavirus A]      98.9            88   \n",
      "2  XP_041588126.1 tyrosine-protein kinase Srms-li...      96.9            65   \n",
      "3  XP_048962668.1 uncharacterized protein LOC1183...      90.3            31   \n",
      "4  WP_288535963.1 M42 family metallopeptidase [un...     100.0            95   \n",
      "\n",
      "   mismatches  gap_opens  q_start  q_end  s_start  s_end  ...  \\\n",
      "0           0          0      316     98        3     75  ...   \n",
      "1           1          0       53    316      600    687  ...   \n",
      "2           2          0      219     25      247    311  ...   \n",
      "3           3          0      291    199      303    333  ...   \n",
      "4           0          0      286      2        7    101  ...   \n",
      "\n",
      "                      sci_name  \\\n",
      "0             Escherichia coli   \n",
      "1                  Rotavirus A   \n",
      "2               Vulpes lagopus   \n",
      "3            Canis lupus dingo   \n",
      "4  uncultured Holdemanella sp.   \n",
      "\n",
      "                                           com_names  superkingdom  \\\n",
      "0  MCU8647632.1 hypothetical protein [Escherichia...      Bacteria   \n",
      "1                       QLI46611.1 VP4 [Rotavirus A]       Viruses   \n",
      "2  XP_041588126.1 tyrosine-protein kinase Srms-li...           NaN   \n",
      "3  XP_048962668.1 uncharacterized protein LOC1183...           NaN   \n",
      "4  WP_288535963.1 M42 family metallopeptidase [un...           NaN   \n",
      "\n",
      "           phylum              family lca_taxid          lca_name  \\\n",
      "0  Proteobacteria  Enterobacteriaceae     562.0  Escherichia coli   \n",
      "1             NaN          Reoviridae   28875.0       Rotavirus A   \n",
      "2             NaN                 NaN       NaN               NaN   \n",
      "3             NaN                 NaN       NaN               NaN   \n",
      "4             NaN                 NaN       NaN               NaN   \n",
      "\n",
      "                       sample                              contig  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459  NODE_10000_length_317_cov_1.612500   \n",
      "1  AIAMA_GOT005_OS_S43_673459  NODE_10003_length_317_cov_1.341667   \n",
      "2  AIAMA_GOT005_OS_S43_673459  NODE_10010_length_317_cov_1.233333   \n",
      "3  AIAMA_GOT005_OS_S43_673459  NODE_10014_length_317_cov_1.229167   \n",
      "4  AIAMA_GOT005_OS_S43_673459  NODE_10015_length_317_cov_1.229167   \n",
      "\n",
      "                                          contig_key  \n",
      "0  aiama_got005_os_s43_673459~node_10000_length_3...  \n",
      "1  aiama_got005_os_s43_673459~node_10003_length_3...  \n",
      "2  aiama_got005_os_s43_673459~node_10010_length_3...  \n",
      "3  aiama_got005_os_s43_673459~node_10014_length_3...  \n",
      "4  aiama_got005_os_s43_673459~node_10015_length_3...  \n",
      "\n",
      "[5 rows x 23 columns]\n",
      "Sample of combined nr_df after dropping NaNs in 'taxid':\n",
      "                                               query  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459~NODE_10000_length_3...   \n",
      "1  AIAMA_GOT005_OS_S43_673459~NODE_10003_length_3...   \n",
      "2  AIAMA_GOT005_OS_S43_673459~NODE_10010_length_3...   \n",
      "3  AIAMA_GOT005_OS_S43_673459~NODE_10014_length_3...   \n",
      "4  AIAMA_GOT005_OS_S43_673459~NODE_10015_length_3...   \n",
      "\n",
      "                                             subject  identity  align_length  \\\n",
      "0  MCU8647632.1 hypothetical protein [Escherichia...     100.0            73   \n",
      "1                       QLI46611.1 VP4 [Rotavirus A]      98.9            88   \n",
      "2  XP_041588126.1 tyrosine-protein kinase Srms-li...      96.9            65   \n",
      "3  XP_048962668.1 uncharacterized protein LOC1183...      90.3            31   \n",
      "4  WP_288535963.1 M42 family metallopeptidase [un...     100.0            95   \n",
      "\n",
      "   mismatches  gap_opens  q_start  q_end  s_start  s_end  ...  \\\n",
      "0           0          0      316     98        3     75  ...   \n",
      "1           1          0       53    316      600    687  ...   \n",
      "2           2          0      219     25      247    311  ...   \n",
      "3           3          0      291    199      303    333  ...   \n",
      "4           0          0      286      2        7    101  ...   \n",
      "\n",
      "                      sci_name  \\\n",
      "0             Escherichia coli   \n",
      "1                  Rotavirus A   \n",
      "2               Vulpes lagopus   \n",
      "3            Canis lupus dingo   \n",
      "4  uncultured Holdemanella sp.   \n",
      "\n",
      "                                           com_names  superkingdom  \\\n",
      "0  MCU8647632.1 hypothetical protein [Escherichia...      Bacteria   \n",
      "1                       QLI46611.1 VP4 [Rotavirus A]       Viruses   \n",
      "2  XP_041588126.1 tyrosine-protein kinase Srms-li...           NaN   \n",
      "3  XP_048962668.1 uncharacterized protein LOC1183...           NaN   \n",
      "4  WP_288535963.1 M42 family metallopeptidase [un...           NaN   \n",
      "\n",
      "           phylum              family lca_taxid          lca_name  \\\n",
      "0  Proteobacteria  Enterobacteriaceae     562.0  Escherichia coli   \n",
      "1             NaN          Reoviridae   28875.0       Rotavirus A   \n",
      "2             NaN                 NaN       NaN               NaN   \n",
      "3             NaN                 NaN       NaN               NaN   \n",
      "4             NaN                 NaN       NaN               NaN   \n",
      "\n",
      "                       sample                              contig  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459  NODE_10000_length_317_cov_1.612500   \n",
      "1  AIAMA_GOT005_OS_S43_673459  NODE_10003_length_317_cov_1.341667   \n",
      "2  AIAMA_GOT005_OS_S43_673459  NODE_10010_length_317_cov_1.233333   \n",
      "3  AIAMA_GOT005_OS_S43_673459  NODE_10014_length_317_cov_1.229167   \n",
      "4  AIAMA_GOT005_OS_S43_673459  NODE_10015_length_317_cov_1.229167   \n",
      "\n",
      "                                          contig_key  \n",
      "0  aiama_got005_os_s43_673459~node_10000_length_3...  \n",
      "1  aiama_got005_os_s43_673459~node_10003_length_3...  \n",
      "2  aiama_got005_os_s43_673459~node_10010_length_3...  \n",
      "3  aiama_got005_os_s43_673459~node_10014_length_3...  \n",
      "4  aiama_got005_os_s43_673459~node_10015_length_3...  \n",
      "\n",
      "[5 rows x 23 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                               query  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459~NODE_10000_length_3...   \n",
      "1  AIAMA_GOT005_OS_S43_673459~NODE_10003_length_3...   \n",
      "2  AIAMA_GOT005_OS_S43_673459~NODE_10010_length_3...   \n",
      "\n",
      "                                             subject  identity  align_length  \\\n",
      "0  MCU8647632.1 hypothetical protein [Escherichia...     100.0            73   \n",
      "1                       QLI46611.1 VP4 [Rotavirus A]      98.9            88   \n",
      "2  XP_041588126.1 tyrosine-protein kinase Srms-li...      96.9            65   \n",
      "\n",
      "   mismatches  gap_opens  q_start  q_end  s_start  s_end  ...  \\\n",
      "0           0          0      316     98        3     75  ...   \n",
      "1           1          0       53    316      600    687  ...   \n",
      "2           2          0      219     25      247    311  ...   \n",
      "\n",
      "               family  lca_taxid          lca_name  \\\n",
      "0  Enterobacteriaceae      562.0  Escherichia coli   \n",
      "1          Reoviridae    28875.0       Rotavirus A   \n",
      "2                 NaN        NaN               NaN   \n",
      "\n",
      "                       sample                              contig  \\\n",
      "0  AIAMA_GOT005_OS_S43_673459  NODE_10000_length_317_cov_1.612500   \n",
      "1  AIAMA_GOT005_OS_S43_673459  NODE_10003_length_317_cov_1.341667   \n",
      "2  AIAMA_GOT005_OS_S43_673459  NODE_10010_length_317_cov_1.233333   \n",
      "\n",
      "                                          contig_key              name  \\\n",
      "0  aiama_got005_os_s43_673459~node_10000_length_3...  Escherichia coli   \n",
      "1  aiama_got005_os_s43_673459~node_10003_length_3...       Rotavirus A   \n",
      "2  aiama_got005_os_s43_673459~node_10010_length_3...    Vulpes lagopus   \n",
      "\n",
      "      rank    kingdom  db  \n",
      "0  species   Bacteria  nr  \n",
      "1  species    Viruses  nr  \n",
      "2  species  Eukaryota  nr  \n",
      "\n",
      "[3 rows x 27 columns]\n",
      "clust2kingdom Dictionary Sample Entries:\n",
      "Cluster 6: Kingdom Eukaryota\n",
      "Cluster 8: Kingdom Eukaryota\n",
      "Cluster 13: Kingdom Eukaryota\n",
      "Cluster 14: Kingdom Viruses\n",
      "Cluster 17: Kingdom Viruses\n",
      "Filtered out 417 rows without 'count' value.\n",
      "Final Merged DataFrame with Counts (after filtering):\n",
      "   cluster  cluster_size  contig_length                name     rank  \\\n",
      "0        6            10          33937  Arvicola amphibius  species   \n",
      "1        8            30          31738    Neotoma mexicana  species   \n",
      "2       13            10          30316  Arvicola amphibius  species   \n",
      "3       14            16          30171           Orf virus  species   \n",
      "4       17             8          29205   Betacoronavirus 1  species   \n",
      "\n",
      "     kingdom    taxid  db  align_length  align_percent  ...  \\\n",
      "0  Eukaryota  1047088  nr          2206            6.5  ...   \n",
      "1  Eukaryota   105149  nr           364            1.1  ...   \n",
      "2  Eukaryota  1047088  nr          2206            7.3  ...   \n",
      "3    Viruses    10258  nr           905            3.0  ...   \n",
      "4    Viruses   694003  nr          4383           15.0  ...   \n",
      "\n",
      "                                           com_names  superkingdom  phylum  \\\n",
      "0  XP_038168257.2 mucin-17-like [Arvicola amphibius]           NaN     NaN   \n",
      "1  UZH93544.1 NADH dehydrogenase subunit 5 [Neoto...           NaN     NaN   \n",
      "2  XP_038168257.2 mucin-17-like [Arvicola amphibius]           NaN     NaN   \n",
      "3  AHZ33783.1 virion core protein P4a precursor [...       Viruses     NaN   \n",
      "4   UVH33440.1 ORF1a polyprotein [Betacoronavirus 1]           NaN     NaN   \n",
      "\n",
      "       family lca_taxid   lca_name                              contig  \\\n",
      "0         NaN       NaN        NaN   NODE_1_length_33937_cov_79.209894   \n",
      "1         NaN       NaN        NaN   NODE_1_length_31738_cov_62.663782   \n",
      "2         NaN       NaN        NaN    NODE_1_length_30316_cov_7.943384   \n",
      "3  Poxviridae   10258.0  Orf virus   NODE_1_length_30171_cov_60.218515   \n",
      "4         NaN       NaN        NaN  NODE_1_length_29205_cov_133.285841   \n",
      "\n",
      "                           sample  \\\n",
      "0    OD_ZAD_AGR005_LIV_S35_673488   \n",
      "1      AINWZ_GOT012_RS_S38_673479   \n",
      "2   OD_ZAD_AGR006_LUNG_S36_673489   \n",
      "3  SENONOWISUGOT002_OS_S76_673660   \n",
      "4  SENONOWISUGOT001_RS_S74_673659   \n",
      "\n",
      "                                          contig_key    count  \n",
      "0  od_zad_agr005_liv_s35_673488~node_1_length_339...  11337.0  \n",
      "1  ainwz_got012_rs_s38_673479~node_1_length_31738...  18632.0  \n",
      "2  od_zad_agr006_lung_s36_673489~node_1_length_30...   5457.0  \n",
      "3  senonowisugot002_os_s76_673660~node_1_length_3...  50072.0  \n",
      "4  senonowisugot001_rs_s74_673659~node_1_length_2...  91014.0  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "from statistics import mode\n",
    "from Bio import Entrez\n",
    "from ete3 import NCBITaxa\n",
    "import logging\n",
    "import sys\n",
    "import re\n",
    "from logging.handlers import RotatingFileHandler\n",
    "\n",
    "# Configure logging\n",
    "handler = RotatingFileHandler('util.log', maxBytes=5*1024*1024, backupCount=2)\n",
    "logging.basicConfig(level=logging.INFO, handlers=[handler],\n",
    "                    format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "# Set Entrez email\n",
    "Entrez.email = \"ifeanyi.omah@ed.ac.uk\"\n",
    "\n",
    "# Initialize NCBI Taxonomy\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "valid_lineages = set()\n",
    "\n",
    "# Updated column names to match your .m9 file\n",
    "nr_column_names = [\n",
    "    'query', 'subject', 'identity', 'alignment_length', 'mismatches', 'gap_opens',\n",
    "    'q_start', 'q_end', 's_start', 's_end', 'evalue', 'bit_score', 'tax_id',\n",
    "    'sci_name', 'com_names', 'superkingdom', 'phylum', 'family',\n",
    "    'lca_taxid', 'lca_name'\n",
    "]\n",
    "\n",
    "# --------------------------------------------------------------------\n",
    "# 1) Change nr_dir to where your NR LCA files actually live\n",
    "#    If these files are named like: LCA_<SAMPLE>_modified_diamond_blast_nr_annotated.m9\n",
    "#    then we can glob them below.\n",
    "# --------------------------------------------------------------------\n",
    "nr_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/\"\n",
    "\n",
    "def load_lca(nr_directory):\n",
    "    \"\"\"\n",
    "    Load and process LCA results from NR files in the specified directory.\n",
    "    Dynamically glob any pattern you want.\n",
    "    \"\"\"\n",
    "    pattern = \"LCA_*_modified_diamond_blast_nr_annotated.m9\"\n",
    "    nr_file_paths = glob(os.path.join(nr_directory, pattern))\n",
    "    logging.info(f\"Found {len(nr_file_paths)} NR files matching '{pattern}' in {nr_directory}.\")\n",
    "\n",
    "    if not nr_file_paths:\n",
    "        logging.error(f\"No files matching {pattern} found in directory: {nr_directory}\")\n",
    "        raise FileNotFoundError(f\"No files matching {pattern} found in directory: {nr_directory}\")\n",
    "\n",
    "    # Process each BLAST output file\n",
    "    nr_df = load_blast_output(nr_file_paths, database='nr')\n",
    "\n",
    "    # Log columns of combined DataFrame\n",
    "    logging.info(f\"Columns in combined nr_df: {nr_df.columns.tolist()}\")\n",
    "\n",
    "    # Debugging: Print the first few rows before dropping NaNs\n",
    "    print(\"Sample of combined nr_df before dropping NaNs in 'taxid':\")\n",
    "    print(nr_df.head())\n",
    "\n",
    "    # Remove rows with missing taxid\n",
    "    initial_nr = len(nr_df)\n",
    "    if 'taxid' not in nr_df.columns:\n",
    "        logging.error(\"'taxid' column is missing in the combined DataFrame after loading.\")\n",
    "        print(\"'taxid' column is missing in the combined DataFrame after loading.\", file=sys.stderr)\n",
    "        raise KeyError(\"'taxid' column is missing in the combined DataFrame after loading.\")\n",
    "    nr_df = nr_df.dropna(subset=['taxid'])\n",
    "    final_nr = len(nr_df)\n",
    "    dropped_nr = initial_nr - final_nr\n",
    "    if dropped_nr > 0:\n",
    "        logging.info(f\"Dropped {dropped_nr} rows with missing taxid from nr_df.\")\n",
    "\n",
    "    # Debugging: Print the first few rows after dropping NaNs\n",
    "    print(\"Sample of combined nr_df after dropping NaNs in 'taxid':\")\n",
    "    print(nr_df.head())\n",
    "\n",
    "    # Convert taxid to integer\n",
    "    try:\n",
    "        nr_df['taxid'] = nr_df['taxid'].astype(int)\n",
    "    except ValueError as e:\n",
    "        logging.error(f\"Error converting 'taxid' to integer: {e}\")\n",
    "        print(f\"Error converting 'taxid' to integer: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    # Build dictionaries for name/rank\n",
    "    taxids = set(nr_df.taxid)\n",
    "    taxids = {t for taxid in taxids for t in ncbi.get_lineage(taxid) if t != 0}\n",
    "\n",
    "    try:\n",
    "        taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error fetching taxid to name mapping: {e}\")\n",
    "        print(f\"Error fetching taxid to name mapping: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    taxid2name[-1] = 'Ambiguous'\n",
    "    taxid2name[0] = 'Unknown'\n",
    "    try:\n",
    "        taxid2rank = ncbi.get_rank(taxids)\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error fetching taxid to rank mapping: {e}\")\n",
    "        print(f\"Error fetching taxid to rank mapping: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    # Collect valid lineages\n",
    "    taxids_in_df = set(nr_df['taxid'])\n",
    "    for tx in taxids_in_df:\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(tx)\n",
    "            for t in lineage:\n",
    "                if t != 0:\n",
    "                    valid_lineages.add(t)\n",
    "        except ValueError:\n",
    "            logging.warning(f\"Skipping {tx}: taxid not found in local DB.\")\n",
    "\n",
    "    # Restrict to valid lineages\n",
    "    taxids = valid_lineages\n",
    "    try:\n",
    "        taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error fetching taxid to name mapping: {e}\")\n",
    "        print(f\"Error fetching taxid to name mapping: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    def taxid2kingdom(taxid):\n",
    "        if taxid == 0:\n",
    "            return 'Unknown'\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(taxid)\n",
    "            if not lineage:\n",
    "                return 'Unknown'\n",
    "            ranks = ncbi.get_rank(lineage)\n",
    "            for taxid_in_lineage in lineage:\n",
    "                rank = ranks.get(taxid_in_lineage, None)\n",
    "                if rank == 'superkingdom':\n",
    "                    kingdom = taxid2name.get(taxid_in_lineage, 'Unknown')\n",
    "                    return kingdom\n",
    "            return 'Unknown'\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error fetching lineage for taxid {taxid}: {e}\")\n",
    "            return 'Unknown'\n",
    "\n",
    "    # Map taxid -> name, rank, kingdom\n",
    "    nr_df['name'] = nr_df['taxid'].map(taxid2name)\n",
    "    nr_df['rank'] = nr_df['taxid'].map(taxid2rank)\n",
    "    nr_df['kingdom'] = nr_df['taxid'].apply(taxid2kingdom)\n",
    "    nr_df['db'] = 'nr'  # Since this is the NR dataset\n",
    "\n",
    "    return nr_df\n",
    "\n",
    "def load_blast_output(files, database='nr'):\n",
    "    \"\"\"\n",
    "    Load BLAST output files into a single DataFrame.\n",
    "    \"\"\"\n",
    "    if database == 'nr':\n",
    "        column_names = nr_column_names\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        try:\n",
    "            # Quick check of columns in the first line\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                columns_in_file = len(first_line.strip().split('\\t'))\n",
    "                expected_columns = len(column_names)\n",
    "                if columns_in_file != expected_columns:\n",
    "                    logging.warning(f\"{file} has {columns_in_file} columns; expected {expected_columns}.\")\n",
    "\n",
    "            # Potentially skip a header if present\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                if 'sci_name' in first_line.lower() or 'lca_taxid' in first_line.lower():\n",
    "                    skiprows = 1\n",
    "                else:\n",
    "                    skiprows = 0\n",
    "\n",
    "            df = pd.read_csv(\n",
    "                file,\n",
    "                sep='\\t',\n",
    "                header=None,\n",
    "                names=column_names,\n",
    "                na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "                skiprows=skiprows,\n",
    "                on_bad_lines='warn'  # or 'error' if you want it strict\n",
    "            )\n",
    "\n",
    "            # Log columns after reading\n",
    "            logging.info(f\"Columns in {file} after reading: {df.columns.tolist()}\")\n",
    "\n",
    "            # Standardize column names to lowercase\n",
    "            df.columns = [col.lower() for col in df.columns]\n",
    "\n",
    "            if 'tax_id' in df.columns:\n",
    "                df.rename(columns={'tax_id': 'taxid'}, inplace=True)\n",
    "                logging.info(f\"'tax_id' renamed to 'taxid' in {file}.\")\n",
    "            elif 'taxid' in df.columns:\n",
    "                logging.info(f\"'taxid' column is already correctly named in {file}.\")\n",
    "            else:\n",
    "                logging.error(f\"No 'tax_id' or 'taxid' column found in {file}.\")\n",
    "                raise KeyError(f\"No 'tax_id' or 'taxid' column found in {file}.\")\n",
    "\n",
    "            # Replace 'Unknown' with '0' in taxid\n",
    "            df['taxid'] = df['taxid'].replace('Unknown', '0')\n",
    "\n",
    "            # Convert taxid to numeric\n",
    "            df['taxid'] = pd.to_numeric(df['taxid'], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in 'taxid'\n",
    "            before_drop = len(df)\n",
    "            df.dropna(subset=['taxid'], inplace=True)\n",
    "            after_drop = len(df)\n",
    "            dropped = before_drop - after_drop\n",
    "            if dropped > 0:\n",
    "                logging.info(f\"Dropped {dropped} rows with invalid taxid in {file}.\")\n",
    "\n",
    "            # Rename 'alignment_length' to 'align_length' if present\n",
    "            if 'alignment_length' in df.columns:\n",
    "                df.rename(columns={'alignment_length': 'align_length'}, inplace=True)\n",
    "                logging.info(f\"'alignment_length' renamed to 'align_length' in {file}.\")\n",
    "\n",
    "            # Derive sample name\n",
    "            df['sample'] = get_sample_from_path(file)\n",
    "            logging.info(f\"Derived 'sample' for {file}: {df['sample'].unique()}\")\n",
    "\n",
    "            # Extract contig\n",
    "            df['contig'] = df.apply(lambda row: extract_contig(row['query'], row['sample']), axis=1)\n",
    "            logging.info(f\"Extracted 'contig' for {file}.\")\n",
    "\n",
    "            # Build contig_key\n",
    "            df['contig_key'] = (df['sample'] + '~' + df['contig']).str.lower().str.strip()\n",
    "            logging.info(f\"Created 'contig_key' for {file}.\")\n",
    "\n",
    "            # Convert numeric columns\n",
    "            numeric_columns = ['identity', 'align_length', 'bit_score', 'evalue']\n",
    "            for col in numeric_columns:\n",
    "                if col in df.columns:\n",
    "                    df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in numeric fields (optional)\n",
    "            before_drop_numeric = len(df)\n",
    "            df = df.dropna(subset=numeric_columns)\n",
    "            dropped_numeric = before_drop_numeric - len(df)\n",
    "            if dropped_numeric > 0:\n",
    "                logging.info(f\"Dropped {dropped_numeric} rows with non-numeric values in {file}.\")\n",
    "\n",
    "            # Clean 'contig' by removing trailing dots and whitespace\n",
    "            df['contig'] = df['contig'].str.rstrip('.').str.strip()\n",
    "\n",
    "            dfs.append(df)\n",
    "\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error reading {file}: {e}\")\n",
    "            continue\n",
    "\n",
    "    combined_df = pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()\n",
    "    logging.info(f\"Combined DataFrame has {len(combined_df)} rows and {len(combined_df.columns)} columns.\")\n",
    "    logging.info(f\"Columns in combined DataFrame: {combined_df.columns.tolist()}\")\n",
    "    return combined_df\n",
    "\n",
    "\n",
    "def get_sample_from_path(path):\n",
    "    \"\"\"\n",
    "    Extract the sample name from the file path.\n",
    "    Assumes filenames are like: LCA_<SAMPLE>_modified_diamond_blast_nr_annotated.m9\n",
    "    \"\"\"\n",
    "    filename = os.path.basename(path)\n",
    "    prefix = \"LCA_\"\n",
    "    suffix = \"_modified_diamond_blast_nr_annotated.m9\"\n",
    "    if filename.startswith(prefix) and filename.endswith(suffix):\n",
    "        sample = filename[len(prefix):-len(suffix)]\n",
    "        return sample\n",
    "    else:\n",
    "        logging.warning(f\"Filename {filename} does not match expected pattern. Assigning 'unknown_sample'.\")\n",
    "        return 'unknown_sample'\n",
    "\n",
    "\n",
    "def extract_contig(query, sample):\n",
    "    \"\"\"\n",
    "    Extract contig information by splitting the query string using '~'.\n",
    "    Remove trailing dots if any.\n",
    "\n",
    "    Parameters:\n",
    "        query (str): The query string from BLAST output.\n",
    "        sample (str): The sample name.\n",
    "\n",
    "    Returns:\n",
    "        str: Extracted contig information or 'unknown_contig'.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        sample_from_query, contig = query.split('~', 1)\n",
    "        if sample_from_query.lower() == sample.lower():\n",
    "            contig = contig.rstrip('.').strip()  # Remove trailing dots and whitespace\n",
    "            if contig:\n",
    "                return contig\n",
    "            else:\n",
    "                logging.warning(f\"Contig extracted from query '{query}' is empty after cleaning.\")\n",
    "                return 'unknown_contig'\n",
    "        else:\n",
    "            logging.warning(\n",
    "                f\"Sample in query '{sample_from_query}' does not match expected '{sample}'. Using 'unknown_contig'.\"\n",
    "            )\n",
    "            return 'unknown_contig'\n",
    "    except ValueError:\n",
    "        logging.warning(f\"Query '{query}' does not contain '~'. Using 'unknown_contig'.\")\n",
    "        return 'unknown_contig'\n",
    "\n",
    "\n",
    "# -------------------------------------------\n",
    "# The rest of your functions remain unchanged\n",
    "# -------------------------------------------\n",
    "\n",
    "def load_cdhit_clusters(filename):\n",
    "    \"\"\"\n",
    "    Load CD-HIT cluster information from a cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        filename (str): Path to the CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        defaultdict: Cluster ID mapped to list of members.\n",
    "    \"\"\"\n",
    "    clusters = defaultdict(list)\n",
    "    cluster_id = None\n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>Cluster'):\n",
    "                cluster_id = line.strip().split()[-1]\n",
    "            else:\n",
    "                if cluster_id is not None:\n",
    "                    try:\n",
    "                        member = parse_cdhit_row(line)\n",
    "                        clusters[cluster_id].append(member)\n",
    "                    except ValueError as e:\n",
    "                        logging.error(f\"Error parsing line: {line.strip()} - {e}\")\n",
    "    return clusters\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    \"\"\"\n",
    "    Parse a single line from a CD-HIT cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        row (str): Line from CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        dict: Parsed member information.\n",
    "    \"\"\"\n",
    "    if '*' in row:\n",
    "        # Reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = True\n",
    "    else:\n",
    "        # Non-reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = False\n",
    "\n",
    "    # Clean and extract fields\n",
    "    # Use regex to extract numeric length\n",
    "    match = re.match(r'(\\d+)', length_str)\n",
    "    if match:\n",
    "        length = int(match.group(1))\n",
    "    else:\n",
    "        logging.warning(f\"Could not extract numeric length from '{length_str}' in row: {row}\")\n",
    "        length = 0  # Assign a default value or handle as needed\n",
    "\n",
    "    name = name.strip('>').strip('.')\n",
    "\n",
    "    # Extract sample and contig\n",
    "    try:\n",
    "        sample, contig = name.split('~')\n",
    "    except ValueError:\n",
    "        # Handle unexpected naming format\n",
    "        sample, contig = 'unknown_sample', name  # Assign a default sample name\n",
    "\n",
    "    return {\n",
    "        'contig': contig,\n",
    "        'sample': sample,\n",
    "        'length': length,\n",
    "        'is_ref': is_ref\n",
    "    }\n",
    "\n",
    "def get_cluster_rep(cluster):\n",
    "    \"\"\"\n",
    "    Get the representative member of a cluster (the reference member).\n",
    "\n",
    "    Parameters:\n",
    "        cluster (list): List of cluster members.\n",
    "\n",
    "    Returns:\n",
    "        dict: Representative member.\n",
    "    \"\"\"\n",
    "    for member in cluster:\n",
    "        if member['is_ref']:\n",
    "            return member\n",
    "    # If no reference member, return the first member\n",
    "    return cluster[0] if cluster else None\n",
    "\n",
    "def merge_clusters_lca(clusters, lca_df):\n",
    "    \"\"\"\n",
    "    Merge cluster information with LCA results.\n",
    "\n",
    "    Parameters:\n",
    "        clusters (defaultdict): Cluster information.\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Merged cluster and LCA information.\n",
    "    \"\"\"\n",
    "    # Prepare cluster data\n",
    "    cluster_ids = [int(id.strip()) for id in clusters.keys()]\n",
    "    cluster_reps = [get_cluster_rep(clusters[str(id)]) for id in cluster_ids]\n",
    "    cluster_lengths = [rep['length'] for rep in cluster_reps]\n",
    "    cluster_sizes = [len(clusters[str(id)]) for id in cluster_ids]\n",
    "    # Correct contig_key formatting using '~'\n",
    "    cluster_contig_keys = [f\"{rep['sample']}~{rep['contig']}\".lower() for rep in cluster_reps]\n",
    "\n",
    "    # Create DataFrame for clusters\n",
    "    df = pd.DataFrame({\n",
    "        'contig_key': cluster_contig_keys,\n",
    "        'cluster': cluster_ids,\n",
    "        'cluster_size': cluster_sizes,\n",
    "        'contig_length': cluster_lengths\n",
    "    })\n",
    "\n",
    "    # Explicitly convert to string, then strip and lowercase\n",
    "    df['contig_key'] = df['contig_key'].astype(str).str.strip().str.lower()\n",
    "\n",
    "    # Debug: Log sample 'contig_key's\n",
    "    logging.info(\"Sample 'contig_key's in clusters DataFrame:\")\n",
    "    logging.info(df['contig_key'].head())\n",
    "\n",
    "    # Merge with LCA DataFrame\n",
    "    merged_df = df.merge(lca_df, on='contig_key', how='inner')\n",
    "\n",
    "    # Debug: Log the number of merged rows\n",
    "    logging.info(f\"Merged DataFrame has {len(merged_df)} rows.\")\n",
    "\n",
    "    if merged_df.empty:\n",
    "        logging.warning(\"merged_df is empty after merging. Check 'contig_key' consistency.\")\n",
    "        print(\"merged_df is empty after merging. Please check 'contig_key' consistency between clusters and LCA DataFrames.\", file=sys.stderr)\n",
    "    else:\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "\n",
    "    # Calculate align_percent\n",
    "    merged_df['align_percent'] = np.round(100 * merged_df['align_length'] / merged_df['contig_length'], 1)\n",
    "\n",
    "    # Use 'identity' as 'percent_identity'\n",
    "    merged_df['percent_identity'] = np.round(merged_df['identity'], 1)\n",
    "\n",
    "    # Define required columns (including additional columns)\n",
    "    required_columns = [\n",
    "        'cluster', 'cluster_size', 'contig_length', 'name', 'rank', \n",
    "        'kingdom', 'taxid', 'db', 'align_length', 'align_percent', \n",
    "        'percent_identity', 'bit_score', 'evalue', 'subject', 'sci_name',\n",
    "        'com_names', 'superkingdom', 'phylum', 'family', 'lca_taxid', 'lca_name',\n",
    "        'contig', 'sample', 'contig_key'\n",
    "    ]\n",
    "\n",
    "    # Check for missing columns\n",
    "    missing_columns = [col for col in required_columns if col not in merged_df.columns]\n",
    "    if missing_columns:\n",
    "        logging.error(f\"Missing columns in merged DataFrame: {missing_columns}\")\n",
    "        logging.info(f\"Available columns: {merged_df.columns.tolist()}\")\n",
    "        raise KeyError(f\"{missing_columns} not in index\")\n",
    "\n",
    "    # Subset the DataFrame to include only required columns\n",
    "    merged_df = merged_df[required_columns]\n",
    "\n",
    "    # Sort by cluster\n",
    "    merged_df = merged_df.sort_values('cluster').reset_index(drop=True)\n",
    "\n",
    "    return merged_df\n",
    "\n",
    "def load_counts_data(counts_file):\n",
    "    \"\"\"\n",
    "    Load and process the counts DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        counts_file (str): Path to the counts file.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed counts DataFrame with 'contig_key'.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Read counts file with header, assuming first row is header\n",
    "        counts = pd.read_csv(\n",
    "            counts_file, \n",
    "            sep=',', \n",
    "            header=0,  # Assumes first row is header\n",
    "            usecols=[0, 1, 2],  # Read 'contig', 'count', 'length'\n",
    "            names=['contig', 'count', 'length'],  # Assign column names\n",
    "            dtype={'contig': str, 'count': str, 'length': str},  # Read as strings initially\n",
    "            na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "            low_memory=False\n",
    "        )\n",
    "        \n",
    "        # Log a sample of the counts DataFrame\n",
    "        logging.info(\"Raw Counts DataFrame sample:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Convert 'count' to numeric, coercing errors to NaN\n",
    "        counts['count'] = pd.to_numeric(counts['count'], errors='coerce')\n",
    "        \n",
    "        # Log the conversion result\n",
    "        logging.info(\"Counts DataFrame after converting 'count' to numeric:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Extract sample and contig from 'contig' column assuming 'sample~contig'\n",
    "        counts[['sample', 'contig']] = counts['contig'].str.split('~', expand=True)\n",
    "        \n",
    "        # Handle cases where 'contig' does not contain '~'\n",
    "        counts['contig'] = counts['contig'].fillna(counts['contig'])\n",
    "        counts['sample'] = counts['sample'].fillna('unknown_sample')\n",
    "        \n",
    "        # Clean 'contig' by removing trailing dots and whitespace\n",
    "        counts['contig'] = counts['contig'].str.rstrip('.').str.strip()\n",
    "        \n",
    "        # Create 'contig_key' as 'sample~contig', lowercase\n",
    "        counts['contig_key'] = (counts['sample'].astype(str) + '~' + counts['contig'].astype(str)).str.lower().str.strip()\n",
    "        \n",
    "        # Debug: Log the processed counts DataFrame\n",
    "        logging.info(\"Processed Counts DataFrame:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        return counts\n",
    "    except pd.errors.ParserError as e:\n",
    "        logging.error(f\"ParserError while reading counts file {counts_file}: {e}\")\n",
    "        print(f\"ParserError while reading counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error while processing counts file {counts_file}: {e}\")\n",
    "        print(f\"Unexpected error while processing counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "def load_taxid2name(lca_df):\n",
    "    \"\"\"\n",
    "    Create a taxid to name mapping from the LCA DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        dict: Mapping from taxid to scientific name.\n",
    "    \"\"\"\n",
    "    taxid2name = dict(zip(lca_df['taxid'], lca_df['name']))\n",
    "    taxid2name[-1] = 'Ambiguous'  # Handle ambiguous taxids\n",
    "    taxid2name[0] = 'Unknown'     # Handle 'Unknown' taxids\n",
    "    return taxid2name\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Example usage\n",
    "    try:\n",
    "        lca_df = load_lca(nr_dir)\n",
    "        if lca_df.empty:\n",
    "            logging.error(\"LCA DataFrame is empty. Exiting.\")\n",
    "            raise ValueError(\"LCA DataFrame is empty.\")\n",
    "        logging.info(f\"LCA DataFrame loaded: {lca_df.shape[0]} rows\")\n",
    "        print(lca_df.head(3))\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load LCA DataFrame: {e}\")\n",
    "        sys.exit(1)\n",
    "        \n",
    "    try:\n",
    "        # Load CD-HIT clusters\n",
    "        clusters = load_cdhit_clusters('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Combined_filtered_500_SENZOR_clsuter.clstr')\n",
    "        logging.info(\"Loaded CD-HIT clusters.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load CD-HIT clusters: {e}\")\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge clusters with LCA\n",
    "        merged_df = merge_clusters_lca(clusters, lca_df)\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "    except KeyError as e:\n",
    "        logging.error(f\"KeyError during merging: {e}\")\n",
    "        print(f\"Error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error during merging: {e}\")\n",
    "        print(f\"Unexpected error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Create clust2kingdom dictionary\n",
    "        clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "        logging.info(\"Created clust2kingdom dictionary.\")\n",
    "\n",
    "        # Save clust2kingdom to CSV\n",
    "        merged_df[['cluster', 'kingdom']].to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv', index=False)\n",
    "        logging.info(\"Saved cluster_kingdoms.csv successfully.\")\n",
    "\n",
    "        # Print sample entries of clust2kingdom\n",
    "        print(\"clust2kingdom Dictionary Sample Entries:\")\n",
    "        for cluster_id, kingdom in list(clust2kingdom.items())[:5]:\n",
    "            print(f\"Cluster {cluster_id}: Kingdom {kingdom}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error creating clust2kingdom: {e}\")\n",
    "        print(f\"Error creating clust2kingdom: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Load and process the counts DataFrame\n",
    "        counts_file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/combined_SENSOR_csvs/1Modified_CSV/merged_contig_stats.csv'\n",
    "        counts_df = load_counts_data(counts_file_path)\n",
    "        logging.info(\"Loaded and processed counts DataFrame.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load counts DataFrame: {e}\")\n",
    "        print(f\"Error loading counts DataFrame: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge counts with merged_df on 'contig_key'\n",
    "        final_df = merged_df.merge(counts_df[['contig_key', 'count']], on='contig_key', how='left')\n",
    "        logging.info(\"Merged counts with merged_df to create final_df.\")\n",
    "\n",
    "        # Convert 'count' to numeric if not already (redundant but safe)\n",
    "        final_df['count'] = pd.to_numeric(final_df['count'], errors='coerce')\n",
    "\n",
    "        # Apply filtering: Remove rows where 'count' is NaN\n",
    "        initial_count = len(final_df)\n",
    "        final_df = final_df.dropna(subset=['count'])\n",
    "        final_count = len(final_df)\n",
    "        removed_rows = initial_count - final_count\n",
    "        if removed_rows > 0:\n",
    "            logging.info(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "            print(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "\n",
    "        # Remove rows where 'count' is zero or negative:\n",
    "        before_zero_drop = len(final_df)\n",
    "        final_df = final_df[final_df['count'] > 0]\n",
    "        after_zero_drop = len(final_df)\n",
    "        removed_zero = before_zero_drop - after_zero_drop\n",
    "        if removed_zero > 0:\n",
    "            logging.info(f\"Filtered out {removed_zero} rows with 'count' <= 0.\")\n",
    "            print(f\"Filtered out {removed_zero} rows with 'count' <= 0.\")\n",
    "\n",
    "        # Verify if final_df is empty after filtering\n",
    "        if final_df.empty:\n",
    "            logging.warning(\"final_df is empty after filtering out rows without 'count'.\")\n",
    "            print(\"final_df is empty after filtering out rows without 'count'.\", file=sys.stderr)\n",
    "        else:\n",
    "            # Display the final merged DataFrame with counts\n",
    "            print(\"Final Merged DataFrame with Counts (after filtering):\")\n",
    "            print(final_df.head())\n",
    "\n",
    "            # Save the final DataFrame to CSV\n",
    "            final_df.to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/final_merged_with_counts.csv', index=False)\n",
    "            logging.info(\"Saved final_merged_with_counts.csv successfully.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error merging counts with merged_df: {e}\")\n",
    "        print(f\"Error merging counts with merged_df: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "986c9352-da29-4cfe-a2e7-b3e63d6f3d61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found NR LCA files: ['/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_GOT005_OS_S43_673459_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_GOT006_RS_S44_673460_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_GOT007_OS_S45_673461_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_LIZ002_OS_S50_673462_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_PIG003_RS_S51_673463_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_SHP004_OS_S46_673464_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_SHP005_RS_S47_673465_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMA_SHP006_OS_S48_673466_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAAGR001_LIV_S46_662105_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMACAT001_S69_657769_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMADOG001_S40_657732_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMADOG002_OS_S64_673467_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAGOAT001_S48_656109_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAGOAT002_S49_656110_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAGOAT003_S50_657720_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAGOAT004_S51_656112_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAPIG001_S63_657713_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAPIG002_S64_657714_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR001_LIV_S21_662106_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR002_LIV_S22_662107_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR003_LIV_S23_662108_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR004_LIV_S24_1_662161_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR005_LIV_S25_1_662162_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR006_LIV_S26_2_662165_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR007_LIV_S27_2_662166_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR008_LIV_S28_1_662167_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR009_LIV_S29_1_662168_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR010_LIV_S30_663038_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR011_LIV_S31_663039_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR012_LIV_S32_663040_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR013_LIV_S33_663041_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR014_LIV_S34_663042_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR015_LIV_S35_663046_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR016_LIV_S36_663047_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR017_LIV_S37_663048_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR018_LIV_S38_663049_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR019_LIV_S39_663050_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMAR020_LIV_S40_663051_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMASHEEP001_S53_656078_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIAMASHEEP002_S54_656079_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMASHEEP003_S55_656080_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGBR021_LIV_S41_663052_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGBR022_LIV_S42_663077_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGBR023_LIV_S43_662128_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGBR024_LIV_S44_662129_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGL_GOT003_OS_S65_673468_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGL_GOT004_RS_S66_673469_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGLDOG001_S59_657733_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGLGOAT001_S60_656113_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMGLGOAT002_S61_656114_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AIMUZSHEEP001_S52_656081_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_CHK003_CS_S49_673471_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_DOG004_RS_S40_673472_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_DOG005_OS_S41_673473_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_DOG006_RS_S42_674096_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT008_OS_S67_673475_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT009_RS_S68_673476_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT010_OS_S69_673481_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT011_RS_S37_673482_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT012_RS_S38_673479_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZ_GOT013_OS_S39_673480_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZAGR001_LIV_S45_662130_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZCHICKEN001_S56_655747_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZCHICKEN002_S57_655748_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZDOG001_S37_657734_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZDOG002_S38_657735_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZDOG003_S39_657736_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT001_S41_656115_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT002_S42_656116_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT003_S43_1_656182_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT004_S44_1_656183_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT005_S45_1_656184_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT006_S46_1_656185_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZGOAT007_S47_2_656739_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZLIZARD001_S58_657718_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZPIG001_S62_657715_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR001_LIV_S1_662131_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR002_LIV_S2_662132_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR003_LIV_S3_662133_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR004_LIV_S4_662134_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR005_LIV_S5_663064_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR006_LIV_S6_663065_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR007_LIV_S7_663066_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR008_LIV_S8_663067_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR009_LIV_S9_663068_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR010_LIV_S10_663069_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR011_LIV_S11_663070_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR012_LIV_S12_663071_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR013_LIV_S13_663072_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR014_LIV_S14_663073_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR015_LIV_S15_663078_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR016_LIV_S16_663079_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR017_LIV_S17_663080_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR018_LIV_S18_663081_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR019_LIV_S19_663082_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_AINWZR020_LIV_S20_663083_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_NE_S70_670946_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_NE_S82_663084_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_NE_S82_673483_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_NTC_S71_670947_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR001_LIV_S31_673484_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR002_LIV_S32_673485_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR003_LIV_S33_673487_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR004_LIV_S34_673490_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR005_LIV_S35_673488_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR006_LUNG_S36_673489_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR007_LUNG_S52_673492_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR008_LUNG_S53_673493_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR009_LUNG_S54_673491_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_AGR010_LUNG_S55_674094_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT020_LIV_S21_673503_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT021_LIV_S22_674095_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT023_LIV_S24_673504_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT024_LIV_S25_673505_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT025_LUNG_S26_673506_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT026_LUNG_S27_673507_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT027_LUNG_S28_673508_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT028_LUNG_S29_673509_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_BAT029_LUNG_S30_673510_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG031_LIV_S1_673511_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG032_LIV_S2_673512_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG033_LIV_S3_673513_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG034_LIV_S4_673514_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG035_LIV_S5_673515_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG036_LUNG_S6_673516_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG037_LUNG_S7_673517_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG038_LUNG_S8_673518_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG039_LUNG_S9_673615_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_CEG040_LUNG_S10_673616_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL021_LIV_S11_673617_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL022_LIV_S12_673618_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL023_LIV_S13_673619_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL024_LIV_S14_673620_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL025_LIV_S15_673621_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL026_LUNG_S16_673622_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL027_LUNG_S17_673623_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL028_LUNG_S18_673624_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL029_LUNG_S19_673625_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_OD_ZAD_SQL030_LUNG_S20_673626_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWGOT003_RS_S79_673627_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWGOT014_OS_S75_673628_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWIFGOT005_RS_S78_673657_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT001_OS_S80_673658_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT001_RS_S74_673659_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT002_OS_S76_673660_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT004_RS_S77_673661_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT005_RS_S71_673662_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWISUGOT006_OS_S73_673663_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWOGGOT016_RS_S81_673664_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SENONOWOGGOT017_OS_S72_673665_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWIS_R001_S79_687907_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWIS_R002_S80_687908_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWIS_R003_S81_687909_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWIS_R004_S82_687910_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWOG_R001_S75_687911_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWOG_R002_S76_687912_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWOG_R003_S77_687913_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZ_ON_OWOG_R004_S78_687914_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWDOG009_S20_657738_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWDOG012_S10_657739_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWDOG01_S7_657737_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWGOAT009_S67_2_656740_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWGOAT010_S65_2_656741_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFCAT001_S34_657717_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFDOG009_S1_657740_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFGOAT007_S68_657731_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFGOAT011_S29_2_656747_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFGOAT012_S30_657726_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFGOT007_RS_S70_673666_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR031_LIV_S71_663085_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR032_LIV_S72_666406_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR033_LIV_S73_666402_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR034_LIV_S74_1_666404_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR037_LIV_S77_666399_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR038_LIV_S78_666400_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR039_LIV_S79_666401_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR040_LIV_S80_666396_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR063_LUNG_S81_666397_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR064_LUNG_S56_673522_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR066_LUNG_S57_673523_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR067_LUNG_S58_673524_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR068_LUNG_S59_673525_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR069_LUNG_S60_673526_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR070_LUNG_S61_673529_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR071_LUNG_S62_673527_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFRR072_LUNG_S63_673528_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFSHEEP001_S33_656082_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWIFSHEEP002_S31_656083_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISDOG002_S4_657741_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISDOG008_S14_657742_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISDOG015_S13_657743_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISDOGOO1_S19_657744_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISGOAT008_S25_657722_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISGOAT013_S24_2_657729_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISLIZARD001_S36_657719_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISSHEEP001_S32_656084_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWISUGOAT003_S26_657724_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG001_S6_657746_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG002_S5_657747_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG004_S12_657748_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG00_S2_657745_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG010_S3_657749_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG013_S18_657750_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG014_S21_657751_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG016_S23_657752_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGDOG017_S22_657753_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGGOAT015_S28_657725_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGGOAT018_S66_1_657730_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR001_LIV_S47_666398_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR002_LIV_S48_666373_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR003_LIV_S49_666374_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR004_LIV_S50_666375_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR005_LIV_S51_666376_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR006_LIV_S52_666377_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR007_LIV_S53_666370_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR008_LIV_S54_666371_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR009_LIV_S55_666372_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR010_LIV_S56_666407_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR081_LUNG_S57_666409_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR082_LUNG_S58_666410_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR083_LUNG_S59_3_666413_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOGR085_LUNG_S61_666414_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR001_LIV_S62_666415_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR002_LIV_S63_666416_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR003_LIV_S64_666417_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR004_LIV_S65_666418_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR005_LIV_S66_666419_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR086_LIV_S67_666422_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR087_LUNG_S68_666423_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR088_LIV_S69_666424_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZONOWOWISR089_LUNG_S70_666425_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWISDOG003_S11_657754_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWISUGOAT002_S27_657721_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGCHICKEN001_S35_655749_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGDOG003_S17_657755_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGDOG005_S15_657756_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGDOG006_S16_1_657760_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGDOG011_S8_1_657761_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SEZOOWOGDOG019_S9_1_657762_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SSP_ERCC_S72_670948_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SSP_ERCC_S83_666426_modified_diamond_blast_nr_annotated.m9', '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/LCA_SSP_ERCC_S83_673530_modified_diamond_blast_nr_annotated.m9']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py:57: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'tuple' object has no attribute 'tb_frame'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m    111\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m         \u001b[0mlogging\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Failed to load LCA DataFrame: {e}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 113\u001b[0;31m         \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/3297609138.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(lca_directory, nr_files)\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;31m# Load BLAST outputs with NR database\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0mnr_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_blast_output\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnr_file_paths\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatabase\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'nr'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m     \u001b[0mnr_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnr_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'taxid'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# remove rows with missing taxid\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m     \u001b[0mnr_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'taxid'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnr_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'taxid'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, axis, how, thresh, subset, inplace, ignore_index)\u001b[0m\n\u001b[1;32m   6668\u001b[0m             \u001b[0mcheck\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindices\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6669\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6670\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcheck\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6671\u001b[0m             \u001b[0magg_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0magg_axis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ['taxid']",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSystemExit\u001b[0m                                Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "Cell \u001b[0;32mIn[55], line 113\u001b[0m\n\u001b[1;32m    112\u001b[0m     logging\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFailed to load LCA DataFrame: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 113\u001b[0m     sys\u001b[38;5;241m.\u001b[39mexit(\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m    115\u001b[0m \u001b[38;5;66;03m# --- Next block (same indentation as above) ---\u001b[39;00m\n",
      "\u001b[0;31mSystemExit\u001b[0m: 1",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/interactiveshell.py:2121\u001b[0m, in \u001b[0;36mInteractiveShell.showtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[1;32m   2118\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m exception_only:\n\u001b[1;32m   2119\u001b[0m     stb \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAn exception has occurred, use \u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mtb to see \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m   2120\u001b[0m            \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthe full traceback.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m-> 2121\u001b[0m     stb\u001b[38;5;241m.\u001b[39mextend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mInteractiveTB\u001b[38;5;241m.\u001b[39mget_exception_only(etype,\n\u001b[1;32m   2122\u001b[0m                                                      value))\n\u001b[1;32m   2123\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   2125\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcontains_exceptiongroup\u001b[39m(val):\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:710\u001b[0m, in \u001b[0;36mListTB.get_exception_only\u001b[0;34m(self, etype, value)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_exception_only\u001b[39m(\u001b[38;5;28mself\u001b[39m, etype, value):\n\u001b[1;32m    703\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Only print the exception type and message, without a traceback.\u001b[39;00m\n\u001b[1;32m    704\u001b[0m \n\u001b[1;32m    705\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    708\u001b[0m \u001b[38;5;124;03m    value : exception value\u001b[39;00m\n\u001b[1;32m    709\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 710\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ListTB\u001b[38;5;241m.\u001b[39mstructured_traceback(\u001b[38;5;28mself\u001b[39m, etype, value)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:568\u001b[0m, in \u001b[0;36mListTB.structured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, context)\u001b[0m\n\u001b[1;32m    565\u001b[0m     chained_exc_ids\u001b[38;5;241m.\u001b[39madd(\u001b[38;5;28mid\u001b[39m(exception[\u001b[38;5;241m1\u001b[39m]))\n\u001b[1;32m    566\u001b[0m     chained_exceptions_tb_offset \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m    567\u001b[0m     out_list \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 568\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstructured_traceback(\n\u001b[1;32m    569\u001b[0m             etype,\n\u001b[1;32m    570\u001b[0m             evalue,\n\u001b[1;32m    571\u001b[0m             (etb, chained_exc_ids),  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m    572\u001b[0m             chained_exceptions_tb_offset,\n\u001b[1;32m    573\u001b[0m             context,\n\u001b[1;32m    574\u001b[0m         )\n\u001b[1;32m    575\u001b[0m         \u001b[38;5;241m+\u001b[39m chained_exception_message\n\u001b[1;32m    576\u001b[0m         \u001b[38;5;241m+\u001b[39m out_list)\n\u001b[1;32m    578\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m out_list\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:1435\u001b[0m, in \u001b[0;36mAutoFormattedTB.structured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1433\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1434\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtb \u001b[38;5;241m=\u001b[39m etb\n\u001b[0;32m-> 1435\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m FormattedTB\u001b[38;5;241m.\u001b[39mstructured_traceback(\n\u001b[1;32m   1436\u001b[0m     \u001b[38;5;28mself\u001b[39m, etype, evalue, etb, tb_offset, number_of_lines_of_context\n\u001b[1;32m   1437\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:1326\u001b[0m, in \u001b[0;36mFormattedTB.structured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1323\u001b[0m mode \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmode\n\u001b[1;32m   1324\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mode \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mverbose_modes:\n\u001b[1;32m   1325\u001b[0m     \u001b[38;5;66;03m# Verbose modes need a full traceback\u001b[39;00m\n\u001b[0;32m-> 1326\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m VerboseTB\u001b[38;5;241m.\u001b[39mstructured_traceback(\n\u001b[1;32m   1327\u001b[0m         \u001b[38;5;28mself\u001b[39m, etype, value, tb, tb_offset, number_of_lines_of_context\n\u001b[1;32m   1328\u001b[0m     )\n\u001b[1;32m   1329\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m mode \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMinimal\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m   1330\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ListTB\u001b[38;5;241m.\u001b[39mget_exception_only(\u001b[38;5;28mself\u001b[39m, etype, value)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:1173\u001b[0m, in \u001b[0;36mVerboseTB.structured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1164\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mstructured_traceback\u001b[39m(\n\u001b[1;32m   1165\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1166\u001b[0m     etype: \u001b[38;5;28mtype\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1170\u001b[0m     number_of_lines_of_context: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m5\u001b[39m,\n\u001b[1;32m   1171\u001b[0m ):\n\u001b[1;32m   1172\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a nice text document describing the traceback.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1173\u001b[0m     formatted_exception \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mformat_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n\u001b[1;32m   1174\u001b[0m                                                            tb_offset)\n\u001b[1;32m   1176\u001b[0m     colors \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mColors  \u001b[38;5;66;03m# just a shorthand + quicker name lookup\u001b[39;00m\n\u001b[1;32m   1177\u001b[0m     colorsnormal \u001b[38;5;241m=\u001b[39m colors\u001b[38;5;241m.\u001b[39mNormal  \u001b[38;5;66;03m# used a lot\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:1063\u001b[0m, in \u001b[0;36mVerboseTB.format_exception_as_a_whole\u001b[0;34m(self, etype, evalue, etb, number_of_lines_of_context, tb_offset)\u001b[0m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(tb_offset, \u001b[38;5;28mint\u001b[39m)\n\u001b[1;32m   1061\u001b[0m head \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprepare_header(\u001b[38;5;28mstr\u001b[39m(etype), \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlong_header)\n\u001b[1;32m   1062\u001b[0m records \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m-> 1063\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_records(etb, number_of_lines_of_context, tb_offset) \u001b[38;5;28;01mif\u001b[39;00m etb \u001b[38;5;28;01melse\u001b[39;00m []\n\u001b[1;32m   1064\u001b[0m )\n\u001b[1;32m   1066\u001b[0m frames \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m   1067\u001b[0m skipped \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/IPython/core/ultratb.py:1131\u001b[0m, in \u001b[0;36mVerboseTB.get_records\u001b[0;34m(self, etb, number_of_lines_of_context, tb_offset)\u001b[0m\n\u001b[1;32m   1129\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m cf \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1130\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1131\u001b[0m         mod \u001b[38;5;241m=\u001b[39m inspect\u001b[38;5;241m.\u001b[39mgetmodule(cf\u001b[38;5;241m.\u001b[39mtb_frame)\n\u001b[1;32m   1132\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m mod \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1133\u001b[0m             mod_name \u001b[38;5;241m=\u001b[39m mod\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'tuple' object has no attribute 'tb_frame'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from glob import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from statistics import mode\n",
    "from Bio import Entrez\n",
    "from ete3 import NCBITaxa\n",
    "import logging\n",
    "import sys\n",
    "import re\n",
    "\n",
    "# Configure logging\n",
    "from logging.handlers import RotatingFileHandler\n",
    "handler = RotatingFileHandler('util.log', maxBytes=5*1024*1024, backupCount=2)\n",
    "logging.basicConfig(level=logging.INFO, handlers=[handler],\n",
    "                    format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "Entrez.email = \"ifeanyi.omah@ed.ac.uk\"\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "# Define column names for NR\n",
    "nr_column_names = [\n",
    "    'query', 'lca_taxid', 'best_hit_sciname', 'subject_title',\n",
    "    'identity', 'aligned_bases', 'contig_length', 'superkingdom',\n",
    "    'bitscore', 'tax_id'\n",
    "]import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "from statistics import mode\n",
    "from Bio import Entrez\n",
    "from ete3 import NCBITaxa\n",
    "import logging\n",
    "import sys\n",
    "import re\n",
    "from logging.handlers import RotatingFileHandler\n",
    "\n",
    "handler = RotatingFileHandler('util.log', maxBytes=5*1024*1024, backupCount=2)\n",
    "logging.basicConfig(level=logging.INFO, handlers=[handler],\n",
    "                    format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "Entrez.email = \"ifeanyi.omah@ed.ac.uk\"\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "valid_lineages = set()\n",
    "\n",
    "# Updated column names to match your .m9 file\n",
    "nr_column_names = [\n",
    "    'query', 'subject', 'identity', 'alignment_length', 'mismatches', 'gap_opens',\n",
    "    'q_start', 'q_end', 's_start', 's_end', 'evalue', 'bit_score', 'tax_id',\n",
    "    'sci_name', 'com_names', 'superkingdom', 'phylum', 'family',\n",
    "    'lca_taxid', 'lca_name'\n",
    "]\n",
    "\n",
    "# --------------------------------------------------------------------\n",
    "# 1) Change nr_dir to where your NR LCA files actually live\n",
    "#    If these files are named like: LCA_<SAMPLE>_modified_diamond_blast_nr_annotated.m9\n",
    "#    then we can glob them below.\n",
    "# --------------------------------------------------------------------\n",
    "nr_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/\"\n",
    "\n",
    "def load_lca(nr_directory):\n",
    "    \"\"\"\n",
    "    Load and process LCA results from NR files in the specified directory.\n",
    "    Dynamically glob any pattern you want.\n",
    "    \"\"\"\n",
    "    # ----------------------------------------------------------------\n",
    "    # 2) Dynamically pick up files named LCA_*_modified_diamond_blast_nr_annotated.m9\n",
    "    #    Adjust the pattern to match your actual file naming.\n",
    "    # ----------------------------------------------------------------\n",
    "    pattern = \"LCA_*_modified_diamond_blast_nr_annotated.m9\"\n",
    "    nr_file_paths = glob(os.path.join(nr_directory, pattern))\n",
    "    logging.info(f\"Found {len(nr_file_paths)} NR files matching '{pattern}' in {nr_directory}.\")\n",
    "\n",
    "    if not nr_file_paths:\n",
    "        logging.error(f\"No files matching {pattern} found in directory: {nr_directory}\")\n",
    "        raise FileNotFoundError(f\"No files matching {pattern} found in directory: {nr_directory}\")\n",
    "\n",
    "    # Process each BLAST output file\n",
    "    nr_df = load_blast_output(nr_file_paths, database='nr')\n",
    "\n",
    "    # Remove rows with missing taxid\n",
    "    initial_nr = len(nr_df)\n",
    "    nr_df = nr_df.dropna(subset=['taxid'])  # <- Make sure 'taxid' actually exists\n",
    "    final_nr = len(nr_df)\n",
    "    dropped_nr = initial_nr - final_nr\n",
    "    if dropped_nr > 0:\n",
    "        logging.info(f\"Dropped {dropped_nr} rows with missing taxid from nr_df.\")\n",
    "\n",
    "    # Convert taxid to integer\n",
    "    nr_df['taxid'] = nr_df['taxid'].astype(int)\n",
    "\n",
    "    # Build dictionaries for name/rank\n",
    "    taxids = set(nr_df.taxid)\n",
    "    taxids = {t for taxid in taxids for t in ncbi.get_lineage(taxid) if t != 0}\n",
    "\n",
    "    taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    taxid2name[-1] = 'Ambiguous'\n",
    "    taxid2name[0] = 'Unknown'\n",
    "    taxid2rank = ncbi.get_rank(taxids)\n",
    "\n",
    "    # Collect valid lineages\n",
    "    taxids_in_df = set(nr_df['taxid'])\n",
    "    for tx in taxids_in_df:\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(tx)\n",
    "            for t in lineage:\n",
    "                if t != 0:\n",
    "                    valid_lineages.add(t)\n",
    "        except ValueError:\n",
    "            logging.warning(f\"Skipping {tx}: taxid not found in local DB.\")\n",
    "\n",
    "    # Restrict to valid lineages\n",
    "    taxids = valid_lineages\n",
    "    taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    \n",
    "    def taxid2kingdom(taxid):\n",
    "        if taxid == 0:\n",
    "            return 'Unknown'\n",
    "        lineage_dict = ncbi.get_rank(ncbi.get_lineage(taxid))\n",
    "        for k, v in lineage_dict.items():\n",
    "            if v == 'superkingdom':\n",
    "                kingdom = taxid2name.get(k, 'Unknown')\n",
    "                return kingdom\n",
    "        return 'Unknown'\n",
    "\n",
    "    # Map taxid -> name, rank, kingdom\n",
    "    nr_df['name'] = nr_df['taxid'].map(taxid2name)\n",
    "    nr_df['rank'] = nr_df['taxid'].map(taxid2rank)\n",
    "    nr_df['kingdom'] = nr_df['taxid'].apply(taxid2kingdom)\n",
    "    nr_df['db'] = 'nr'  # Since this is the NR dataset\n",
    "\n",
    "    return nr_df\n",
    "\n",
    "def load_blast_output(files, database='nr'):\n",
    "    \"\"\"\n",
    "    Load BLAST output files into a single DataFrame.\n",
    "    \"\"\"\n",
    "    if database == 'nr':\n",
    "        column_names = nr_column_names\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        try:\n",
    "            # Quick check of columns in the first line\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                columns_in_file = len(first_line.strip().split('\\t'))\n",
    "                expected_columns = len(column_names)\n",
    "                if columns_in_file != expected_columns:\n",
    "                    logging.warning(f\"{file} has {columns_in_file} columns; expected {expected_columns}.\")\n",
    "\n",
    "            # Potentially skip a header if present\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                if 'sci_name' in first_line.lower() or 'lca_taxid' in first_line.lower():\n",
    "                    skiprows = 1\n",
    "                else:\n",
    "                    skiprows = 0\n",
    "\n",
    "            df = pd.read_csv(\n",
    "                file,\n",
    "                sep='\\t',\n",
    "                header=None,\n",
    "                names=column_names,\n",
    "                na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "                skiprows=skiprows,\n",
    "                on_bad_lines='warn'\n",
    "            )\n",
    "\n",
    "            if 'tax_id' not in df.columns:\n",
    "                # If your data is guaranteed to have 'tax_id', raise KeyError.\n",
    "                logging.error(f\"'tax_id' column is missing in {file}.\")\n",
    "                raise KeyError(f\"'tax_id' column is missing in {file}.\")\n",
    "\n",
    "            # Rename 'tax_id' -> 'taxid'\n",
    "            df.rename(columns={'tax_id': 'taxid'}, inplace=True)\n",
    "            df['taxid'] = df['taxid'].replace('Unknown', '0')\n",
    "\n",
    "            # Convert taxid to numeric\n",
    "            df['taxid'] = pd.to_numeric(df['taxid'], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in 'taxid'\n",
    "            before_drop = len(df)\n",
    "            df.dropna(subset=['taxid'], inplace=True)\n",
    "            after_drop = len(df)\n",
    "            if (d := (before_drop - after_drop)) > 0:\n",
    "                logging.info(f\"Dropped {d} rows with invalid taxid in {file}.\")\n",
    "\n",
    "            # alignment_length -> align_length\n",
    "            if 'alignment_length' in df.columns:\n",
    "                df.rename(columns={'alignment_length': 'align_length'}, inplace=True)\n",
    "                \n",
    "                # Build contig_key\n",
    "                df['contig_key'] = (df['sample'] + '~' + df['contig']).str.lower()\n",
    "\n",
    "                # Convert numeric columns\n",
    "                numeric_columns = ['identity', 'align_length', 'bit_score', 'evalue']\n",
    "                for col in numeric_columns:\n",
    "                    if col in df.columns:\n",
    "                        df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "\n",
    "                # Drop rows with NaN in numeric fields (optional)\n",
    "                before_drop_numeric = len(df)\n",
    "                df = df.dropna(subset=numeric_columns)\n",
    "                dropped_numeric = before_drop_numeric - len(df)\n",
    "                if dropped_numeric > 0:\n",
    "                    logging.info(f\"Dropped {dropped_numeric} rows with non-numeric in {file}.\")\n",
    "\n",
    "            # Build a contig_key if needed\n",
    "            # For example: df['contig_key'] = (df['query'] + '~' + df['subject']).str.lower()\n",
    "            # Or keep your existing method if you rely on a different scheme\n",
    "\n",
    "            dfs.append(df)\n",
    "\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error reading {file}: {e}\")\n",
    "            continue\n",
    "\n",
    "    combined_df = pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()\n",
    "    return combined_df\n",
    "\n",
    "\n",
    "# ----------------------------------------------------------------------\n",
    "# Below are the rest of your original functions (merge_clusters_lca, etc.)\n",
    "# You can keep them exactly as is. The main difference is that now your\n",
    "# `load_lca()` is truly dynamic, picking up all the .m9 files in nr_dir.\n",
    "# ----------------------------------------------------------------------\n",
    "\n",
    "\n",
    "def extract_contig(query, sample):\n",
    "    \"\"\"\n",
    "    Extract contig information by removing the sample prefix from the query.\n",
    "\n",
    "    Parameters:\n",
    "        query (str): The query string from BLAST output.\n",
    "        sample (str): The sample name.\n",
    "\n",
    "    Returns:\n",
    "        str: Extracted contig information or 'unknown_contig'.\n",
    "    \"\"\"\n",
    "    prefix = sample + '_'\n",
    "    if query.startswith(prefix):\n",
    "        return query[len(prefix):]\n",
    "    else:\n",
    "        logging.warning(f\"Query '{query}' does not start with sample prefix '{prefix}'. Using 'unknown_contig'.\")\n",
    "        return 'unknown_contig'\n",
    "\n",
    "# -------------------------------------------\n",
    "# The rest of your functions remain unchanged\n",
    "# -------------------------------------------\n",
    "\n",
    "\n",
    "def load_cdhit_clusters(filename):\n",
    "    \"\"\"\n",
    "    Load CD-HIT cluster information from a cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        filename (str): Path to the CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        defaultdict: Cluster ID mapped to list of members.\n",
    "    \"\"\"\n",
    "    clusters = defaultdict(list)\n",
    "    cluster_id = None\n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>Cluster'):\n",
    "                cluster_id = line.strip().split()[-1]\n",
    "            else:\n",
    "                if cluster_id is not None:\n",
    "                    try:\n",
    "                        member = parse_cdhit_row(line)\n",
    "                        clusters[cluster_id].append(member)\n",
    "                    except ValueError as e:\n",
    "                        logging.error(f\"Error parsing line: {line.strip()} - {e}\")\n",
    "    return clusters\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    \"\"\"\n",
    "    Parse a single line from a CD-HIT cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        row (str): Line from CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        dict: Parsed member information.\n",
    "    \"\"\"\n",
    "    if '*' in row:\n",
    "        # Reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = True\n",
    "    else:\n",
    "        # Non-reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = False\n",
    "\n",
    "    # Clean and extract fields\n",
    "    # Use regex to extract numeric length\n",
    "    match = re.match(r'(\\d+)', length_str)\n",
    "    if match:\n",
    "        length = int(match.group(1))\n",
    "    else:\n",
    "        logging.warning(f\"Could not extract numeric length from '{length_str}' in row: {row}\")\n",
    "        length = 0  # Assign a default value or handle as needed\n",
    "\n",
    "    name = name.strip('>').strip('.')\n",
    "\n",
    "    # Extract sample and contig\n",
    "    try:\n",
    "        sample, contig = name.split('~')\n",
    "    except ValueError:\n",
    "        # Handle unexpected naming format\n",
    "        sample, contig = 'unknown_sample', name  # Assign a default sample name\n",
    "\n",
    "    return {\n",
    "        'contig': contig,\n",
    "        'sample': sample,\n",
    "        'length': length,\n",
    "        'is_ref': is_ref\n",
    "    }\n",
    "\n",
    "def get_cluster_rep(cluster):\n",
    "    \"\"\"\n",
    "    Get the representative member of a cluster (the reference member).\n",
    "\n",
    "    Parameters:\n",
    "        cluster (list): List of cluster members.\n",
    "\n",
    "    Returns:\n",
    "        dict: Representative member.\n",
    "    \"\"\"\n",
    "    for member in cluster:\n",
    "        if member['is_ref']:\n",
    "            return member\n",
    "    # If no reference member, return the first member\n",
    "    return cluster[0] if cluster else None\n",
    "\n",
    "def merge_clusters_lca(clusters, lca_df):\n",
    "    \"\"\"\n",
    "    Merge cluster information with LCA results.\n",
    "\n",
    "    Parameters:\n",
    "        clusters (defaultdict): Cluster information.\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Merged cluster and LCA information.\n",
    "    \"\"\"\n",
    "    # Prepare cluster data\n",
    "    cluster_ids = [int(id.strip()) for id in clusters.keys()]\n",
    "    cluster_reps = [get_cluster_rep(clusters[str(id)]) for id in cluster_ids]\n",
    "    cluster_lengths = [rep['length'] for rep in cluster_reps]\n",
    "    cluster_sizes = [len(clusters[str(id)]) for id in cluster_ids]\n",
    "    # Correct contig_key formatting using '~'\n",
    "    cluster_contig_keys = [f\"{rep['sample']}~{rep['contig']}\".lower() for rep in cluster_reps]\n",
    "\n",
    "    # Create DataFrame for clusters\n",
    "    df = pd.DataFrame({\n",
    "        'contig_key': cluster_contig_keys,\n",
    "        'cluster': cluster_ids,\n",
    "        'cluster_size': cluster_sizes,\n",
    "        'contig_length': cluster_lengths\n",
    "    })\n",
    "\n",
    "    # Explicitly convert to string, then strip and lowercase\n",
    "    df['contig_key'] = df['contig_key'].astype(str).str.strip().str.lower()\n",
    "\n",
    "    # Debug: Log sample 'contig_key's\n",
    "    logging.info(\"Sample 'contig_key's in clusters DataFrame:\")\n",
    "    logging.info(df['contig_key'].head())\n",
    "\n",
    "    # Merge with LCA DataFrame\n",
    "    merged_df = df.merge(lca_df, on='contig_key', how='inner')\n",
    "\n",
    "    # Debug: Log the number of merged rows\n",
    "    logging.info(f\"Merged DataFrame has {len(merged_df)} rows.\")\n",
    "\n",
    "    if merged_df.empty:\n",
    "        logging.warning(\"merged_df is empty after merging. Check 'contig_key' consistency.\")\n",
    "        print(\"merged_df is empty after merging. Please check 'contig_key' consistency between clusters and LCA DataFrames.\", file=sys.stderr)\n",
    "    else:\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "\n",
    "    # Calculate align_percent\n",
    "    merged_df['align_percent'] = np.round(100 * merged_df['align_length'] / merged_df['contig_length'], 1)\n",
    "\n",
    "    # Use 'identity' as 'percent_identity'\n",
    "    merged_df['percent_identity'] = np.round(merged_df['identity'], 1)\n",
    "\n",
    "    # Define required columns (including additional columns)\n",
    "    required_columns = [\n",
    "        'cluster', 'cluster_size', 'contig_length', 'name', 'rank', \n",
    "        'kingdom', 'taxid', 'db', 'align_length', 'align_percent', \n",
    "        'percent_identity', 'bit_score', 'evalue', 'subject', 'sci_name',\n",
    "        'com_names', 'superkingdom', 'phylum', 'family', 'lca_taxid', 'lca_name',\n",
    "        'contig', 'sample', 'contig_key'\n",
    "    ]\n",
    "\n",
    "    # Check for missing columns\n",
    "    missing_columns = [col for col in required_columns if col not in merged_df.columns]\n",
    "    if missing_columns:\n",
    "        logging.error(f\"Missing columns in merged DataFrame: {missing_columns}\")\n",
    "        logging.info(f\"Available columns: {merged_df.columns.tolist()}\")\n",
    "        raise KeyError(f\"{missing_columns} not in index\")\n",
    "\n",
    "    # Subset the DataFrame to include only required columns\n",
    "    merged_df = merged_df[required_columns]\n",
    "\n",
    "    # Sort by cluster\n",
    "    merged_df = merged_df.sort_values('cluster').reset_index(drop=True)\n",
    "\n",
    "    return merged_df\n",
    "\n",
    "def load_counts_data(counts_file):\n",
    "    \"\"\"\n",
    "    Load and process the counts DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        counts_file (str): Path to the counts file.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed counts DataFrame with 'contig_key'.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Read counts file with header, assuming first row is header\n",
    "        counts = pd.read_csv(\n",
    "            counts_file, \n",
    "            sep=',', \n",
    "            header=0,  # Assumes first row is header\n",
    "            usecols=[0, 1, 2],  # Read 'contig', 'count', 'length'\n",
    "            names=['contig', 'count', 'length'],  # Assign column names\n",
    "            dtype={'contig': str, 'count': str, 'length': str},  # Read as strings initially\n",
    "            na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "            low_memory=False\n",
    "        )\n",
    "        \n",
    "        # Log a sample of the counts DataFrame\n",
    "        logging.info(\"Raw Counts DataFrame sample:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Convert 'count' to numeric, coercing errors to NaN\n",
    "        counts['count'] = pd.to_numeric(counts['count'], errors='coerce')\n",
    "        \n",
    "        # Log the conversion result\n",
    "        logging.info(\"Counts DataFrame after converting 'count' to numeric:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Extract sample and contig from 'contig' column assuming 'sample~contig'\n",
    "        counts[['sample', 'contig']] = counts['contig'].str.split('~', expand=True)\n",
    "        \n",
    "        # Handle cases where 'contig' does not contain '~'\n",
    "        counts['contig'] = counts['contig'].fillna(counts['contig'])\n",
    "        counts['sample'] = counts['sample'].fillna('unknown_sample')\n",
    "        \n",
    "        # Create 'contig_key' as 'sample~contig', lowercase\n",
    "        counts['contig_key'] = (counts['sample'].astype(str) + '~' + counts['contig'].astype(str)).str.strip().str.lower()\n",
    "        \n",
    "        # Debug: Log the processed counts DataFrame\n",
    "        logging.info(\"Processed Counts DataFrame:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        return counts\n",
    "    except pd.errors.ParserError as e:\n",
    "        logging.error(f\"ParserError while reading counts file {counts_file}: {e}\")\n",
    "        print(f\"ParserError while reading counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error while processing counts file {counts_file}: {e}\")\n",
    "        print(f\"Unexpected error while processing counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "def load_taxid2name(lca_df):\n",
    "    \"\"\"\n",
    "    Create a taxid to name mapping from the LCA DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        dict: Mapping from taxid to scientific name.\n",
    "    \"\"\"\n",
    "    taxid2name = dict(zip(lca_df['taxid'], lca_df['name']))\n",
    "    taxid2name[-1] = 'Ambiguous'  # Handle ambiguous taxids\n",
    "    taxid2name[0] = 'Unknown'     # Handle 'Unknown' taxids\n",
    "    return taxid2name\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Example usage\n",
    "    try:\n",
    "        lca_df = load_lca(nr_dir)\n",
    "        if lca_df.empty:\n",
    "            logging.error(\"LCA DataFrame is empty. Exiting.\")\n",
    "            raise ValueError(\"LCA DataFrame is empty.\")\n",
    "        logging.info(f\"LCA DataFrame loaded: {lca_df.shape[0]} rows\")\n",
    "        print(lca_df.head(3))\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load LCA DataFrame: {e}\")\n",
    "        sys.exit(1)\n",
    "        \n",
    "    try:\n",
    "        # Load CD-HIT clusters\n",
    "        clusters = load_cdhit_clusters('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Combined_filtered_500_SENZOR_clsuter.clstr')\n",
    "        logging.info(\"Loaded CD-HIT clusters.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load CD-HIT clusters: {e}\")\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge clusters with LCA\n",
    "        merged_df = merge_clusters_lca(clusters, lca_df)\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "    except KeyError as e:\n",
    "        logging.error(f\"KeyError during merging: {e}\")\n",
    "        print(f\"Error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error during merging: {e}\")\n",
    "        print(f\"Unexpected error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Create clust2kingdom dictionary\n",
    "        clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "        logging.info(\"Created clust2kingdom dictionary.\")\n",
    "\n",
    "        # Save clust2kingdom to CSV\n",
    "        merged_df[['cluster', 'kingdom']].to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv', index=False)\n",
    "        logging.info(\"Saved cluster_kingdoms.csv successfully.\")\n",
    "\n",
    "        # Print sample entries of clust2kingdom\n",
    "        print(\"clust2kingdom Dictionary Sample Entries:\")\n",
    "        for cluster_id, kingdom in list(clust2kingdom.items())[:5]:\n",
    "            print(f\"Cluster {cluster_id}: Kingdom {kingdom}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error creating clust2kingdom: {e}\")\n",
    "        print(f\"Error creating clust2kingdom: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Load and process the counts DataFrame\n",
    "        counts_file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/combined_SENSOR_csvs/1Modified_CSV/merged_contig_stats.csv'\n",
    "        counts_df = load_counts_data(counts_file_path)\n",
    "        logging.info(\"Loaded and processed counts DataFrame.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load counts DataFrame: {e}\")\n",
    "        print(f\"Error loading counts DataFrame: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge counts with merged_df on 'contig_key'\n",
    "        final_df = merged_df.merge(counts_df[['contig_key', 'count']], on='contig_key', how='left')\n",
    "        logging.info(\"Merged counts with merged_df to create final_df.\")\n",
    "\n",
    "        # Convert 'count' to numeric if not already (redundant but safe)\n",
    "        final_df['count'] = pd.to_numeric(final_df['count'], errors='coerce')\n",
    "\n",
    "        # Apply filtering: Remove rows where 'count' is NaN\n",
    "        initial_count = len(final_df)\n",
    "        final_df = final_df.dropna(subset=['count'])\n",
    "        final_count = len(final_df)\n",
    "        removed_rows = initial_count - final_count\n",
    "        if removed_rows > 0:\n",
    "            logging.info(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "            print(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "\n",
    "        # Remove rows where 'count' is zero or negative:\n",
    "        logging.info(\"Filtered out rows with 'count' <= 0.\")\n",
    "\n",
    "        # Verify if final_df is empty after filtering\n",
    "        if final_df.empty:\n",
    "            logging.warning(\"final_df is empty after filtering out rows without 'count'.\")\n",
    "            print(\"final_df is empty after filtering out rows without 'count'.\", file=sys.stderr)\n",
    "        else:\n",
    "            # Display the final merged DataFrame with counts\n",
    "            print(\"Final Merged DataFrame with Counts (after filtering):\")\n",
    "            print(final_df.head())\n",
    "\n",
    "            # Save the final DataFrame to CSV\n",
    "            final_df.to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/final_merged_with_counts.csv', index=False)\n",
    "            logging.info(\"Saved final_merged_with_counts.csv successfully.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error merging counts with merged_df: {e}\")\n",
    "        print(f\"Error merging counts with merged_df: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    # ...the rest of your pipeline (e.g., load CD-HIT clusters, merge, etc.)\n",
    "    # ...\n",
    "\n",
    "\n",
    "# Suppose you no longer have NT files or simply skip them\n",
    "nt_lca_files = []\n",
    "\n",
    "def load_lca(lca_directory, nr_files):\n",
    "    \"\"\"\n",
    "    Simplified function that only loads NR LCA files.\n",
    "    \"\"\"\n",
    "    nr_file_paths = [os.path.join(lca_directory, filename) for filename in nr_files]\n",
    "    # Load BLAST outputs with NR database\n",
    "    nr_df = load_blast_output(nr_file_paths, database='nr')\n",
    "    nr_df = nr_df.dropna(subset=['taxid'])  # remove rows with missing taxid\n",
    "    nr_df['taxid'] = nr_df['taxid'].astype(int)\n",
    "\n",
    "    # (Optionally do lineage lookups, merges, etc.)\n",
    "    return nr_df\n",
    "\n",
    "def load_blast_output(files, database='nr'):\n",
    "    \"\"\"\n",
    "    Loads BLAST output files into a single DataFrame.\n",
    "    \"\"\"\n",
    "    if database == 'nr':\n",
    "        column_names = nr_column_names\n",
    "    else:\n",
    "        raise ValueError(\"Unknown DB\")\n",
    "\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        try:\n",
    "            df = pd.read_csv(\n",
    "                file,\n",
    "                sep='\\t',\n",
    "                header=None,\n",
    "                names=column_names,\n",
    "                na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "                on_bad_lines='warn'\n",
    "            )\n",
    "            if 'tax_id' in df.columns:\n",
    "                df.rename(columns={'tax_id': 'taxid'}, inplace=True)\n",
    "            df['taxid'] = df['taxid'].replace('Unknown', '0')\n",
    "            df['taxid'] = pd.to_numeric(df['taxid'], errors='coerce')\n",
    "            df.dropna(subset=['taxid'], inplace=True)\n",
    "\n",
    "            # You can further process the sample/contig name here:\n",
    "            df['sample'] = [os.path.basename(file).replace('_modified_diamond_blast_nr_annotated.m9','')] * len(df)\n",
    "            df['contig'] = df['query']  # or other logic\n",
    "\n",
    "            # Create contig_key\n",
    "            df['contig_key'] = df['sample'].str.lower() + \"_\" + df['contig'].str.lower()\n",
    "\n",
    "            dfs.append(df)\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error reading {file}: {e}\")\n",
    "            continue\n",
    "\n",
    "    combined_df = pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()\n",
    "    return combined_df\n",
    "\n",
    "# Make sure these helper functions exist in your code\n",
    "def load_cdhit_clusters(file_path):\n",
    "    # ... your implementation ...\n",
    "    return {}\n",
    "\n",
    "def merge_clusters_lca(clusters, lca_df):\n",
    "    # ... your implementation ...\n",
    "    return pd.DataFrame()\n",
    "\n",
    "def load_counts_data(file_path):\n",
    "    # ... your implementation ...\n",
    "    return pd.DataFrame()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    lca_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA\"\n",
    "    \n",
    "    # Dynamically fetch .m9 files that match your naming pattern\n",
    "    nr_lca_files = glob(os.path.join(lca_dir, \"LCA_*_modified_diamond_blast_nr_annotated.m9\"))\n",
    "    print(\"Found NR LCA files:\", nr_lca_files)\n",
    "\n",
    "    # Now load and process only the NR files\n",
    "    try:\n",
    "        lca_df = load_lca(lca_dir, [os.path.basename(f) for f in nr_lca_files])\n",
    "        print(\"NR LCA DataFrame shape:\", lca_df.shape)\n",
    "        print(lca_df.head())\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load LCA DataFrame: {e}\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    # --- Next block (same indentation as above) ---\n",
    "    try:\n",
    "        # Load CD-HIT clusters\n",
    "        clusters = load_cdhit_clusters(\n",
    "            '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Combined_filtered_500_SENZOR_clsuter.clstr'\n",
    "        )\n",
    "        logging.info(\"Loaded CD-HIT clusters.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load CD-HIT clusters: {e}\")\n",
    "        raise e\n",
    "\n",
    "    try:\n",
    "        # Merge clusters with LCA\n",
    "        merged_df = merge_clusters_lca(clusters, lca_df)\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "    except KeyError as e:\n",
    "        logging.error(f\"KeyError during merging: {e}\")\n",
    "        print(f\"Error during merging: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error during merging: {e}\")\n",
    "        print(f\"Unexpected error during merging: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    try:\n",
    "        # Create clust2kingdom dictionary\n",
    "        clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "        logging.info(\"Created clust2kingdom dictionary.\")\n",
    "\n",
    "        # Save clust2kingdom to CSV\n",
    "        merged_df[['cluster', 'kingdom']].to_csv(\n",
    "            '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv',\n",
    "            index=False\n",
    "        )\n",
    "        logging.info(\"Saved cluster_kingdoms.csv successfully.\")\n",
    "\n",
    "        # Print sample entries of clust2kingdom\n",
    "        print(\"clust2kingdom Dictionary Sample Entries:\")\n",
    "        for cluster_id, kingdom in list(clust2kingdom.items())[:5]:\n",
    "            print(f\"Cluster {cluster_id}: Kingdom {kingdom}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error creating clust2kingdom: {e}\")\n",
    "        print(f\"Error creating clust2kingdom: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    try:\n",
    "        # Load and process the counts DataFrame\n",
    "        counts_file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/bowtie_csp_counts_1000.txt'\n",
    "        counts_df = load_counts_data(counts_file_path)\n",
    "        logging.info(\"Loaded and processed counts DataFrame.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load counts DataFrame: {e}\")\n",
    "        print(f\"Error loading counts DataFrame: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "    try:\n",
    "        # Merge counts with merged_df on 'contig_key'\n",
    "        final_df = merged_df.merge(counts_df[['contig_key', 'count']], on='contig_key', how='left')\n",
    "        logging.info(\"Merged counts with merged_df to create final_df.\")\n",
    "\n",
    "        # Apply filtering: Remove rows where 'count' is NaN\n",
    "        initial_count = len(final_df)\n",
    "        final_df = final_df.dropna(subset=['count'])\n",
    "        final_count = len(final_df)\n",
    "        removed_rows = initial_count - final_count\n",
    "        if removed_rows > 0:\n",
    "            logging.info(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "            print(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "\n",
    "        # Optionally remove rows where 'count' is zero or negative\n",
    "        final_df = final_df[final_df['count'] > 0]\n",
    "        logging.info(\"Filtered out rows with 'count' <= 0.\")\n",
    "\n",
    "        # Verify if final_df is empty after filtering\n",
    "        if final_df.empty:\n",
    "            logging.warning(\"final_df is empty after filtering out rows without 'count'.\")\n",
    "            print(\"final_df is empty after filtering out rows without 'count'.\", file=sys.stderr)\n",
    "        else:\n",
    "            # Display the final merged DataFrame with counts\n",
    "            print(\"Final Merged DataFrame with Counts (after filtering):\")\n",
    "            print(final_df.head())\n",
    "\n",
    "            # Save the final DataFrame to CSV\n",
    "            final_df.to_csv(\n",
    "                '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/final_merged_with_counts.csv',\n",
    "                index=False\n",
    "            )\n",
    "            logging.info(\"Saved final_merged_with_counts.csv successfully.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error merging counts with merged_df: {e}\")\n",
    "        print(f\"Error merging counts with merged_df: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "897854a6-8096-470b-961c-4f9cae90213d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clust2kingdom Dictionary Sample Entries:\n",
      "Cluster 0: Kingdom Bacteria\n",
      "Cluster 1: Kingdom Viruses\n",
      "Cluster 2: Kingdom Bacteria\n",
      "Cluster 3: Kingdom Bacteria\n",
      "Cluster 4: Kingdom Bacteria\n",
      "Filtered out 1735 rows without 'count' value.\n",
      "Final Merged DataFrame with Counts (after filtering):\n",
      "   cluster  cluster_size  contig_length                              name  \\\n",
      "0        0             1          64751                Duncaniella sp. C9   \n",
      "1        1             1          37429  Acinetobacter phage Aristophanes   \n",
      "2        2             3          36456              Cutibacterium avidum   \n",
      "3        3             1          35911          uncultured Treponema sp.   \n",
      "4        4             1          34801          uncultured Treponema sp.   \n",
      "\n",
      "      rank   kingdom    taxid  db  align_length  align_percent  ...  \\\n",
      "0  species  Bacteria  2530392  nr          2662            4.1  ...   \n",
      "1  species   Viruses  2759203  nr           784            2.1  ...   \n",
      "2  species  Bacteria    33010  nr          1334            3.7  ...   \n",
      "3  species  Bacteria   162155  nr           923            2.6  ...   \n",
      "4  species  Bacteria   162155  nr           588            1.7  ...   \n",
      "\n",
      "                                           com_names  superkingdom  \\\n",
      "0  WP_136408717.1 hypothetical protein [Duncaniel...           NaN   \n",
      "1  QNO11440.1 putative DNA polymerase [Acinetobac...           NaN   \n",
      "2  MDU5547358.1 tape measure protein [Cutibacteri...           NaN   \n",
      "3  WP_308693184.1 glycoside hydrolase family 3 N-...           NaN   \n",
      "4  WP_288567344.1 asparagine synthase B [uncultur...      Bacteria   \n",
      "\n",
      "         phylum           family lca_taxid                  lca_name  \\\n",
      "0           NaN              NaN       NaN                       NaN   \n",
      "1           NaN              NaN       NaN                       NaN   \n",
      "2           NaN              NaN       NaN                       NaN   \n",
      "3           NaN              NaN       NaN                       NaN   \n",
      "4  Spirochaetes  Spirochaetaceae  162155.0  uncultured Treponema sp.   \n",
      "\n",
      "                              contig                         sample  \\\n",
      "0   NODE_1_length_64751_cov_9.584764  SEZONOWIFGOAT011_S29_2_656747   \n",
      "1  NODE_1_length_37429_cov_18.557909        AIAMAGOAT001_S48_656109   \n",
      "2  NODE_1_length_36456_cov_34.186784        AINWZGOAT001_S41_656115   \n",
      "3   NODE_2_length_35911_cov_4.193866  SEZONOWIFGOAT011_S29_2_656747   \n",
      "4   NODE_3_length_34801_cov_4.368276  SEZONOWIFGOAT011_S29_2_656747   \n",
      "\n",
      "                                          contig_key    count  \n",
      "0  sezonowifgoat011_s29_2_656747~node_1_length_64...  11643.0  \n",
      "1  aiamagoat001_s48_656109~node_1_length_37429_co...  14491.0  \n",
      "2  ainwzgoat001_s41_656115~node_1_length_36456_co...  27866.0  \n",
      "3  sezonowifgoat011_s29_2_656747~node_2_length_35...   2779.0  \n",
      "4  sezonowifgoat011_s29_2_656747~node_3_length_34...   2796.0  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "from statistics import mode\n",
    "from Bio import Entrez\n",
    "from ete3 import NCBITaxa\n",
    "import logging\n",
    "import sys\n",
    "import re\n",
    "from logging.handlers import RotatingFileHandler\n",
    "\n",
    "handler = RotatingFileHandler('util.log', maxBytes=5*1024*1024, backupCount=2)\n",
    "logging.basicConfig(level=logging.INFO, handlers=[handler], \n",
    "                    format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "# Set Entrez email\n",
    "Entrez.email = \"ifeanyi.omah@ed.ac.uk\"\n",
    "\n",
    "# Initialize NCBI Taxonomy\n",
    "ncbi = NCBITaxa()\n",
    "# Optionally: ncbi.update_taxonomy_database()\n",
    "\n",
    "valid_lineages = set()\n",
    "\n",
    "# Updated column names to match your .m9 file\n",
    "nr_column_names = [\n",
    "    'query', 'subject', 'identity', 'alignment_length', 'mismatches', 'gap_opens',\n",
    "    'q_start', 'q_end', 's_start', 's_end', 'evalue', 'bit_score', 'tax_id',\n",
    "    'sci_name', 'com_names', 'superkingdom', 'phylum', 'family',\n",
    "    'lca_taxid', 'lca_name'\n",
    "]\n",
    "\n",
    "# Define NR directory\n",
    "nr_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/\"\n",
    "\n",
    "def load_lca(nr_directory):\n",
    "    \"\"\"\n",
    "    Load and process LCA results from NR files in the specified directory.\n",
    "\n",
    "    Parameters:\n",
    "        nr_directory (str): Path to the directory containing NR LCA files.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed LCA DataFrame.\n",
    "    \"\"\"\n",
    "    # CHANGED: Search specifically for 'LCA_*.m9' files\n",
    "    nr_file_paths = glob(f\"{nr_directory}/LCA_*.m9\")\n",
    "    logging.info(f\"Found {len(nr_file_paths)} NR files matching 'LCA_*.m9' in {nr_directory}.\")\n",
    "\n",
    "    if not nr_file_paths:\n",
    "        logging.error(f\"No files matching LCA_*.m9 found in directory: {nr_directory}\")\n",
    "        raise FileNotFoundError(f\"No files matching LCA_*.m9 found in directory: {nr_directory}\")\n",
    "\n",
    "    # Process each BLAST output file\n",
    "    nr_df = load_blast_output(nr_file_paths, database='nr')\n",
    "\n",
    "    # Remove rows with missing taxid\n",
    "    initial_nr = len(nr_df)\n",
    "    nr_df = nr_df.dropna(subset=['taxid'])\n",
    "    final_nr = len(nr_df)\n",
    "    dropped_nr = initial_nr - final_nr\n",
    "    if dropped_nr > 0:\n",
    "        logging.info(f\"Dropped {dropped_nr} rows with missing taxid from nr_df.\")\n",
    "\n",
    "    # Convert taxid to integer\n",
    "    nr_df['taxid'] = nr_df['taxid'].astype(int)\n",
    "\n",
    "    # Get all lineages for valid taxids\n",
    "    taxids = set(nr_df.taxid)\n",
    "    taxids = {t for taxid in taxids for t in ncbi.get_lineage(taxid) if t != 0}\n",
    "\n",
    "    # Build dictionaries for name/rank\n",
    "    taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    taxid2name[-1] = 'Ambiguous'\n",
    "    taxid2name[0] = 'Unknown'\n",
    "    taxid2rank = ncbi.get_rank(taxids)\n",
    "\n",
    "    # Verify valid lineages\n",
    "    taxids = set(nr_df['taxid'])\n",
    "    for taxid in taxids:\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(taxid)\n",
    "            for t in lineage:\n",
    "                if t != 0:\n",
    "                    valid_lineages.add(t)\n",
    "        except ValueError:\n",
    "            logging.warning(f\"Skipping {taxid}: taxid not found in local DB.\")\n",
    "\n",
    "    # Now limit to valid_lineages only\n",
    "    taxids = valid_lineages\n",
    "    taxid2name = ncbi.get_taxid_translator(taxids)\n",
    "    \n",
    "    def taxid2kingdom(taxid):\n",
    "        \"\"\"Retrieve the kingdom for a given taxid.\"\"\"\n",
    "        if taxid == 0:\n",
    "            return 'Unknown'\n",
    "        lineage_dict = ncbi.get_rank(ncbi.get_lineage(taxid))\n",
    "        for k, v in lineage_dict.items():\n",
    "            if v == 'superkingdom':\n",
    "                kingdom = taxid2name.get(k, 'Unknown')\n",
    "                if kingdom != 'Unknown' and taxid != k:\n",
    "                    logging.warning(f\"taxid {taxid} was translated into {kingdom}\")\n",
    "                return kingdom\n",
    "        return 'Unknown'\n",
    "\n",
    "    # Map taxid -> name, rank, kingdom\n",
    "    nr_df['name'] = nr_df['taxid'].map(taxid2name)\n",
    "    nr_df['rank'] = nr_df['taxid'].map(taxid2rank)\n",
    "    nr_df['kingdom'] = nr_df['taxid'].apply(taxid2kingdom)\n",
    "    nr_df['db'] = 'nr'  # Since this is the NR dataset\n",
    "\n",
    "    return nr_df\n",
    "\n",
    "def load_blast_output(files, database='nr'):\n",
    "    \"\"\"\n",
    "    Load BLAST output files into a single DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        files (list): List of file paths to BLAST output files.\n",
    "        database (str): Type of database ('nr').\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Concatenated BLAST results.\n",
    "    \"\"\"\n",
    "    def get_sample_from_path(path):\n",
    "        filename = path.rsplit('/', 1)[-1]  # Extract just the filename\n",
    "        prefix = \"LCA_\"\n",
    "        suffix = \"_modified_diamond_blast_nr_annotated.m9\"\n",
    "        # Remove the known prefix\n",
    "        if filename.startswith(prefix):\n",
    "            filename = filename[len(prefix):]\n",
    "        # Remove the known suffix\n",
    "        if filename.endswith(suffix):\n",
    "            filename = filename[:-len(suffix)]\n",
    "        return filename\n",
    "\n",
    "    if database == 'nr':\n",
    "        column_names = nr_column_names\n",
    "    else:\n",
    "        logging.error(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "        raise ValueError(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        try:\n",
    "            # Quick check of columns in the first line\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                columns_in_file = len(first_line.strip().split('\\t'))\n",
    "                expected_columns = len(column_names)\n",
    "                if columns_in_file != expected_columns:\n",
    "                    logging.warning(f\"{file} has {columns_in_file} columns; expected {expected_columns}.\")\n",
    "\n",
    "            # Determine if the first line contains headers\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                if 'sci_name' in first_line.lower() or 'lca_taxid' in first_line.lower():\n",
    "                    skiprows = 1\n",
    "                else:\n",
    "                    skiprows = 0\n",
    "\n",
    "            df = pd.read_csv(\n",
    "                file,\n",
    "                sep='\\t',\n",
    "                header=None,\n",
    "                names=column_names,\n",
    "                na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "                skiprows=skiprows,\n",
    "                on_bad_lines='warn'  # or 'error' if you want it strict\n",
    "            )\n",
    "\n",
    "            # Rename 'tax_id' -> 'taxid' for consistency\n",
    "            if 'tax_id' in df.columns:\n",
    "                df.rename(columns={'tax_id': 'taxid'}, inplace=True)\n",
    "            else:\n",
    "                logging.error(f\"'tax_id' column is missing in {file}.\")\n",
    "                raise KeyError(f\"'tax_id' column is missing in {file}.\")\n",
    "\n",
    "            # Replace 'Unknown' with '0' in taxid\n",
    "            df['taxid'] = df['taxid'].replace('Unknown', '0')\n",
    "\n",
    "            # Convert taxid to numeric\n",
    "            df['taxid'] = pd.to_numeric(df['taxid'], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in 'taxid'\n",
    "            before_drop = len(df)\n",
    "            df = df.dropna(subset=['taxid'])\n",
    "            after_drop = len(df)\n",
    "            dropped = before_drop - after_drop\n",
    "            if dropped > 0:\n",
    "                logging.info(f\"Dropped {dropped} rows with invalid taxid in {file}.\")\n",
    "\n",
    "            # Derive sample name\n",
    "            df['sample'] = get_sample_from_path(file)\n",
    "\n",
    "            # Extract contig\n",
    "            df['contig'] = df.apply(lambda row: extract_contig(row['query'], row['sample']), axis=1)\n",
    "\n",
    "            # alignment_length -> align_length\n",
    "            df.rename(columns={'alignment_length': 'align_length'}, inplace=True)\n",
    "\n",
    "            # Build contig_key\n",
    "            df['contig_key'] = (df['sample'] + '~' + df['contig']).str.lower()\n",
    "\n",
    "            # Convert numeric columns\n",
    "            numeric_columns = ['identity', 'align_length', 'bit_score', 'evalue']\n",
    "            for col in numeric_columns:\n",
    "                if col in df.columns:\n",
    "                    df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in numeric fields (optional)\n",
    "            before_drop_numeric = len(df)\n",
    "            df = df.dropna(subset=numeric_columns)\n",
    "            dropped_numeric = before_drop_numeric - len(df)\n",
    "            if dropped_numeric > 0:\n",
    "                logging.info(f\"Dropped {dropped_numeric} rows with non-numeric in {file}.\")\n",
    "\n",
    "            dfs.append(df)\n",
    "\n",
    "        except pd.errors.ParserError as e:\n",
    "            logging.error(f\"ParserError while reading {file}: {e}\")\n",
    "            print(f\"ParserError while reading {file}: {e}\", file=sys.stderr)\n",
    "        except KeyError as e:\n",
    "            logging.error(f\"KeyError while processing {file}: {e}\")\n",
    "            print(f\"KeyError while processing {file}: {e}\", file=sys.stderr)\n",
    "            raise e\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Unexpected error while reading {file}: {e}\")\n",
    "            print(f\"Unexpected error while reading {file}: {e}\", file=sys.stderr)\n",
    "            raise e\n",
    "\n",
    "    if dfs:\n",
    "        combined_df = pd.concat(dfs, ignore_index=True)\n",
    "    else:\n",
    "        combined_df = pd.DataFrame()\n",
    "        logging.warning(\"No DataFrames were created; returning empty DataFrame.\")\n",
    "\n",
    "    return combined_df\n",
    "\n",
    "def extract_contig(query, sample):\n",
    "    \"\"\"\n",
    "    Extract contig information by removing the sample prefix from the query.\n",
    "\n",
    "    Parameters:\n",
    "        query (str): The query string from BLAST output.\n",
    "        sample (str): The sample name.\n",
    "\n",
    "    Returns:\n",
    "        str: Extracted contig information or 'unknown_contig'.\n",
    "    \"\"\"\n",
    "    prefix = sample + '_'\n",
    "    if query.startswith(prefix):\n",
    "        return query[len(prefix):]\n",
    "    else:\n",
    "        logging.warning(f\"Query '{query}' does not start with sample prefix '{prefix}'. Using 'unknown_contig'.\")\n",
    "        return 'unknown_contig'\n",
    "\n",
    "# -------------------------------------------\n",
    "# The rest of your functions remain unchanged\n",
    "# -------------------------------------------\n",
    "\n",
    "\n",
    "def load_cdhit_clusters(filename):\n",
    "    \"\"\"\n",
    "    Load CD-HIT cluster information from a cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        filename (str): Path to the CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        defaultdict: Cluster ID mapped to list of members.\n",
    "    \"\"\"\n",
    "    clusters = defaultdict(list)\n",
    "    cluster_id = None\n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>Cluster'):\n",
    "                cluster_id = line.strip().split()[-1]\n",
    "            else:\n",
    "                if cluster_id is not None:\n",
    "                    try:\n",
    "                        member = parse_cdhit_row(line)\n",
    "                        clusters[cluster_id].append(member)\n",
    "                    except ValueError as e:\n",
    "                        logging.error(f\"Error parsing line: {line.strip()} - {e}\")\n",
    "    return clusters\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    \"\"\"\n",
    "    Parse a single line from a CD-HIT cluster file.\n",
    "\n",
    "    Parameters:\n",
    "        row (str): Line from CD-HIT cluster file.\n",
    "\n",
    "    Returns:\n",
    "        dict: Parsed member information.\n",
    "    \"\"\"\n",
    "    if '*' in row:\n",
    "        # Reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = True\n",
    "    else:\n",
    "        # Non-reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = False\n",
    "\n",
    "    # Clean and extract fields\n",
    "    # Use regex to extract numeric length\n",
    "    match = re.match(r'(\\d+)', length_str)\n",
    "    if match:\n",
    "        length = int(match.group(1))\n",
    "    else:\n",
    "        logging.warning(f\"Could not extract numeric length from '{length_str}' in row: {row}\")\n",
    "        length = 0  # Assign a default value or handle as needed\n",
    "\n",
    "    name = name.strip('>').strip('.')\n",
    "\n",
    "    # Extract sample and contig\n",
    "    try:\n",
    "        sample, contig = name.split('~')\n",
    "    except ValueError:\n",
    "        # Handle unexpected naming format\n",
    "        sample, contig = 'unknown_sample', name  # Assign a default sample name\n",
    "\n",
    "    return {\n",
    "        'contig': contig,\n",
    "        'sample': sample,\n",
    "        'length': length,\n",
    "        'is_ref': is_ref\n",
    "    }\n",
    "\n",
    "def get_cluster_rep(cluster):\n",
    "    \"\"\"\n",
    "    Get the representative member of a cluster (the reference member).\n",
    "\n",
    "    Parameters:\n",
    "        cluster (list): List of cluster members.\n",
    "\n",
    "    Returns:\n",
    "        dict: Representative member.\n",
    "    \"\"\"\n",
    "    for member in cluster:\n",
    "        if member['is_ref']:\n",
    "            return member\n",
    "    # If no reference member, return the first member\n",
    "    return cluster[0] if cluster else None\n",
    "\n",
    "def merge_clusters_lca(clusters, lca_df):\n",
    "    \"\"\"\n",
    "    Merge cluster information with LCA results.\n",
    "\n",
    "    Parameters:\n",
    "        clusters (defaultdict): Cluster information.\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Merged cluster and LCA information.\n",
    "    \"\"\"\n",
    "    # Prepare cluster data\n",
    "    cluster_ids = [int(id.strip()) for id in clusters.keys()]\n",
    "    cluster_reps = [get_cluster_rep(clusters[str(id)]) for id in cluster_ids]\n",
    "    cluster_lengths = [rep['length'] for rep in cluster_reps]\n",
    "    cluster_sizes = [len(clusters[str(id)]) for id in cluster_ids]\n",
    "    # Correct contig_key formatting using '~'\n",
    "    cluster_contig_keys = [f\"{rep['sample']}~{rep['contig']}\".lower() for rep in cluster_reps]\n",
    "\n",
    "    # Create DataFrame for clusters\n",
    "    df = pd.DataFrame({\n",
    "        'contig_key': cluster_contig_keys,\n",
    "        'cluster': cluster_ids,\n",
    "        'cluster_size': cluster_sizes,\n",
    "        'contig_length': cluster_lengths\n",
    "    })\n",
    "\n",
    "    # Explicitly convert to string, then strip and lowercase\n",
    "    df['contig_key'] = df['contig_key'].astype(str).str.strip().str.lower()\n",
    "\n",
    "    # Debug: Log sample 'contig_key's\n",
    "    logging.info(\"Sample 'contig_key's in clusters DataFrame:\")\n",
    "    logging.info(df['contig_key'].head())\n",
    "\n",
    "    # Merge with LCA DataFrame\n",
    "    merged_df = df.merge(lca_df, on='contig_key', how='inner')\n",
    "\n",
    "    # Debug: Log the number of merged rows\n",
    "    logging.info(f\"Merged DataFrame has {len(merged_df)} rows.\")\n",
    "\n",
    "    if merged_df.empty:\n",
    "        logging.warning(\"merged_df is empty after merging. Check 'contig_key' consistency.\")\n",
    "        print(\"merged_df is empty after merging. Please check 'contig_key' consistency between clusters and LCA DataFrames.\", file=sys.stderr)\n",
    "    else:\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "\n",
    "    # Calculate align_percent\n",
    "    merged_df['align_percent'] = np.round(100 * merged_df['align_length'] / merged_df['contig_length'], 1)\n",
    "\n",
    "    # Use 'identity' as 'percent_identity'\n",
    "    merged_df['percent_identity'] = np.round(merged_df['identity'], 1)\n",
    "\n",
    "    # Define required columns (including additional columns)\n",
    "    required_columns = [\n",
    "        'cluster', 'cluster_size', 'contig_length', 'name', 'rank', \n",
    "        'kingdom', 'taxid', 'db', 'align_length', 'align_percent', \n",
    "        'percent_identity', 'bit_score', 'evalue', 'subject', 'sci_name',\n",
    "        'com_names', 'superkingdom', 'phylum', 'family', 'lca_taxid', 'lca_name',\n",
    "        'contig', 'sample', 'contig_key'\n",
    "    ]\n",
    "\n",
    "    # Check for missing columns\n",
    "    missing_columns = [col for col in required_columns if col not in merged_df.columns]\n",
    "    if missing_columns:\n",
    "        logging.error(f\"Missing columns in merged DataFrame: {missing_columns}\")\n",
    "        logging.info(f\"Available columns: {merged_df.columns.tolist()}\")\n",
    "        raise KeyError(f\"{missing_columns} not in index\")\n",
    "\n",
    "    # Subset the DataFrame to include only required columns\n",
    "    merged_df = merged_df[required_columns]\n",
    "\n",
    "    # Sort by cluster\n",
    "    merged_df = merged_df.sort_values('cluster').reset_index(drop=True)\n",
    "\n",
    "    return merged_df\n",
    "\n",
    "def load_counts_data(counts_file):\n",
    "    \"\"\"\n",
    "    Load and process the counts DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        counts_file (str): Path to the counts file.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed counts DataFrame with 'contig_key'.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Read counts file with header, assuming first row is header\n",
    "        counts = pd.read_csv(\n",
    "            counts_file, \n",
    "            sep=',', \n",
    "            header=0,  # Assumes first row is header\n",
    "            usecols=[0, 1, 2],  # Read 'contig', 'count', 'length'\n",
    "            names=['contig', 'count', 'length'],  # Assign column names\n",
    "            dtype={'contig': str, 'count': str, 'length': str},  # Read as strings initially\n",
    "            na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "            low_memory=False\n",
    "        )\n",
    "        \n",
    "        # Log a sample of the counts DataFrame\n",
    "        logging.info(\"Raw Counts DataFrame sample:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Convert 'count' to numeric, coercing errors to NaN\n",
    "        counts['count'] = pd.to_numeric(counts['count'], errors='coerce')\n",
    "        \n",
    "        # Log the conversion result\n",
    "        logging.info(\"Counts DataFrame after converting 'count' to numeric:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        # Extract sample and contig from 'contig' column assuming 'sample~contig'\n",
    "        counts[['sample', 'contig']] = counts['contig'].str.split('~', expand=True)\n",
    "        \n",
    "        # Handle cases where 'contig' does not contain '~'\n",
    "        counts['contig'] = counts['contig'].fillna(counts['contig'])\n",
    "        counts['sample'] = counts['sample'].fillna('unknown_sample')\n",
    "        \n",
    "        # Create 'contig_key' as 'sample~contig', lowercase\n",
    "        counts['contig_key'] = (counts['sample'].astype(str) + '~' + counts['contig'].astype(str)).str.strip().str.lower()\n",
    "        \n",
    "        # Debug: Log the processed counts DataFrame\n",
    "        logging.info(\"Processed Counts DataFrame:\")\n",
    "        logging.info(counts.head())\n",
    "        \n",
    "        return counts\n",
    "    except pd.errors.ParserError as e:\n",
    "        logging.error(f\"ParserError while reading counts file {counts_file}: {e}\")\n",
    "        print(f\"ParserError while reading counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error while processing counts file {counts_file}: {e}\")\n",
    "        print(f\"Unexpected error while processing counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "def load_taxid2name(lca_df):\n",
    "    \"\"\"\n",
    "    Create a taxid to name mapping from the LCA DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        lca_df (pd.DataFrame): LCA DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        dict: Mapping from taxid to scientific name.\n",
    "    \"\"\"\n",
    "    taxid2name = dict(zip(lca_df['taxid'], lca_df['name']))\n",
    "    taxid2name[-1] = 'Ambiguous'  # Handle ambiguous taxids\n",
    "    taxid2name[0] = 'Unknown'     # Handle 'Unknown' taxids\n",
    "    return taxid2name\n",
    "\n",
    "# Example usage within util.py (optional)\n",
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        # Load LCA results from NR database\n",
    "        lca_df = load_lca(nr_dir)\n",
    "        if lca_df.empty:\n",
    "            logging.error(\"LCA DataFrame is empty. Exiting.\")\n",
    "            raise ValueError(\"LCA DataFrame is empty.\")\n",
    "        logging.info(\"Loaded LCA DataFrame.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load LCA DataFrame: {e}\")\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Load CD-HIT clusters\n",
    "        clusters = load_cdhit_clusters('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Combined_filtered_500_SENZOR_clsuter.clstr')\n",
    "        logging.info(\"Loaded CD-HIT clusters.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load CD-HIT clusters: {e}\")\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge clusters with LCA\n",
    "        merged_df = merge_clusters_lca(clusters, lca_df)\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "    except KeyError as e:\n",
    "        logging.error(f\"KeyError during merging: {e}\")\n",
    "        print(f\"Error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error during merging: {e}\")\n",
    "        print(f\"Unexpected error during merging: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Create clust2kingdom dictionary\n",
    "        clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "        logging.info(\"Created clust2kingdom dictionary.\")\n",
    "\n",
    "        # Save clust2kingdom to CSV\n",
    "        merged_df[['cluster', 'kingdom']].to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv', index=False)\n",
    "        logging.info(\"Saved cluster_kingdoms.csv successfully.\")\n",
    "\n",
    "        # Print sample entries of clust2kingdom\n",
    "        print(\"clust2kingdom Dictionary Sample Entries:\")\n",
    "        for cluster_id, kingdom in list(clust2kingdom.items())[:5]:\n",
    "            print(f\"Cluster {cluster_id}: Kingdom {kingdom}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error creating clust2kingdom: {e}\")\n",
    "        print(f\"Error creating clust2kingdom: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Load and process the counts DataFrame\n",
    "        counts_file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/combined_SENSOR_csvs/1Modified_CSV/merged_contig_stats.csv'\n",
    "        counts_df = load_counts_data(counts_file_path)\n",
    "        logging.info(\"Loaded and processed counts DataFrame.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load counts DataFrame: {e}\")\n",
    "        print(f\"Error loading counts DataFrame: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate\n",
    "\n",
    "    try:\n",
    "        # Merge counts with merged_df on 'contig_key'\n",
    "        final_df = merged_df.merge(counts_df[['contig_key', 'count']], on='contig_key', how='left')\n",
    "        logging.info(\"Merged counts with merged_df to create final_df.\")\n",
    "\n",
    "        # Convert 'count' to numeric if not already (redundant but safe)\n",
    "        final_df['count'] = pd.to_numeric(final_df['count'], errors='coerce')\n",
    "\n",
    "        # Apply filtering: Remove rows where 'count' is NaN\n",
    "        initial_count = len(final_df)\n",
    "        final_df = final_df.dropna(subset=['count'])\n",
    "        final_count = len(final_df)\n",
    "        removed_rows = initial_count - final_count\n",
    "        if removed_rows > 0:\n",
    "            logging.info(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "            print(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "\n",
    "        # Remove rows where 'count' is zero or negative:\n",
    "        logging.info(\"Filtered out rows with 'count' <= 0.\")\n",
    "\n",
    "        # Verify if final_df is empty after filtering\n",
    "        if final_df.empty:\n",
    "            logging.warning(\"final_df is empty after filtering out rows without 'count'.\")\n",
    "            print(\"final_df is empty after filtering out rows without 'count'.\", file=sys.stderr)\n",
    "        else:\n",
    "            # Display the final merged DataFrame with counts\n",
    "            print(\"Final Merged DataFrame with Counts (after filtering):\")\n",
    "            print(final_df.head())\n",
    "\n",
    "            # Save the final DataFrame to CSV\n",
    "            final_df.to_csv('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/1final_merged_with_counts.csv', index=False)\n",
    "            logging.info(\"Saved final_merged_with_counts.csv successfully.\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error merging counts with merged_df: {e}\")\n",
    "        print(f\"Error merging counts with merged_df: {e}\", file=sys.stderr)\n",
    "        raise e  # Let the exception propagate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e92b255-0d5d-4151-907d-a69df1b7ee66",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def filter_fasta(input_file, output_file, min_length=300):\n",
    "    \"\"\"Filters a FASTA file, keeping only sequences longer than the specified minimum length.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        output_file (str): Path to the output FASTA file.\n",
    "        min_length (int, optional): Minimum sequence length. Defaults to 500.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(output_file, 'w') as outfile:\n",
    "        for record in SeqIO.parse(input_file, 'fasta'):\n",
    "            if len(record.seq) >= min_length:\n",
    "                SeqIO.write(record, outfile, 'fasta')\n",
    "\n",
    "# Example usage:\n",
    "input_file = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/All_contigs_combined.fasta\"\n",
    "output_file = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/All_contigs_combined.fa\"\n",
    "filter_fasta(input_file, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6816f80b-77b2-4bd6-9db1-be68ce3aa7be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully loaded data from /Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/contig_stats_filtered_lca.tsv. Total records: 721594\n",
      "Filtered records containing 'Viruses'. Total virus records: 38974\n",
      "After applying 'contig_length' >= 500, remaining records: 3049\n",
      "Filtered data saved to /Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/500_contig_stats_filtered_viruses.csv successfully.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Specify the path to your input and output files\n",
    "input_file = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/blast_with_clusters.tsv'  # Replace with your input file path\n",
    "output_file = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/500_contig_stats_filtered_viruses.csv'  # The file where viruses will be saved\n",
    "\n",
    "try:\n",
    "    # Load the data into a pandas DataFrame with the correct separator\n",
    "    df = pd.read_csv(input_file, sep='\\t')\n",
    "    print(f\"Successfully loaded data from {input_file}. Total records: {len(df)}\")\n",
    "    \n",
    "    # Check if 'taxon_group' column exists\n",
    "    if 'taxon_group' not in df.columns:\n",
    "        raise KeyError(\"'taxon_group' column not found in the input file.\")\n",
    "    \n",
    "    # Check if 'contig_length' column exists\n",
    "    if 'contig_length' not in df.columns:\n",
    "        raise KeyError(\"'contig_length' column not found in the input file.\")\n",
    "    \n",
    "    # Ensure 'contig_length' is numeric\n",
    "    if not pd.api.types.is_numeric_dtype(df['contig_length']):\n",
    "        df['contig_length'] = pd.to_numeric(df['contig_length'], errors='coerce')\n",
    "        num_non_numeric = df['contig_length'].isna().sum()\n",
    "        if num_non_numeric > 0:\n",
    "            print(f\"Warning: {num_non_numeric} records have non-numeric 'contig_length' and were set to NaN.\")\n",
    "    \n",
    "    # Filter rows where 'taxon_group' contains 'Viruses' (case-insensitive)\n",
    "    virus_df = df[df['taxon_group'].str.contains('Viruses', case=False, na=False)]\n",
    "    print(f\"Filtered records containing 'Viruses'. Total virus records: {len(virus_df)}\")\n",
    "    \n",
    "    # Further filter rows where 'contig_length' is >= 500\n",
    "    filtered_virus_df = virus_df[virus_df['contig_length'] >= 500]\n",
    "    print(f\"After applying 'contig_length' >= 500, remaining records: {len(filtered_virus_df)}\")\n",
    "    \n",
    "    # Save the filtered DataFrame to a new CSV file\n",
    "    filtered_virus_df.to_csv(output_file, index=False)\n",
    "    print(f\"Filtered data saved to {output_file} successfully.\")\n",
    "    \n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: The file {input_file} was not found. Please check the file path.\")\n",
    "except pd.errors.ParserError:\n",
    "    print(f\"Error: Could not parse the file {input_file}. Please ensure it's a valid TSV file.\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: {e}\")\n",
    "except ValueError as e:\n",
    "    print(f\"Value Error: {e}\")\n",
    "except Exception as e:\n",
    "    print(f\"An unexpected error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "02b5a273-6ff4-42fa-9088-60abe2a40f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def filter_fasta(input_file, output_file, min_length=500):\n",
    "    \"\"\"Filters a FASTA file, keeping only sequences longer than the specified minimum length.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        output_file (str): Path to the output FASTA file.\n",
    "        min_length (int, optional): Minimum sequence length to keep. Defaults to 500.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(output_file, 'w') as output_handle:\n",
    "        for record in SeqIO.parse(input_file, 'fasta'):\n",
    "            if len(record.seq) >= min_length:\n",
    "                SeqIO.write(record, output_handle, 'fasta')\n",
    "\n",
    "# Example usage:\n",
    "input_file = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/Non_rodents_500.fasta'\n",
    "output_file = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/Non_Rodents_500_cluster_filter.fasta'\n",
    "filter_fasta(input_file, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bc62a3db-ad7e-43bf-bafa-5dd26bba74a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def merge_txt_files(file1, file2, output_file):\n",
    "    \"\"\"Merges two text files with identical headers.\n",
    "\n",
    "    Args:\n",
    "        file1 (str): Path to the first text file.\n",
    "        file2 (str): Path to the second text file.\n",
    "        output_file (str): Path to the output merged file.\n",
    "    \"\"\"\n",
    "\n",
    "    df1 = pd.read_csv(file1, sep='\\t')\n",
    "    df2 = pd.read_csv(file2, sep='\\t')\n",
    "\n",
    "    # Merge the DataFrames, assuming 'contig_name' is the unique identifier\n",
    "    merged_df = pd.concat([df1, df2], ignore_index=True)\n",
    "\n",
    "    # Write the merged DataFrame to a new text file\n",
    "    merged_df.to_csv(output_file, sep='\\t', index=False)\n",
    "\n",
    "# Example usage:\n",
    "file1 = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/Non_rodents_bowtie_csp_1000.txt\"\n",
    "file2 = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/merged_SENSOR_bowtie_csp_1000.txt\"\n",
    "output_file = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Cluster_contigs/1merged_SENSOR_bowtie_csp_1000.txt\"\n",
    "merge_txt_files(file1, file2, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "528c7501-9689-4b17-8e60-044ce8ebc856",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2663848 was translated into 2664091\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991675 was translated into 3383036\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2968474 was translated into 2968473\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 56232 was translated into 3370561\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3035934 was translated into 3031728\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 3026421 was translated into 3026419\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2705430 was translated into 2675323\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 96101 was translated into 46163\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771547 was translated into 351659\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762244 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2762239 was translated into 2760084\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 553056 was translated into 3097545\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 159320 was translated into 3371122\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2771544 was translated into 351673\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2916757 was translated into 2779506\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 103587 was translated into 3371154\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique LCA contig_keys: 1616406\n",
      "Number of unique cluster contig_keys: 320917\n",
      "Number of LCA contig_keys not found in clusters: 1517383\n",
      "Sample missing contig_keys (cluster): ['aiamagoat002_s49_656110~node_5986_length_472_cov_1.179747', 'aiamagoat003_s50_657720~node_66920_length_250_cov_0.855491', 'ainwzchicken001_s56_655747~node_5002_length_340_cov_0.855513', 'sezoowogdog006_s16_1_657760~node_12378_length_265_cov_0.787234', 'ainwzdog003_s39_657736~node_962_length_439_cov_2.287293', 'ainwzgoat001_s41_656115~node_3119_length_306_cov_1.039301', 'sezonowgoat010_s65_2_656741~node_38131_length_111_cov_34.264706', 'sezonowifgoat011_s29_2_656747~node_25293_length_429_cov_1.511364', 'sezonowowisr086_liv_s67_666422~node_4828_length_377_cov_0.666667', 'ainwzchicken001_s56_655747~node_21540_length_247_cov_0.864706'] ...\n",
      "clust2kingdom Dictionary Sample Entries:\n",
      "Cluster 0: Kingdom Bacteria\n",
      "Cluster 1: Kingdom Viruses\n",
      "Cluster 2: Kingdom Bacteria\n",
      "Cluster 3: Kingdom Bacteria\n",
      "Cluster 4: Kingdom Bacteria\n",
      "Number of LCA contig_keys not found in counts: 894812\n",
      "Sample missing contig_keys (counts): ['ainwzgoat005_s45_1_656184~node_31671_length_255_cov_0.831461', 'aiamar019_liv_s39_663050~node_35970_length_274_cov_0.751269', 'sezonowoggoat018_s66_1_657730~node_11448_length_271_cov_0.855670', 'aiamagoat003_s50_657720~node_66920_length_250_cov_0.855491', 'ainwzchicken001_s56_655747~node_5002_length_340_cov_0.855513', 'sezoowogdog006_s16_1_657760~node_12378_length_265_cov_0.787234', 'sezonowowisr086_liv_s67_666422~node_4828_length_377_cov_0.666667', 'aiamagoat001_s48_656109~node_25735_length_259_cov_0.813187', 'aiamagoat003_s50_657720~node_66150_length_251_cov_0.850575', 'sezonowifgoat012_s30_657726~node_23564_length_238_cov_0.919255'] ...\n",
      "Filtered out 1735 rows without 'count' value.\n",
      "Final Merged DataFrame with Counts (after filtering):\n",
      "   cluster  cluster_size  contig_length                              name  \\\n",
      "0        0             1          64751                Duncaniella sp. C9   \n",
      "1        1             1          37429  Acinetobacter phage Aristophanes   \n",
      "2        2             3          36456              Cutibacterium avidum   \n",
      "3        3             1          35911          uncultured Treponema sp.   \n",
      "4        4             1          34801          uncultured Treponema sp.   \n",
      "\n",
      "      rank   kingdom    taxid  db  align_length  align_percent  ...  \\\n",
      "0  species  Bacteria  2530392  nr          2662            4.1  ...   \n",
      "1  species   Viruses  2759203  nr           784            2.1  ...   \n",
      "2  species  Bacteria    33010  nr          1334            3.7  ...   \n",
      "3  species  Bacteria   162155  nr           923            2.6  ...   \n",
      "4  species  Bacteria   162155  nr           588            1.7  ...   \n",
      "\n",
      "                                           com_names  superkingdom  \\\n",
      "0  WP_136408717.1 hypothetical protein [Duncaniel...           NaN   \n",
      "1  QNO11440.1 putative DNA polymerase [Acinetobac...           NaN   \n",
      "2  MDU5547358.1 tape measure protein [Cutibacteri...           NaN   \n",
      "3  WP_308693184.1 glycoside hydrolase family 3 N-...           NaN   \n",
      "4  WP_288567344.1 asparagine synthase B [uncultur...      Bacteria   \n",
      "\n",
      "         phylum           family lca_taxid                  lca_name  \\\n",
      "0           NaN              NaN       NaN                       NaN   \n",
      "1           NaN              NaN       NaN                       NaN   \n",
      "2           NaN              NaN       NaN                       NaN   \n",
      "3           NaN              NaN       NaN                       NaN   \n",
      "4  Spirochaetes  Spirochaetaceae  162155.0  uncultured Treponema sp.   \n",
      "\n",
      "                              contig                         sample  \\\n",
      "0   NODE_1_length_64751_cov_9.584764  SEZONOWIFGOAT011_S29_2_656747   \n",
      "1  NODE_1_length_37429_cov_18.557909        AIAMAGOAT001_S48_656109   \n",
      "2  NODE_1_length_36456_cov_34.186784        AINWZGOAT001_S41_656115   \n",
      "3   NODE_2_length_35911_cov_4.193866  SEZONOWIFGOAT011_S29_2_656747   \n",
      "4   NODE_3_length_34801_cov_4.368276  SEZONOWIFGOAT011_S29_2_656747   \n",
      "\n",
      "                                          contig_key    count  \n",
      "0  sezonowifgoat011_s29_2_656747~node_1_length_64...  11643.0  \n",
      "1  aiamagoat001_s48_656109~node_1_length_37429_co...  14491.0  \n",
      "2  ainwzgoat001_s41_656115~node_1_length_36456_co...  27866.0  \n",
      "3  sezonowifgoat011_s29_2_656747~node_2_length_35...   2779.0  \n",
      "4  sezonowifgoat011_s29_2_656747~node_3_length_34...   2796.0  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "from statistics import mode\n",
    "from Bio import Entrez\n",
    "from ete3 import NCBITaxa\n",
    "import logging\n",
    "import sys\n",
    "import re\n",
    "from logging.handlers import RotatingFileHandler\n",
    "\n",
    "handler = RotatingFileHandler('util.log', maxBytes=5*1024*1024, backupCount=2)\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    handlers=[handler],\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s'\n",
    ")\n",
    "\n",
    "# Set Entrez email\n",
    "Entrez.email = \"ifeanyi.omah@ed.ac.uk\"\n",
    "\n",
    "# Initialize NCBI Taxonomy\n",
    "ncbi = NCBITaxa()\n",
    "# Optionally update taxonomy database\n",
    "# ncbi.update_taxonomy_database()\n",
    "\n",
    "valid_lineages = set()\n",
    "\n",
    "# Define column names for NR files (updated to match actual .m9 file columns)\n",
    "nr_column_names = [\n",
    "    'query', 'subject', 'identity', 'alignment_length', 'mismatches', 'gap_opens',\n",
    "    'q_start', 'q_end', 's_start', 's_end', 'evalue', 'bit_score', 'tax_id',\n",
    "    'sci_name', 'com_names', 'superkingdom', 'phylum', 'family',\n",
    "    'lca_taxid', 'lca_name'\n",
    "]\n",
    "\n",
    "# Define NR directory\n",
    "nr_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA/\"\n",
    "\n",
    "\n",
    "def load_lca(nr_directory):\n",
    "    \"\"\"\n",
    "    Load and process LCA results from NR files in the specified directory.\n",
    "\n",
    "    Parameters:\n",
    "        nr_directory (str): Path to the directory containing NR LCA files.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed LCA DataFrame.\n",
    "    \"\"\"\n",
    "    # Dynamically list all NR .m9 files that start with 'LCA_'\n",
    "    nr_file_paths = glob(f\"{nr_directory}/*.m9\")\n",
    "    logging.info(f\"Found {len(nr_file_paths)} NR files to process.\")\n",
    "\n",
    "    if not nr_file_paths:\n",
    "        logging.error(f\"No NR files found in directory: {nr_directory}\")\n",
    "        raise FileNotFoundError(f\"No NR files found in directory: {nr_directory}\")\n",
    "\n",
    "    # Load BLAST outputs for NR database\n",
    "    nr_df = load_blast_output(nr_file_paths, database='nr')\n",
    "\n",
    "    # Remove rows with missing taxid\n",
    "    initial_nr = len(nr_df)\n",
    "    nr_df = nr_df.dropna(subset=['taxid'])\n",
    "    final_nr = len(nr_df)\n",
    "    dropped_nr = initial_nr - final_nr\n",
    "    if dropped_nr > 0:\n",
    "        logging.info(f\"Dropped {dropped_nr} rows with missing taxid from nr_df.\")\n",
    "\n",
    "    # Convert taxid to integer\n",
    "    nr_df['taxid'] = nr_df['taxid'].astype(int)\n",
    "\n",
    "    # Retrieve all relevant taxids (including lineage)\n",
    "    taxids = set(nr_df['taxid'])\n",
    "    expanded_lineage = set()\n",
    "    for taxid in taxids:\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(taxid)\n",
    "            expanded_lineage.update(lineage)\n",
    "        except ValueError:\n",
    "            logging.warning(f\"Skipping {taxid}: taxid not found in local DB.\")\n",
    "    # Remove zero/invalid\n",
    "    expanded_lineage.discard(0)\n",
    "\n",
    "    # Build dictionaries for name/rank\n",
    "    taxid2name = ncbi.get_taxid_translator(expanded_lineage)\n",
    "    taxid2name[-1] = 'Ambiguous'\n",
    "    taxid2name[0] = 'Unknown'\n",
    "    taxid2rank = ncbi.get_rank(expanded_lineage)\n",
    "\n",
    "    def taxid2kingdom(tid):\n",
    "        \"\"\"\n",
    "        Retrieve the kingdom for a given taxid.\n",
    "        \"\"\"\n",
    "        if tid == 0:\n",
    "            return 'Unknown'\n",
    "        try:\n",
    "            lineage = ncbi.get_lineage(tid)\n",
    "            ranks = ncbi.get_rank(lineage)\n",
    "            for k, v in ranks.items():\n",
    "                if v == 'superkingdom':\n",
    "                    # Map taxid -> name\n",
    "                    kingdom_name = taxid2name.get(k, 'Unknown')\n",
    "                    if kingdom_name != 'Unknown' and tid != k:\n",
    "                        logging.warning(f\"taxid {tid} was translated into {kingdom_name}\")\n",
    "                    return kingdom_name\n",
    "        except ValueError:\n",
    "            return 'Unknown'\n",
    "        return 'Unknown'\n",
    "\n",
    "    # Map taxid -> name, rank, kingdom\n",
    "    nr_df['name'] = nr_df['taxid'].map(taxid2name)\n",
    "    nr_df['rank'] = nr_df['taxid'].map(taxid2rank)\n",
    "    nr_df['kingdom'] = nr_df['taxid'].apply(taxid2kingdom)\n",
    "    nr_df['db'] = 'nr'  # Tag as NR data\n",
    "\n",
    "    return nr_df\n",
    "\n",
    "\n",
    "def load_blast_output(files, database='nr'):\n",
    "    \"\"\"\n",
    "    Load BLAST output files into a single DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        files (list): List of file paths to BLAST output files.\n",
    "        database (str): Type of database ('nr').\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Concatenated BLAST results.\n",
    "    \"\"\"\n",
    "\n",
    "    def get_sample_from_path(path):\n",
    "        \"\"\"\n",
    "        Extract sample name from filename using regex:\n",
    "        LCA_SOMESAMPLE_modified_diamond_blast_nr_annotated.m9\n",
    "        \"\"\"\n",
    "        filename = path.split('/')[-1]\n",
    "        # Regex capturing portion after \"LCA_\" and before \"_modified...\"\n",
    "        match = re.match(r'^LCA_(.+?)_modified_diamond_blast_nr_annotated\\.m9$', filename)\n",
    "        if match:\n",
    "            return match.group(1)  # Everything in the capturing group\n",
    "        else:\n",
    "            logging.warning(f\"Could not extract sample name from filename: {filename}\")\n",
    "            return \"unknown_sample\"\n",
    "\n",
    "    # Define column names based on database\n",
    "    if database == 'nr':\n",
    "        column_names = nr_column_names\n",
    "    else:\n",
    "        logging.error(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "        raise ValueError(f\"Unknown database type: {database}. Expected 'nr'.\")\n",
    "\n",
    "    dfs = []\n",
    "\n",
    "    for file in files:\n",
    "        try:\n",
    "            # Quick check of columns in the first line\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                columns_in_file = len(first_line.strip().split('\\t'))\n",
    "                expected_columns = len(column_names)\n",
    "                if columns_in_file != expected_columns:\n",
    "                    logging.warning(\n",
    "                        f\"{file} has {columns_in_file} columns; \"\n",
    "                        f\"expected {expected_columns}.\"\n",
    "                    )\n",
    "\n",
    "            # Determine if the file has a header row\n",
    "            with open(file, 'r') as f:\n",
    "                first_line = f.readline()\n",
    "                if 'sci_name' in first_line.lower() or 'lca_taxid' in first_line.lower():\n",
    "                    skiprows = 1\n",
    "                else:\n",
    "                    skiprows = 0\n",
    "\n",
    "            df = pd.read_csv(\n",
    "                file,\n",
    "                sep='\\t',\n",
    "                header=None,\n",
    "                names=column_names,\n",
    "                na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "                skiprows=skiprows,\n",
    "                on_bad_lines='warn'  # or 'error' if you want strict\n",
    "            )\n",
    "\n",
    "            # Debug: Log the first few rows\n",
    "            logging.info(f\"First few rows of {file} after reading:\")\n",
    "            logging.info(df.head())\n",
    "\n",
    "            # Rename 'tax_id' to 'taxid' if present\n",
    "            if 'tax_id' in df.columns:\n",
    "                df.rename(columns={'tax_id': 'taxid'}, inplace=True)\n",
    "            elif 'taxid' not in df.columns:\n",
    "                logging.error(f\"'tax_id' or 'taxid' column not found in {file}\")\n",
    "                raise KeyError(f\"'tax_id' or 'taxid' column not found in {file}.\")\n",
    "\n",
    "            # Replace 'Unknown' with '0'\n",
    "            df['taxid'] = df['taxid'].replace('Unknown', '0')\n",
    "\n",
    "            # Convert 'taxid' to numeric\n",
    "            df['taxid'] = pd.to_numeric(df['taxid'], errors='coerce')\n",
    "\n",
    "            # Drop rows with NaN in 'taxid'\n",
    "            before_drop = len(df)\n",
    "            df = df.dropna(subset=['taxid'])\n",
    "            after_drop = len(df)\n",
    "            dropped = before_drop - after_drop\n",
    "            if dropped > 0:\n",
    "                logging.info(f\"Dropped {dropped} rows with invalid taxid in {file}.\")\n",
    "\n",
    "            # Derive sample name from the filename\n",
    "            df['sample'] = get_sample_from_path(file)\n",
    "\n",
    "            # Extract contig from 'query' by removing sample prefix\n",
    "            df['contig'] = df.apply(\n",
    "                lambda row: extract_contig(row['query'], row['sample']),\n",
    "                axis=1\n",
    "            )\n",
    "\n",
    "            # alignment_length -> align_length\n",
    "            df.rename(columns={'alignment_length': 'align_length'}, inplace=True)\n",
    "\n",
    "            # Build contig_key\n",
    "            df['contig_key'] = (df['sample'] + '~' + df['contig']).str.lower()\n",
    "\n",
    "            # Debug: Log the created 'contig_key'\n",
    "            logging.info(f\"Created 'contig_key' for {file}:\")\n",
    "            logging.info(df[['contig_key']].head())\n",
    "\n",
    "            # Convert numeric columns\n",
    "            numeric_columns = ['identity', 'align_length', 'bit_score', 'evalue']\n",
    "            for col in numeric_columns:\n",
    "                if col in df.columns:\n",
    "                    df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "                else:\n",
    "                    logging.warning(f\"Column '{col}' not found in {file}.\")\n",
    "\n",
    "            # Optionally drop rows where numeric columns are NaN\n",
    "            before_num_drop = len(df)\n",
    "            df = df.dropna(subset=numeric_columns)\n",
    "            after_num_drop = len(df)\n",
    "            dropped_num = before_num_drop - after_num_drop\n",
    "            if dropped_num > 0:\n",
    "                logging.info(f\"Dropped {dropped_num} rows with non-numeric values in {file}.\")\n",
    "\n",
    "            dfs.append(df)\n",
    "        except pd.errors.ParserError as e:\n",
    "            logging.error(f\"ParserError while reading {file}: {e}\")\n",
    "            print(f\"ParserError while reading {file}: {e}\", file=sys.stderr)\n",
    "        except KeyError as e:\n",
    "            logging.error(f\"KeyError while processing {file}: {e}\")\n",
    "            print(f\"KeyError while processing {file}: {e}\", file=sys.stderr)\n",
    "            raise e\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Unexpected error while reading {file}: {e}\")\n",
    "            print(f\"Unexpected error while reading {file}: {e}\", file=sys.stderr)\n",
    "            raise e\n",
    "\n",
    "    if dfs:\n",
    "        combined_df = pd.concat(dfs, ignore_index=True)\n",
    "    else:\n",
    "        combined_df = pd.DataFrame()\n",
    "        logging.warning(\"No dataframes to concatenate. Returning empty DataFrame.\")\n",
    "\n",
    "    return combined_df\n",
    "\n",
    "\n",
    "def extract_contig(query, sample):\n",
    "    \"\"\"\n",
    "    Extract contig information by removing the sample prefix from the query.\n",
    "    If the query doesn't start with 'sample_', label it 'unknown_contig'.\n",
    "    \"\"\"\n",
    "    prefix = sample + '_'\n",
    "    if query.startswith(prefix):\n",
    "        return query[len(prefix):]\n",
    "    else:\n",
    "        logging.warning(\n",
    "            f\"Query '{query}' does not start with sample prefix '{prefix}'. \"\n",
    "            f\"Assigning 'unknown_contig'.\"\n",
    "        )\n",
    "        return 'unknown_contig'\n",
    "\n",
    "\n",
    "def load_cdhit_clusters(filename):\n",
    "    \"\"\"\n",
    "    Load CD-HIT cluster information from a cluster file.\n",
    "    \"\"\"\n",
    "    clusters = defaultdict(list)\n",
    "    cluster_id = None\n",
    "\n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>Cluster'):\n",
    "                cluster_id = line.strip().split()[-1]\n",
    "            else:\n",
    "                if cluster_id is not None:\n",
    "                    try:\n",
    "                        member = parse_cdhit_row(line)\n",
    "                        clusters[cluster_id].append(member)\n",
    "                    except ValueError as e:\n",
    "                        logging.error(f\"Error parsing line: {line.strip()} - {e}\")\n",
    "    return clusters\n",
    "\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    \"\"\"\n",
    "    Parse a single line from a CD-HIT cluster file.\n",
    "    \"\"\"\n",
    "    if '*' in row:\n",
    "        # Reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = True\n",
    "    else:\n",
    "        # Non-reference member\n",
    "        parts = row.strip().split()\n",
    "        if len(parts) < 4:\n",
    "            raise ValueError(f\"Unexpected format in row: {row}\")\n",
    "        index, length_str, name, percent_id = parts[:4]\n",
    "        is_ref = False\n",
    "\n",
    "    # Extract numeric length\n",
    "    match_len = re.match(r'(\\d+)', length_str)\n",
    "    if match_len:\n",
    "        length = int(match_len.group(1))\n",
    "    else:\n",
    "        logging.warning(f\"Could not extract numeric length from '{length_str}' in row: {row}\")\n",
    "        length = 0\n",
    "\n",
    "    # Cleanup name\n",
    "    name = name.strip('>').strip('.')\n",
    "\n",
    "    # Extract sample, contig\n",
    "    try:\n",
    "        sample, contig = name.split('~')\n",
    "    except ValueError:\n",
    "        # Unexpected naming\n",
    "        sample, contig = 'unknown_sample', name\n",
    "\n",
    "    return {\n",
    "        'contig': contig,\n",
    "        'sample': sample,\n",
    "        'length': length,\n",
    "        'is_ref': is_ref\n",
    "    }\n",
    "\n",
    "\n",
    "def get_cluster_rep(cluster):\n",
    "    \"\"\"\n",
    "    Get the representative member of a cluster (the reference member).\n",
    "    \"\"\"\n",
    "    for member in cluster:\n",
    "        if member['is_ref']:\n",
    "            return member\n",
    "    # If no reference member, just return the first\n",
    "    return cluster[0] if cluster else None\n",
    "\n",
    "\n",
    "def merge_clusters_lca(clusters, lca_df):\n",
    "    \"\"\"\n",
    "    Merge cluster information with LCA results.\n",
    "    \"\"\"\n",
    "    cluster_ids = [int(cid.strip()) for cid in clusters.keys()]\n",
    "    cluster_reps = [get_cluster_rep(clusters[str(cid)]) for cid in cluster_ids]\n",
    "    cluster_lengths = [rep['length'] for rep in cluster_reps]\n",
    "    cluster_sizes = [len(clusters[str(cid)]) for cid in cluster_ids]\n",
    "    # Build contig_key\n",
    "    cluster_contig_keys = [f\"{rep['sample']}~{rep['contig']}\".lower() for rep in cluster_reps]\n",
    "\n",
    "    df_clusters = pd.DataFrame({\n",
    "        'contig_key': cluster_contig_keys,\n",
    "        'cluster': cluster_ids,\n",
    "        'cluster_size': cluster_sizes,\n",
    "        'contig_length': cluster_lengths\n",
    "    })\n",
    "\n",
    "    # Lowercase & strip\n",
    "    df_clusters['contig_key'] = df_clusters['contig_key'].str.strip().str.lower()\n",
    "\n",
    "    # Debug\n",
    "    logging.info(\"Sample 'contig_key's in clusters DataFrame:\")\n",
    "    logging.info(df_clusters['contig_key'].head())\n",
    "\n",
    "    # Merge\n",
    "    merged_df = df_clusters.merge(lca_df, on='contig_key', how='inner')\n",
    "    logging.info(f\"Merged DataFrame has {len(merged_df)} rows.\")\n",
    "\n",
    "    if merged_df.empty:\n",
    "        logging.warning(\"merged_df is empty after merging. Check 'contig_key' consistency.\")\n",
    "        print(\n",
    "            \"merged_df is empty after merging. \"\n",
    "            \"Please check 'contig_key' consistency between clusters and LCA DataFrames.\",\n",
    "            file=sys.stderr\n",
    "        )\n",
    "    else:\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "\n",
    "    # Add align_percent\n",
    "    merged_df['align_percent'] = np.round(\n",
    "        100 * merged_df['align_length'] / merged_df['contig_length'], 1\n",
    "    )\n",
    "    # Use 'identity' as 'percent_identity'\n",
    "    merged_df['percent_identity'] = np.round(merged_df['identity'], 1)\n",
    "\n",
    "    # Required columns\n",
    "    required_columns = [\n",
    "        'cluster', 'cluster_size', 'contig_length', 'name', 'rank',\n",
    "        'kingdom', 'taxid', 'db', 'align_length', 'align_percent',\n",
    "        'percent_identity', 'bit_score', 'evalue', 'subject', 'sci_name',\n",
    "        'com_names', 'superkingdom', 'phylum', 'family', 'lca_taxid', 'lca_name',\n",
    "        'contig', 'sample', 'contig_key'\n",
    "    ]\n",
    "\n",
    "    missing_columns = [c for c in required_columns if c not in merged_df.columns]\n",
    "    if missing_columns:\n",
    "        logging.error(f\"Missing columns in merged DataFrame: {missing_columns}\")\n",
    "        logging.info(f\"Available columns: {merged_df.columns.tolist()}\")\n",
    "        raise KeyError(f\"{missing_columns} not in index\")\n",
    "\n",
    "    # Subset only required columns\n",
    "    merged_df = merged_df[required_columns]\n",
    "    merged_df = merged_df.sort_values('cluster').reset_index(drop=True)\n",
    "\n",
    "    return merged_df\n",
    "\n",
    "\n",
    "def load_counts_data(counts_file):\n",
    "    \"\"\"\n",
    "    Load and process the counts DataFrame.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        counts = pd.read_csv(\n",
    "            counts_file,\n",
    "            sep=',',\n",
    "            header=0,\n",
    "            usecols=[0, 1, 2],\n",
    "            names=['contig', 'count', 'length'],\n",
    "            dtype={'contig': str, 'count': str, 'length': str},\n",
    "            na_values=['', 'NA', 'nan', 'Unknown'],\n",
    "            low_memory=False\n",
    "        )\n",
    "\n",
    "        # Debug\n",
    "        logging.info(\"Raw Counts DataFrame sample:\")\n",
    "        logging.info(counts.head())\n",
    "\n",
    "        counts['count'] = pd.to_numeric(counts['count'], errors='coerce')\n",
    "\n",
    "        logging.info(\"Counts DataFrame after converting 'count' to numeric:\")\n",
    "        logging.info(counts.head())\n",
    "\n",
    "        # Extract sample, contig from 'contig'\n",
    "        counts[['sample', 'contig']] = counts['contig'].str.split('~', expand=True)\n",
    "        counts['contig'] = counts['contig'].fillna(counts['contig'])\n",
    "        counts['sample'] = counts['sample'].fillna('unknown_sample')\n",
    "\n",
    "        # Build contig_key\n",
    "        counts['contig_key'] = (\n",
    "            counts['sample'] + '~' + counts['contig']\n",
    "        ).str.strip().str.lower()\n",
    "\n",
    "        # Debug\n",
    "        logging.info(\"Processed Counts DataFrame:\")\n",
    "        logging.info(counts.head())\n",
    "\n",
    "        return counts\n",
    "    except pd.errors.ParserError as e:\n",
    "        logging.error(f\"ParserError while reading counts file {counts_file}: {e}\")\n",
    "        print(f\"ParserError while reading counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error while processing counts file {counts_file}: {e}\")\n",
    "        print(f\"Unexpected error while processing counts file {counts_file}: {e}\", file=sys.stderr)\n",
    "        raise e\n",
    "\n",
    "\n",
    "def load_taxid2name(lca_df):\n",
    "    \"\"\"\n",
    "    Create a taxid to name mapping from the LCA DataFrame.\n",
    "    \"\"\"\n",
    "    taxid2name = dict(zip(lca_df['taxid'], lca_df['name']))\n",
    "    taxid2name[-1] = 'Ambiguous'\n",
    "    taxid2name[0] = 'Unknown'\n",
    "    return taxid2name\n",
    "\n",
    "\n",
    "#\n",
    "# Main execution (Example usage)\n",
    "#\n",
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        # 1) Load LCA results\n",
    "        lca_df = load_lca(nr_dir)\n",
    "        if lca_df.empty:\n",
    "            logging.error(\"LCA DataFrame is empty. Exiting.\")\n",
    "            raise ValueError(\"LCA DataFrame is empty.\")\n",
    "        logging.info(\"Loaded LCA DataFrame.\")\n",
    "\n",
    "        # Debugging step: check LCA keys\n",
    "        unique_lca_keys = set(lca_df['contig_key'])\n",
    "        print(\"Number of unique LCA contig_keys:\", len(unique_lca_keys))\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load LCA DataFrame: {e}\")\n",
    "        print(f\"Error loading LCA DataFrame: {e}\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    try:\n",
    "        # 2) Load CD-HIT clusters\n",
    "        clusters = load_cdhit_clusters('/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Combined_filtered_500_SENZOR_clsuter.clstr'\n",
    "        )\n",
    "        logging.info(\"Loaded CD-HIT clusters.\")\n",
    "\n",
    "        # Debug: convert clusters -> DataFrame just to see how many contig_keys it has\n",
    "        cluster_ids = clusters.keys()\n",
    "        cluster_reps = [get_cluster_rep(clusters[cid]) for cid in cluster_ids]\n",
    "        df_clusters = pd.DataFrame({\n",
    "            'sample': [rep['sample'] for rep in cluster_reps],\n",
    "            'contig': [rep['contig'] for rep in cluster_reps]\n",
    "        })\n",
    "        df_clusters['contig_key'] = (\n",
    "            df_clusters['sample'] + '~' + df_clusters['contig']\n",
    "        ).str.lower()\n",
    "\n",
    "        unique_cluster_keys = set(df_clusters['contig_key'])\n",
    "        print(\"Number of unique cluster contig_keys:\", len(unique_cluster_keys))\n",
    "\n",
    "        # Check what's missing from LCA but not in cluster\n",
    "        missing_in_clusters = unique_lca_keys.difference(unique_cluster_keys)\n",
    "        print(f\"Number of LCA contig_keys not found in clusters: {len(missing_in_clusters)}\")\n",
    "        if missing_in_clusters:\n",
    "            print(\"Sample missing contig_keys (cluster):\", list(missing_in_clusters)[:10], \"...\")\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load CD-HIT clusters: {e}\")\n",
    "        print(f\"Error loading CD-HIT clusters: {e}\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    try:\n",
    "        # 3) Merge clusters with LCA\n",
    "        merged_df = merge_clusters_lca(clusters, lca_df)\n",
    "        logging.info(\"Successfully merged clusters with LCA results.\")\n",
    "    except KeyError as e:\n",
    "        logging.error(f\"KeyError during merging: {e}\")\n",
    "        print(f\"Error during merging: {e}\", file=sys.stderr)\n",
    "        sys.exit(1)\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Unexpected error during merging: {e}\")\n",
    "        print(f\"Unexpected error during merging: {e}\", file=sys.stderr)\n",
    "        sys.exit(1)\n",
    "\n",
    "    try:\n",
    "        # 4) Create clust2kingdom dictionary\n",
    "        clust2kingdom = dict(zip(merged_df['cluster'], merged_df['kingdom']))\n",
    "        logging.info(\"Created clust2kingdom dictionary.\")\n",
    "\n",
    "        # Save clust2kingdom to CSV\n",
    "        merged_df[['cluster', 'kingdom']].to_csv(\n",
    "            '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/cluster_kingdoms.csv',\n",
    "            index=False\n",
    "        )\n",
    "        logging.info(\"Saved cluster_kingdoms.csv successfully.\")\n",
    "\n",
    "        # Print sample entries of clust2kingdom\n",
    "        print(\"clust2kingdom Dictionary Sample Entries:\")\n",
    "        for cluster_id, kingdom in list(clust2kingdom.items())[:5]:\n",
    "            print(f\"Cluster {cluster_id}: Kingdom {kingdom}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error creating clust2kingdom: {e}\")\n",
    "        print(f\"Error creating clust2kingdom: {e}\", file=sys.stderr)\n",
    "        sys.exit(1)\n",
    "\n",
    "    try:\n",
    "        # 5) Load and process the counts DataFrame\n",
    "        counts_file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/combined_SENSOR_csvs/1Modified_CSV/merged_contig_stats.csv'\n",
    "        counts_df = load_counts_data(counts_file_path)\n",
    "        logging.info(\"Loaded and processed counts DataFrame.\")\n",
    "\n",
    "        # Debug: check how many contig_keys in counts\n",
    "        unique_counts_keys = set(counts_df['contig_key'])\n",
    "        missing_in_counts = unique_lca_keys.difference(unique_counts_keys)\n",
    "        print(f\"Number of LCA contig_keys not found in counts: {len(missing_in_counts)}\")\n",
    "        if missing_in_counts:\n",
    "            print(\"Sample missing contig_keys (counts):\", list(missing_in_counts)[:10], \"...\")\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Failed to load counts DataFrame: {e}\")\n",
    "        print(f\"Error loading counts DataFrame: {e}\", file=sys.stderr)\n",
    "        sys.exit(1)\n",
    "\n",
    "    try:\n",
    "        # 6) Merge counts with merged_df on 'contig_key'\n",
    "        final_df = merged_df.merge(counts_df[['contig_key', 'count']], on='contig_key', how='left')\n",
    "        logging.info(\"Merged counts with merged_df to create final_df.\")\n",
    "\n",
    "        # Convert 'count' to numeric if not already\n",
    "        final_df['count'] = pd.to_numeric(final_df['count'], errors='coerce')\n",
    "\n",
    "        # Filter: remove rows where 'count' is NaN\n",
    "        initial_count = len(final_df)\n",
    "        final_df = final_df.dropna(subset=['count'])\n",
    "        final_count = len(final_df)\n",
    "        removed_rows = initial_count - final_count\n",
    "        if removed_rows > 0:\n",
    "            logging.info(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "            print(f\"Filtered out {removed_rows} rows without 'count' value.\")\n",
    "\n",
    "        # Filter out rows where 'count' <= 0\n",
    "        final_df = final_df[final_df['count'] > 0]\n",
    "        logging.info(\"Filtered out rows with 'count' <= 0.\")\n",
    "\n",
    "        if final_df.empty:\n",
    "            logging.warning(\"final_df is empty after filtering out rows with no/low 'count'.\")\n",
    "            print(\"final_df is empty after filtering out rows without a valid 'count'.\", file=sys.stderr)\n",
    "        else:\n",
    "            # Show final DataFrame\n",
    "            print(\"Final Merged DataFrame with Counts (after filtering):\")\n",
    "            print(final_df.head())\n",
    "\n",
    "            # Save the final DataFrame\n",
    "            final_df.to_csv(\n",
    "                '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/Complete_contig_counts.csv',\n",
    "                index=False\n",
    "            )\n",
    "            logging.info(\"Saved final_merged_with_counts.csv successfully.\")\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error merging counts with merged_df: {e}\")\n",
    "        print(f\"Error merging counts with merged_df: {e}\", file=sys.stderr)\n",
    "        sys.exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "101873d4-1646-47cc-9f5b-18598369fded",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SEZONOWIFGOAT011_S29_2_656747', 'AIAMAGOAT001_S48_656109',\n",
       "       'AINWZGOAT001_S41_656115', 'AIAMAGOAT002_S49_656110',\n",
       "       'SEZOOWOGCHICKEN001_S35_655749', 'SEZONOWOGDOG014_S21_657751',\n",
       "       'SEZONOWGOAT009_S67_2_656740', 'AINWZGOAT004_S44_1_656183',\n",
       "       'SEZONOWIFRR033_LIV_S73_666402', 'AINWZGOAT007_S47_2_656739',\n",
       "       'AIMUZSHEEP001_S52_656081', 'SEZONOWOGR004_LIV_S50_666375',\n",
       "       'AIAMAR011_LIV_S31_663039', 'AINWZGOAT005_S45_1_656184',\n",
       "       'SEZONOWDOG01_S7_657737', 'AINWZLIZARD001_S58_657718',\n",
       "       'AINWZGOAT006_S46_1_656185', 'SEZONOWISSHEEP001_S32_656084',\n",
       "       'SEZONOWISDOG002_S4_657741', 'AINWZDOG001_S37_657734',\n",
       "       'SEZONOWISDOG015_S13_657743', 'AIMGBR023_LIV_S43_662128',\n",
       "       'SEZONOWISLIZARD001_S36_657719', 'SEZONOWOGR006_LIV_S52_666377',\n",
       "       'AIMGBR021_LIV_S41_663052', 'AINWZR015_LIV_S15_663078',\n",
       "       'AINWZR008_LIV_S8_663067', 'AINWZPIG001_S62_657715',\n",
       "       'SEZONOWOGR002_LIV_S48_666373', 'AINWZR009_LIV_S9_663068',\n",
       "       'SEZONOWDOG012_S10_657739', 'SEZONOWOWISR088_LIV_S69_666424',\n",
       "       'AINWZR017_LIV_S17_663080', 'SEZONOWIFSHEEP001_S33_656082',\n",
       "       'SEZONOWOGR005_LIV_S51_666376', 'AINWZR003_LIV_S3_662133',\n",
       "       'SEZONOWIFRR063_LUNG_S81_666397', 'AINWZR001_LIV_S1_662131',\n",
       "       'AIAMADOG001_S40_657732', 'SEZONOWIFRR038_LIV_S78_666400',\n",
       "       'AIAMAR012_LIV_S32_663040', 'SEZONOWISDOG008_S14_657742',\n",
       "       'AIMGLGOAT001_S60_656113', 'SEZONOWISDOGOO1_S19_657744',\n",
       "       'SEZONOWOGDOG002_S5_657747', 'AIAMAR004_LIV_S24_1_662161',\n",
       "       'AIAMAGOAT003_S50_657720', 'SEZONOWIFRR040_LIV_S80_666396',\n",
       "       'SEZONOWOGR081_LUNG_S57_666409', 'AINWZR006_LIV_S6_663065',\n",
       "       'AIAMAR019_LIV_S39_663050', 'AIAMAR017_LIV_S37_663048',\n",
       "       'AIAMAR002_LIV_S22_662107', 'AINWZR002_LIV_S2_662132',\n",
       "       'AINWZR011_LIV_S11_663070', 'SEZONOWIFSHEEP002_S31_656083',\n",
       "       'AINWZR014_LIV_S14_663073', 'AINWZR007_LIV_S7_663066',\n",
       "       'AIAMAR020_LIV_S40_663051', 'SEZONOWIFCAT001_S34_657717',\n",
       "       'SEZONOWOWISR089_LUNG_S70_666425', 'AIMGBR024_LIV_S44_662129',\n",
       "       'AINWZR016_LIV_S16_663079', 'SEZONOWISUGOAT003_S26_657724',\n",
       "       'AIAMAR009_LIV_S29_1_662168', 'AINWZR005_LIV_S5_663064',\n",
       "       'AINWZR010_LIV_S10_663069', 'AINWZR020_LIV_S20_663083',\n",
       "       'SEZONOWIFDOG009_S1_657740', 'SEZOOWISDOG003_S11_657754',\n",
       "       'AIAMAR003_LIV_S23_662108', 'AINWZCHICKEN001_S56_655747',\n",
       "       'SEZONOWOGR008_LIV_S54_666371', 'AIAMACAT001_S69_657769',\n",
       "       'SEZONOWOGGOAT015_S28_657725', 'SEZONOWOGR082_LUNG_S58_666410',\n",
       "       'SEZOOWOGDOG006_S16_1_657760', 'AIMGBR022_LIV_S42_663077',\n",
       "       'AIAMAR013_LIV_S33_663041', 'AINWZR018_LIV_S18_663081',\n",
       "       'SEZONOWOWISR087_LUNG_S68_666423', 'SEZOOWOGDOG005_S15_657756',\n",
       "       'SEZONOWOGGOAT018_S66_1_657730', 'SEZONOWOGDOG010_S3_657749',\n",
       "       'AIAMAPIG001_S63_657713', 'SEZONOWOWISR005_LIV_S66_666419',\n",
       "       'SEZONOWOGDOG00_S2_657745', 'SEZONOWOGR009_LIV_S55_666372',\n",
       "       'AIAMAAGR001_LIV_S46_662105', 'SEZONOWOGDOG017_S22_657753',\n",
       "       'SEZONOWGOAT010_S65_2_656741', 'AIAMAR001_LIV_S21_662106',\n",
       "       'SEZOOWISUGOAT002_S27_657721', 'AIAMAR015_LIV_S35_663046',\n",
       "       'SEZONOWOGR003_LIV_S49_666374', 'AIAMAR008_LIV_S28_1_662167',\n",
       "       'SEZONOWOWISR086_LIV_S67_666422', 'AIAMAR016_LIV_S36_663047',\n",
       "       'AIAMAR018_LIV_S38_663049', 'AINWZR013_LIV_S13_663072',\n",
       "       'SEZONOWIFRR032_LIV_S72_666406', 'SEZONOWIFGOAT007_S68_657731',\n",
       "       'SEZONOWISGOAT013_S24_2_657729', 'AIAMASHEEP002_S54_656079',\n",
       "       'AINWZR012_LIV_S12_663071', 'AIMGLGOAT002_S61_656114',\n",
       "       'AIAMAR005_LIV_S25_1_662162', 'SEZOOWOGDOG003_S17_657755',\n",
       "       'AINWZDOG002_S38_657735', 'AIAMAR006_LIV_S26_2_662165',\n",
       "       'SEZONOWOGR010_LIV_S56_666407', 'AINWZDOG003_S39_657736',\n",
       "       'AIAMAR007_LIV_S27_2_662166', 'SEZONOWIFRR031_LIV_S71_663085',\n",
       "       'SEZONOWIFGOAT012_S30_657726', 'SEZONOWIFRR039_LIV_S79_666401',\n",
       "       'SEZONOWOGR007_LIV_S53_666370', 'AINWZAGR001_LIV_S45_662130',\n",
       "       'SEZONOWOWISR003_LIV_S64_666417', 'SSP_ERCC_S72_670948',\n",
       "       'SEZONOWOWISR002_LIV_S63_666416', 'SSP_ERCC_S83_666426',\n",
       "       'SEZONOWDOG009_S20_657738', 'AIMASHEEP003_S55_656080',\n",
       "       'SEZOOWOGDOG011_S8_1_657761', 'AIAMAR010_LIV_S30_663038',\n",
       "       'SEZONOWIFRR037_LIV_S77_666399', 'SEZONOWOGDOG001_S6_657746',\n",
       "       'AINWZR004_LIV_S4_662134', 'AIAMASHEEP001_S53_656078',\n",
       "       'SEZONOWOGDOG016_S23_657752', 'AIAMAGOAT004_S51_656112',\n",
       "       'AIAMAR014_LIV_S34_663042', 'SEZONOWOGDOG004_S12_657748',\n",
       "       'AIAMAPIG002_S64_657714', 'SEZONOWISGOAT008_S25_657722',\n",
       "       'AIMGLDOG001_S59_657733', 'SEZONOWOGDOG013_S18_657750',\n",
       "       'SEZONOWOGR083_LUNG_S59_3_666413', 'AINWZCHICKEN002_S57_655748',\n",
       "       'AINWZGOAT003_S43_1_656182', 'SEZONOWIFRR034_LIV_S74_1_666404',\n",
       "       'SEZONOWOWISR001_LIV_S62_666415', 'SEZOOWOGDOG019_S9_1_657762',\n",
       "       'SEZONOWOWISR004_LIV_S65_666418', 'AINWZR019_LIV_S19_663082',\n",
       "       'SEZONOWOGR001_LIV_S47_666398', 'AINWZGOAT002_S42_656116',\n",
       "       'NE_S70_670946', 'SEZONOWOGR085_LUNG_S61_666414', 'NE_S82_663084',\n",
       "       'NTC_S71_670947'], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File path\n",
    "file_path = '/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_results/final_merged_with_counts.csv'\n",
    "\n",
    "# Load the CSV file\n",
    "data = pd.read_csv(file_path)\n",
    "\n",
    "# Get unique values for 'sample' column\n",
    "unique_samples = data['sample'].unique()\n",
    "\n",
    "# Display the unique values\n",
    "unique_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ec74aa7e-138c-4e1e-a0ed-3c187da6d3de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample\n",
       "SEZONOWIFGOAT011_S29_2_656747     18230\n",
       "AIAMAGOAT003_S50_657720            7213\n",
       "AIAMAGOAT002_S49_656110            4227\n",
       "AINWZLIZARD001_S58_657718          2717\n",
       "AIMGLDOG001_S59_657733             2609\n",
       "                                  ...  \n",
       "SEZONOWOWISR001_LIV_S62_666415       23\n",
       "NTC_S71_670947                       19\n",
       "AINWZR019_LIV_S19_663082             12\n",
       "SEZONOWOGR001_LIV_S47_666398          5\n",
       "NE_S82_663084                         3\n",
       "Name: count, Length: 152, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_samples_count = data['sample'].value_counts()\n",
    "\n",
    "# Display the counts of unique values\n",
    "unique_samples_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7e39215e-ed45-48fd-bda2-be820608d78c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/4469 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59451 was translated into 3371016\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2650730 was translated into 261694\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 4469/4469 [00:00<00:00, 139303.09it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/8069 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 143291 was translated into 132908\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 8069/8069 [00:00<00:00, 201286.08it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/7037 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 228257 was translated into 3082113\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 7037/7037 [00:00<00:00, 203683.15it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/7899 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 90247 was translated into 3371155\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 7899/7899 [00:00<00:00, 390900.92it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/7820 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1906744 was translated into 1796644\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 7820/7820 [00:00<00:00, 284167.43it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/4546 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991676 was translated into 3383035\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 85948 was translated into 147567\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 4546/4546 [00:00<00:00, 180245.84it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 26952/26952 [00:00<00:00, 492002.18it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6103/6103 [00:00<00:00, 204615.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 505/505 [00:00<00:00, 156538.58it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3892/3892 [00:00<00:00, 381603.42it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 10313/10313 [00:00<00:00, 441422.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9125/9125 [00:00<00:00, 400341.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                     | 0/15092 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2991674 was translated into 3383038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 15092/15092 [00:00<00:00, 301585.75it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                     | 0/28432 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2052317 was translated into 3107663\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 7604 was translated into 133437\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 28432/28432 [00:00<00:00, 502456.20it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 36352/36352 [00:00<00:00, 782686.99it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7366/7366 [00:00<00:00, 593215.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8268/8268 [00:00<00:00, 399154.07it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                     | 0/10461 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1914757 was translated into 1768011\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2789583 was translated into 2841044\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 10461/10461 [00:00<00:00, 201719.50it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1914/1914 [00:00<00:00, 341598.14it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2641/2641 [00:00<00:00, 543850.99it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1986/1986 [00:00<00:00, 526674.74it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/2387 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 39086 was translated into 3370998\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 2387/2387 [00:00<00:00, 329889.08it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2379/2379 [00:00<00:00, 635233.59it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2533/2533 [00:00<00:00, 561145.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2379/2379 [00:00<00:00, 472701.18it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2544/2544 [00:00<00:00, 689920.43it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2190/2190 [00:00<00:00, 607990.85it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3040/3040 [00:00<00:00, 270267.59it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1979/1979 [00:00<00:00, 531846.45it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2214/2214 [00:00<00:00, 610210.87it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2111/2111 [00:00<00:00, 596843.66it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2104/2104 [00:00<00:00, 694593.91it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2385/2385 [00:00<00:00, 756401.89it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2400/2400 [00:00<00:00, 288350.89it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2076/2076 [00:00<00:00, 685620.09it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2329/2329 [00:00<00:00, 823792.72it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2042/2042 [00:00<00:00, 730595.31it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2033/2033 [00:00<00:00, 658966.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9428/9428 [00:00<00:00, 425628.83it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4236/4236 [00:00<00:00, 538037.42it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/5892 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817377 was translated into 2817376\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2811023 was translated into 1903157\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 139420 was translated into 2498619\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 5892/5892 [00:00<00:00, 195319.81it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2457/2457 [00:00<00:00, 771189.47it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2485/2485 [00:00<00:00, 662609.37it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2479/2479 [00:00<00:00, 542789.71it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2050/2050 [00:00<00:00, 613246.07it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 18791/18791 [00:00<00:00, 813173.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9527/9527 [00:00<00:00, 908451.19it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                     | 0/14198 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 436010 was translated into 1759441\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 14198/14198 [00:00<00:00, 262589.03it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7693/7693 [00:00<00:00, 888451.48it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5431/5431 [00:00<00:00, 791606.37it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2805/2805 [00:00<00:00, 558775.72it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 13103/13103 [00:00<00:00, 798668.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 598/598 [00:00<00:00, 402663.96it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8891/8891 [00:00<00:00, 1009079.90it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8077/8077 [00:00<00:00, 508852.94it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 13932/13932 [00:00<00:00, 808834.31it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9695/9695 [00:00<00:00, 711975.65it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 11086/11086 [00:00<00:00, 523762.40it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1291/1291 [00:00<00:00, 482778.75it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1294/1294 [00:00<00:00, 493492.40it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 739/739 [00:00<00:00, 415606.15it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 522/522 [00:00<00:00, 382700.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8853/8853 [00:00<00:00, 489741.14it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1993/1993 [00:00<00:00, 293482.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5370/5370 [00:00<00:00, 492886.02it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1270/1270 [00:00<00:00, 341941.59it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4180/4180 [00:00<00:00, 546259.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7042/7042 [00:00<00:00, 791772.70it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 19173/19173 [00:00<00:00, 742796.62it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9581/9581 [00:00<00:00, 903169.56it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6461/6461 [00:00<00:00, 686825.78it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 10473/10473 [00:00<00:00, 671861.03it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7688/7688 [00:00<00:00, 843755.64it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9494/9494 [00:00<00:00, 1008936.92it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8958/8958 [00:00<00:00, 562303.76it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3699/3699 [00:00<00:00, 291444.01it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1948/1948 [00:00<00:00, 860415.35it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2285/2285 [00:00<00:00, 865214.83it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2167/2167 [00:00<00:00, 745432.36it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2414/2414 [00:00<00:00, 800715.69it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2316/2316 [00:00<00:00, 835469.86it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2007/2007 [00:00<00:00, 841207.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2172/2172 [00:00<00:00, 873360.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2525/2525 [00:00<00:00, 1012487.34it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2441/2441 [00:00<00:00, 954708.70it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2723/2723 [00:00<00:00, 247322.15it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2267/2267 [00:00<00:00, 839305.07it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2283/2283 [00:00<00:00, 902761.95it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2506/2506 [00:00<00:00, 651436.37it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2554/2554 [00:00<00:00, 956280.34it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2168/2168 [00:00<00:00, 895886.81it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2491/2491 [00:00<00:00, 973719.60it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2220/2220 [00:00<00:00, 764165.36it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2257/2257 [00:00<00:00, 751850.06it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 133/133 [00:00<00:00, 80952.32it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3107/3107 [00:00<00:00, 790374.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 378/378 [00:00<00:00, 208501.70it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 114/114 [00:00<00:00, 73358.49it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3394/3394 [00:00<00:00, 313881.50it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 28/28 [00:00<00:00, 30727.50it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 676/676 [00:00<00:00, 333022.02it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 394/394 [00:00<00:00, 649079.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 403/403 [00:00<00:00, 374541.22it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 126/126 [00:00<00:00, 107153.75it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 517/517 [00:00<00:00, 546485.68it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 456/456 [00:00<00:00, 207823.82it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6777/6777 [00:00<00:00, 582463.44it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5926/5926 [00:00<00:00, 675181.20it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9517/9517 [00:00<00:00, 935244.98it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9626/9626 [00:00<00:00, 600692.88it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                       | 0/431 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 77222 was translated into 3370833\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 431/431 [00:00<00:00, 52845.68it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 77/77 [00:00<00:00, 99098.31it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 91/91 [00:00<00:00, 124488.47it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 478/478 [00:00<00:00, 249331.84it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 142/142 [00:00<00:00, 319009.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 155/155 [00:00<00:00, 113696.59it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 145/145 [00:00<00:00, 221476.36it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 131/131 [00:00<00:00, 155961.91it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 233/233 [00:00<00:00, 311928.77it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4537/4537 [00:00<00:00, 742443.03it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4309/4309 [00:00<00:00, 947632.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4455/4455 [00:00<00:00, 861565.12it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3868/3868 [00:00<00:00, 731582.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3192/3192 [00:00<00:00, 724118.04it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4514/4514 [00:00<00:00, 987126.60it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/4347 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 381198 was translated into 8845\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 4347/4347 [00:00<00:00, 585787.61it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4006/4006 [00:00<00:00, 922194.39it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3811/3811 [00:00<00:00, 1125509.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5397/5397 [00:00<00:00, 1124635.27it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2937/2937 [00:00<00:00, 731469.08it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3459/3459 [00:00<00:00, 761539.95it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2541/2541 [00:00<00:00, 809980.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2487/2487 [00:00<00:00, 809375.70it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2779/2779 [00:00<00:00, 692488.76it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3403/3403 [00:00<00:00, 507311.77it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3235/3235 [00:00<00:00, 969737.95it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/3285 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1114980 was translated into 2144\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 3285/3285 [00:00<00:00, 382157.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3403/3403 [00:00<00:00, 718981.29it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/3374 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 9413 was translated into 3371112\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 3374/3374 [00:00<00:00, 402536.74it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5592/5592 [00:00<00:00, 402300.95it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7208/7208 [00:00<00:00, 727338.29it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8320/8320 [00:00<00:00, 914601.21it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9716/9716 [00:00<00:00, 567693.22it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 11219/11219 [00:00<00:00, 918468.50it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8566/8566 [00:00<00:00, 914185.59it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9212/9212 [00:00<00:00, 773346.31it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8105/8105 [00:00<00:00, 943776.62it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7388/7388 [00:00<00:00, 1174125.41it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9373/9373 [00:00<00:00, 1088375.50it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8029/8029 [00:00<00:00, 614123.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 169/169 [00:00<00:00, 124729.43it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1680/1680 [00:00<00:00, 675204.17it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1869/1869 [00:00<00:00, 748582.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 289/289 [00:00<00:00, 152318.91it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5500/5500 [00:00<00:00, 974471.85it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5047/5047 [00:00<00:00, 1074496.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5073/5073 [00:00<00:00, 1031546.24it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5034/5034 [00:00<00:00, 467312.79it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4193/4193 [00:00<00:00, 477886.92it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3020/3020 [00:00<00:00, 526685.99it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3660/3660 [00:00<00:00, 597181.69it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 12226/12226 [00:00<00:00, 965280.49it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7728/7728 [00:00<00:00, 1136441.39it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2891/2891 [00:00<00:00, 649929.40it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4874/4874 [00:00<00:00, 671893.70it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 11261/11261 [00:00<00:00, 982568.28it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 55587/55587 [00:00<00:00, 945361.26it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6298/6298 [00:00<00:00, 877161.77it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 26032/26032 [00:00<00:00, 889369.56it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4523/4523 [00:00<00:00, 948968.89it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3097/3097 [00:00<00:00, 348288.27it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1457/1457 [00:00<00:00, 630465.38it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 558/558 [00:00<00:00, 314361.54it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1501/1501 [00:00<00:00, 646437.04it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2198/2198 [00:00<00:00, 796120.91it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3438/3438 [00:00<00:00, 905438.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/1146 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 59452 was translated into 3371021\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 1146/1146 [00:00<00:00, 215816.83it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1587/1587 [00:00<00:00, 644621.39it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1909/1909 [00:00<00:00, 657329.15it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2613/2613 [00:00<00:00, 827398.18it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1692/1692 [00:00<00:00, 663248.82it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 528/528 [00:00<00:00, 118522.48it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2426/2426 [00:00<00:00, 729574.93it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2328/2328 [00:00<00:00, 831361.41it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1467/1467 [00:00<00:00, 743480.42it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1224/1224 [00:00<00:00, 679887.18it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8718/8718 [00:00<00:00, 466027.84it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4912/4912 [00:00<00:00, 597137.01it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3658/3658 [00:00<00:00, 745481.95it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/3782 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 302711 was translated into 225400\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 3782/3782 [00:00<00:00, 459992.97it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3770/3770 [00:00<00:00, 726211.36it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2955/2955 [00:00<00:00, 618471.47it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6203/6203 [00:00<00:00, 936310.78it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6290/6290 [00:00<00:00, 996192.73it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3757/3757 [00:00<00:00, 724139.52it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1977/1977 [00:00<00:00, 625632.94it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7791/7791 [00:00<00:00, 864058.34it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4575/4575 [00:00<00:00, 685074.64it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2896/2896 [00:00<00:00, 610571.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3607/3607 [00:00<00:00, 656064.81it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3467/3467 [00:00<00:00, 672290.89it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8549/8549 [00:00<00:00, 859800.14it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3325/3325 [00:00<00:00, 725224.17it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9634/9634 [00:00<00:00, 505484.49it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3711/3711 [00:00<00:00, 787386.79it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4006/4006 [00:00<00:00, 586695.83it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group:   0%|                                                                                                                      | 0/5059 [00:00<?, ?it/s]/Users/ifeanyi/anaconda3/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 686832 was translated into 76867\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "Updating taxon_group: 100%|| 5059/5059 [00:00<00:00, 256879.10it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 7676/7676 [00:00<00:00, 1242445.01it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 63/63 [00:00<00:00, 208721.29it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2687/2687 [00:00<00:00, 870950.14it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2082/2082 [00:00<00:00, 808269.25it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2252/2252 [00:00<00:00, 575809.11it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2072/2072 [00:00<00:00, 863018.66it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1555/1555 [00:00<00:00, 851121.33it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2132/2132 [00:00<00:00, 788836.99it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1865/1865 [00:00<00:00, 706437.01it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1675/1675 [00:00<00:00, 804104.29it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2726/2726 [00:00<00:00, 370381.36it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2366/2366 [00:00<00:00, 889142.84it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2964/2964 [00:00<00:00, 854368.57it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1429/1429 [00:00<00:00, 461121.74it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 338/338 [00:00<00:00, 209158.27it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 145/145 [00:00<00:00, 116263.44it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1658/1658 [00:00<00:00, 676474.32it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2162/2162 [00:00<00:00, 710052.87it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 12677/12677 [00:00<00:00, 1271916.37it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1721/1721 [00:00<00:00, 839934.51it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1852/1852 [00:00<00:00, 861754.05it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1799/1799 [00:00<00:00, 602679.94it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 1620/1620 [00:00<00:00, 864035.16it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 2537/2537 [00:00<00:00, 845593.55it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3499/3499 [00:00<00:00, 761789.24it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 6083/6083 [00:00<00:00, 1069050.16it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4215/4215 [00:00<00:00, 598557.40it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3831/3831 [00:00<00:00, 738436.52it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 4010/4010 [00:00<00:00, 623231.89it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 9866/9866 [00:00<00:00, 883171.56it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3320/3320 [00:00<00:00, 748539.98it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 3322/3322 [00:00<00:00, 724984.54it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8220/8220 [00:00<00:00, 599478.00it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 5529/5529 [00:00<00:00, 963412.69it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:373: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined.rename(columns={\n",
      "Updating taxon_group: 100%|| 8361/8361 [00:00<00:00, 841901.76it/s]\n",
      "/var/folders/xt/vz8fznds4hdbfcxyds95mtp00000gp/T/ipykernel_87864/2777863969.py:386: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       sample  \\\n",
      "0  AIAMAAGR001_LIV_S46_662105   \n",
      "1  AIAMAAGR001_LIV_S46_662105   \n",
      "2  AIAMAAGR001_LIV_S46_662105   \n",
      "3  AIAMAAGR001_LIV_S46_662105   \n",
      "4  AIAMAAGR001_LIV_S46_662105   \n",
      "\n",
      "                                         contig_name  contig_length  \\\n",
      "0  AIAMAAGR001_LIV_S46_662105_NODE_1000_length_48...            480   \n",
      "1  AIAMAAGR001_LIV_S46_662105_NODE_1001_length_48...            480   \n",
      "2  AIAMAAGR001_LIV_S46_662105_NODE_1002_length_48...            480   \n",
      "3  AIAMAAGR001_LIV_S46_662105_NODE_1003_length_48...            480   \n",
      "4  AIAMAAGR001_LIV_S46_662105_NODE_1004_length_47...            479   \n",
      "\n",
      "   read_count  \n",
      "0          12  \n",
      "1          12  \n",
      "2           8  \n",
      "3           4  \n",
      "4          13  \n",
      "                       sample  \\\n",
      "0  AIAMAAGR001_LIV_S46_662105   \n",
      "1  AIAMAAGR001_LIV_S46_662105   \n",
      "2  AIAMAAGR001_LIV_S46_662105   \n",
      "3  AIAMAAGR001_LIV_S46_662105   \n",
      "4  AIAMAAGR001_LIV_S46_662105   \n",
      "\n",
      "                                         contig_name  contig_length  \\\n",
      "0  AIAMAAGR001_LIV_S46_662105_NODE_1006_length_47...            479   \n",
      "1  AIAMAAGR001_LIV_S46_662105_NODE_1008_length_47...            479   \n",
      "2  AIAMAAGR001_LIV_S46_662105_NODE_1014_length_47...            478   \n",
      "3  AIAMAAGR001_LIV_S46_662105_NODE_1019_length_47...            477   \n",
      "4  AIAMAAGR001_LIV_S46_662105_NODE_1021_length_47...            477   \n",
      "\n",
      "   read_count    nr nt_or_nr  bitscore  align_length taxon_group  tax_id  \\\n",
      "0           6  True       nr      52.4            32     Metazoa   10029   \n",
      "1           8  True       nr     188.0           166     Metazoa  109678   \n",
      "2           8  True       nr      51.6            40     Metazoa   10089   \n",
      "3         183  True       nr     254.0           158     Metazoa   61156   \n",
      "4          13  True       nr     136.0           111     Viruses   11855   \n",
      "\n",
      "   identity                   sci_name  \\\n",
      "0      78.1         Cricetulus griseus   \n",
      "1      54.8       Phodopus roborovskii   \n",
      "2      62.5                 Mus caroli   \n",
      "3      81.0      Arvicanthis niloticus   \n",
      "4      63.1  Mason-Pfizer monkey virus   \n",
      "\n",
      "                                           com_names phylum family lca_taxid  \\\n",
      "0  XP_035304239.1 mitogen-activated protein kinas...    NaN    NaN       NaN   \n",
      "1         CAH7434577.1 Gm8113 [Phodopus roborovskii]    NaN    NaN       NaN   \n",
      "2  XP_021035475.1 centromere protein P isoform X2...    NaN    NaN       NaN   \n",
      "3  XP_034344067.1 LOW QUALITY PROTEIN: uncharacte...    NaN    NaN       NaN   \n",
      "4     2D4N_A Chain A, DU [Mason-Pfizer monkey virus]    NaN    NaN       NaN   \n",
      "\n",
      "  lca_taxid lca_name  common_nr  \n",
      "0       NaN      NaN       True  \n",
      "1       NaN      NaN       True  \n",
      "2       NaN      NaN       True  \n",
      "3       NaN      NaN       True  \n",
      "4       NaN      NaN       True  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "from ete3 import NCBITaxa\n",
    "from tqdm import tqdm  # For progress bars\n",
    "import logging\n",
    "import glob\n",
    "\n",
    "# ===========================\n",
    "# Configuration\n",
    "# ===========================\n",
    "\n",
    "# Directories\n",
    "json_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_Contig_quality/\"\n",
    "lca_dir = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_SENZOR_Blast_NR/LCA\"\n",
    "\n",
    "# Output TSV file paths\n",
    "contig_stats_all_tsv = os.path.join(json_dir, \"contig_stats.tsv\")\n",
    "complete_summary_tsv = os.path.join(lca_dir, \"contig_stats_lca.tsv\")\n",
    "\n",
    "# ===========================\n",
    "# Logging Configuration\n",
    "# ===========================\n",
    "\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s %(levelname)s:%(message)s',\n",
    "    handlers=[\n",
    "        logging.StreamHandler()\n",
    "    ]\n",
    ")\n",
    "\n",
    "# ===========================\n",
    "# Function Definitions\n",
    "# ===========================\n",
    "\n",
    "def load_json_data(json_path):\n",
    "    \"\"\"\n",
    "    Load JSON data from the given path and process contig names.\n",
    "\n",
    "    Args:\n",
    "        json_path (str): Path to the JSON file.\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary with contig_name as keys and read_count as values.\n",
    "    \"\"\"\n",
    "    with open(json_path, 'r') as f:\n",
    "        data = json.load(f)\n",
    "    \n",
    "    # Replace \"~\" with \"_\" in contig names and exclude any key named \"*\"\n",
    "    processed_data = {key.replace(\"~\", \"_\"): value for key, value in data.items() if key != \"*\"}\n",
    "    return processed_data\n",
    "\n",
    "def load_lca_data(lca_path):\n",
    "    \"\"\"\n",
    "    Load LCA data from the given path with specified data types.\n",
    "\n",
    "    Args:\n",
    "        lca_path (str): Path to the LCA file.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame containing LCA data.\n",
    "    \"\"\"\n",
    "    dtype_dict = {\n",
    "        \"query\": str,\n",
    "        \"subject\": str,\n",
    "        \"identity\": float,\n",
    "        \"alignment_length\": int,\n",
    "        \"mismatches\": int,\n",
    "        \"gap_opens\": int,\n",
    "        \"q_start\": int,\n",
    "        \"q_end\": int,\n",
    "        \"s_start\": int,\n",
    "        \"s_end\": int,\n",
    "        \"evalue\": float,\n",
    "        \"bit_score\": float,\n",
    "        \"tax_id\": str,\n",
    "        \"sci_name\": str,\n",
    "        \"com_names\": str,\n",
    "        \"superkingdom\": str,\n",
    "        \"phylum\": str,\n",
    "        \"family\": str,\n",
    "        \"lca_taxid\": str,\n",
    "        \"lca_name\": str\n",
    "    }\n",
    "    return pd.read_csv(\n",
    "        lca_path, \n",
    "        sep=\"\\t\", \n",
    "        header=0,  # Use the first row as header\n",
    "        dtype=dtype_dict,\n",
    "        na_values=['Unknown'],  # Treat 'Unknown' as NaN if desired\n",
    "        keep_default_na=False  # Prevent 'Unknown' from being treated as NaN\n",
    "    )\n",
    "\n",
    "def extract_contig_length(contig_name):\n",
    "    \"\"\"\n",
    "    Extract contig length from contig name using regex.\n",
    "    Assumes contig_name contains 'length_<number>'.\n",
    "\n",
    "    Args:\n",
    "        contig_name (str): The contig name string.\n",
    "\n",
    "    Returns:\n",
    "        int or float: Extracted contig length or NaN if not found.\n",
    "    \"\"\"\n",
    "    match = re.search(r\"length_(\\d+)\", contig_name)\n",
    "    if match:\n",
    "        return int(match.group(1))\n",
    "    else:\n",
    "        return pd.NA  # Use pandas NA for missing values\n",
    "\n",
    "def get_tax_group(taxid, ncbi, taxon_groups, taxon_groups_id):\n",
    "    \"\"\"\n",
    "    Determine the taxon group for a given taxid.\n",
    "\n",
    "    Args:\n",
    "        taxid (int or str): NCBI taxid.\n",
    "        ncbi (NCBITaxa): Initialized NCBITaxa object.\n",
    "        taxon_groups (list): List of taxon group names.\n",
    "        taxon_groups_id (list): List of taxon group taxids.\n",
    "\n",
    "    Returns:\n",
    "        str: Taxon group name or \"Ambiguous\" if not found.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        taxid = int(taxid)\n",
    "        lineage = ncbi.get_lineage(taxid)\n",
    "        for i, tax in enumerate(taxon_groups_id):\n",
    "            if tax is not None and tax in lineage:\n",
    "                return taxon_groups[i]\n",
    "        return \"Ambiguous\"\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error processing taxid {taxid}: {e}\")\n",
    "        return \"Ambiguous\"\n",
    "\n",
    "# ===========================\n",
    "# Initialize NCBITaxa\n",
    "# ===========================\n",
    "\n",
    "logging.info(\"Initializing NCBITaxa...\")\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "# Define taxon groups and retrieve their taxids\n",
    "taxon_groups = [\"Viruses\", \"Bacteria\", \"Archaea\", \"Metazoa\", \"Eukaryota\"]\n",
    "taxon_groups_id = []\n",
    "\n",
    "for group in taxon_groups:\n",
    "    name_translator = ncbi.get_name_translator([group])\n",
    "    if group in name_translator:\n",
    "        taxid = name_translator[group][0]\n",
    "        taxon_groups_id.append(taxid)\n",
    "    else:\n",
    "        logging.warning(f\"Taxon group '{group}' not found in NCBITaxa database.\")\n",
    "        taxon_groups_id.append(None)  # Append None if taxon group not found\n",
    "\n",
    "# ===========================\n",
    "# Initialize Data Containers\n",
    "# ===========================\n",
    "\n",
    "# List to hold DataFrames for contig_stats_all.tsv\n",
    "contig_stats_all_data = []\n",
    "\n",
    "# List to hold DataFrames for contig_stats_lca.tsv\n",
    "complete_summary_data = []\n",
    "\n",
    "# Initialize a cache for taxid to taxon_group mapping to optimize performance\n",
    "taxid_cache = {}\n",
    "\n",
    "def get_tax_group_cached(taxid):\n",
    "    \"\"\"\n",
    "    Determine the taxon group for a given taxid with caching.\n",
    "\n",
    "    Args:\n",
    "        taxid (int or str): NCBI taxid.\n",
    "\n",
    "    Returns:\n",
    "        str: Taxon group name or \"Ambiguous\" if not found.\n",
    "    \"\"\"\n",
    "    if taxid in taxid_cache:\n",
    "        return taxid_cache[taxid]\n",
    "    else:\n",
    "        group = get_tax_group(taxid, ncbi, taxon_groups, taxon_groups_id)\n",
    "        taxid_cache[taxid] = group\n",
    "        return group\n",
    "\n",
    "# ===========================\n",
    "# Dynamic File Listing\n",
    "# ===========================\n",
    "\n",
    "# Dynamically list all JSON files in json_dir\n",
    "json_pattern = os.path.join(json_dir, \"*_contigs_stats.json\")\n",
    "json_files = glob.glob(json_pattern)\n",
    "\n",
    "if not json_files:\n",
    "    logging.error(f\"No JSON files found in directory: {json_dir}\")\n",
    "    exit(1)\n",
    "\n",
    "# Dynamically list all NR BLAST result files in lca_dir\n",
    "# Assuming NR blast files have 'blast_nr' in their filename\n",
    "nr_lca_pattern = os.path.join(lca_dir, \"*_modified_diamond_blast_nr_annotated.m9\")\n",
    "nr_lca_files = glob.glob(nr_lca_pattern)\n",
    "\n",
    "if not nr_lca_files:\n",
    "    logging.error(f\"No NR BLAST result files found in directory: {lca_dir}\")\n",
    "    exit(1)\n",
    "\n",
    "# Create a mapping from sample name to NR BLAST file\n",
    "# Assuming sample name can be extracted from JSON and NR BLAST filenames\n",
    "# For example:\n",
    "# JSON: AIAMACAT001_S69_657769_contig_stats.json\n",
    "# NR BLAST: AIAMACAT001_S69_657769_modified_diamond_blast_nr_annotated.m9\n",
    "\n",
    "nr_blast_mapping = {}\n",
    "for nr_file in nr_lca_files:\n",
    "    basename = os.path.basename(nr_file)\n",
    "    # Extract sample name by removing suffix '_modified_diamond_blast_nr_annotated.m9'\n",
    "    match = re.match(r\"(.+?)_modified_diamond_blast_nr_annotated\\.m9\", basename)\n",
    "    if match:\n",
    "        sample_name = match.group(1)  # \"LCA_AIAMA_GOT005_OS_S43_673459\"\n",
    "        # Remove the \"LCA_\" prefix if present\n",
    "        sample_name = re.sub(r\"^LCA_\", \"\", sample_name)\n",
    "        nr_blast_mapping[sample_name] = nr_file\n",
    "    else:\n",
    "        logging.warning(f\"NR BLAST file name does not match expected pattern: {basename}\")\n",
    "\n",
    "# Log the mapping for verification\n",
    "logging.info(\"Mapping of samples to NR BLAST files:\")\n",
    "for sample, nr_file in nr_blast_mapping.items():\n",
    "    logging.info(f\"Sample: {sample} => NR BLAST File: {os.path.basename(nr_file)}\")\n",
    "\n",
    "# ===========================\n",
    "# Processing JSON and NR BLAST Files\n",
    "# ===========================\n",
    "\n",
    "for json_file in json_files:\n",
    "    basename = os.path.basename(json_file)\n",
    "    # Extract sample name by removing suffix '_contig_stats.json'\n",
    "    sample_name = basename.replace(\"_contigs_stats.json\", \"\")\n",
    "    \n",
    "    logging.info(f\"\\nProcessing sample: {sample_name}\")\n",
    "    \n",
    "    # Find corresponding NR BLAST file\n",
    "    if sample_name not in nr_blast_mapping:\n",
    "        logging.error(f\"No corresponding NR BLAST file found for sample: {sample_name}\")\n",
    "        continue\n",
    "    \n",
    "    nr_lca_file = nr_blast_mapping[sample_name]\n",
    "    \n",
    "    # ---------------------------\n",
    "    # Process JSON File\n",
    "    # ---------------------------\n",
    "    \n",
    "    # Full path to the JSON file\n",
    "    json_path = os.path.join(json_dir, json_file)\n",
    "    \n",
    "    # Check if JSON file exists\n",
    "    if not os.path.exists(json_path):\n",
    "        logging.error(f\"JSON file not found: {json_path}\")\n",
    "        continue\n",
    "    \n",
    "    # Load and process JSON data\n",
    "    json_data = load_json_data(json_path)\n",
    "    num_contigs = len(json_data)\n",
    "    logging.info(f\"Loaded {num_contigs} contigs from {basename}\")\n",
    "    \n",
    "    # Convert to DataFrame for contig_stats_all.tsv\n",
    "    contig_stats = pd.DataFrame.from_dict(json_data, orient='index', columns=['read_count']).reset_index()\n",
    "    contig_stats.rename(columns={'index': 'contig_name'}, inplace=True)\n",
    "    \n",
    "    # Extract contig_length\n",
    "    contig_stats['contig_length'] = contig_stats['contig_name'].apply(extract_contig_length)\n",
    "    \n",
    "    # Handle contig_length extraction failures\n",
    "    missing_lengths = contig_stats['contig_length'].isna().sum()\n",
    "    if missing_lengths > 0:\n",
    "        logging.warning(f\"{missing_lengths} contigs in {basename} did not have a 'length_' pattern.\")\n",
    "    \n",
    "    # Add sample name\n",
    "    contig_stats['sample'] = sample_name\n",
    "    \n",
    "    # Reorder columns for contig_stats_all.tsv\n",
    "    contig_stats = contig_stats[['sample', 'contig_name', 'contig_length', 'read_count']]\n",
    "    \n",
    "    # Append to contig_stats_all_data list\n",
    "    contig_stats_all_data.append(contig_stats)\n",
    "    \n",
    "    # ---------------------------\n",
    "    # Process NR LCA File\n",
    "    # ---------------------------\n",
    "    \n",
    "    # Full path to NR LCA file\n",
    "    nr_lca_path = nr_lca_file\n",
    "    \n",
    "    # Check if NR LCA file exists\n",
    "    if not os.path.exists(nr_lca_path):\n",
    "        logging.error(f\"NR LCA file not found: {nr_lca_path}\")\n",
    "        continue\n",
    "    \n",
    "    # Load LCA data\n",
    "    nr_lca = load_lca_data(nr_lca_path)\n",
    "    logging.info(f\"Loaded NR LCA data for {sample_name}\")\n",
    "    \n",
    "    # Standardize contig names in LCA files by replacing \"~\" with \"_\"\n",
    "    nr_lca[\"query\"] = nr_lca[\"query\"].str.replace(\"~\", \"_\")\n",
    "    \n",
    "    # Check for matching contigs\n",
    "    contig_names = set(contig_stats[\"contig_name\"])\n",
    "    nr_queries = set(nr_lca[\"query\"])\n",
    "    common_nr = contig_names.intersection(nr_queries)\n",
    "    logging.info(f\"Common contigs in NR LCA for {sample_name}: {len(common_nr)}\")\n",
    "    \n",
    "    # Merge NR LCA data with contig stats\n",
    "    nr_merged = pd.merge(\n",
    "        contig_stats, \n",
    "        nr_lca, \n",
    "        left_on=\"contig_name\", \n",
    "        right_on=\"query\", \n",
    "        how=\"inner\",\n",
    "        suffixes=('_stats', '_nr')\n",
    "    )\n",
    "    nr_merged[\"nr\"] = True\n",
    "    nr_merged[\"nt_or_nr\"] = \"nr\"\n",
    "    logging.info(f\"NR merged rows for {sample_name}: {len(nr_merged)}\")\n",
    "    \n",
    "    # Add 'common_nr' column\n",
    "    # This column will be True if the contig is present in NR, else False\n",
    "    nr_merged[\"common_nr\"] = nr_merged[\"contig_name\"].isin(common_nr)\n",
    "    logging.info(f\"Added 'common_nr' column for {sample_name}\")\n",
    "    \n",
    "    # Handle duplicate 'contig_length' columns\n",
    "    # After merging, there might be 'contig_length_stats' and 'contig_length_nr'\n",
    "    # We'll retain one 'contig_length' column and drop the others\n",
    "    if 'contig_length_stats' in nr_merged.columns:\n",
    "        nr_merged.rename(columns={'contig_length_stats': 'contig_length'}, inplace=True)\n",
    "        # Drop the other contig_length columns\n",
    "        nr_merged.drop(['contig_length_nr'], axis=1, inplace=True, errors='ignore')\n",
    "    elif 'contig_length_nr' in nr_merged.columns:\n",
    "        nr_merged.rename(columns={'contig_length_nr': 'contig_length'}, inplace=True)\n",
    "        nr_merged.drop(['contig_length_stats'], axis=1, inplace=True, errors='ignore')\n",
    "    \n",
    "    # Define selected_columns with the updated 'contig_length'\n",
    "    selected_columns = [\n",
    "        \"sample\", \n",
    "        \"contig_name\", \n",
    "        \"contig_length\",  # Updated column\n",
    "        \"read_count\", \n",
    "        \"nr\", \n",
    "        \"nt_or_nr\",\n",
    "        \"lca_taxid\", \n",
    "        \"bit_score\", \n",
    "        \"alignment_length\", \n",
    "        \"superkingdom\", \n",
    "        \"tax_id\",\n",
    "        \"identity\",\n",
    "        \"sci_name\",\n",
    "        \"com_names\",\n",
    "        \"phylum\",\n",
    "        \"family\",\n",
    "        \"lca_taxid\",\n",
    "        \"lca_name\",\n",
    "        \"common_nr\"  # Include the new column\n",
    "    ]\n",
    "    \n",
    "    # Select only columns that are present\n",
    "    available_columns = nr_merged.columns.tolist()\n",
    "    selected_columns = [col for col in selected_columns if col in available_columns]\n",
    "    \n",
    "    # Now, select the relevant columns\n",
    "    combined = nr_merged[selected_columns]\n",
    "    \n",
    "    # Rename columns for clarity\n",
    "    combined.rename(columns={\n",
    "        \"bit_score\": \"bitscore\",\n",
    "        \"alignment_length\": \"align_length\"\n",
    "    }, inplace=True)\n",
    "    \n",
    "    # ===========================\n",
    "    # Update taxon_group Using taxid\n",
    "    # ===========================\n",
    "    \n",
    "    logging.info(\"\\nUpdating 'taxon_group' based on 'taxid' using ETE3's NCBITaxa...\")\n",
    "    \n",
    "    # Apply the cached get_tax_group function with progress bar\n",
    "    tqdm.pandas(desc=\"Updating taxon_group\")\n",
    "    combined['taxon_group'] = combined['tax_id'].progress_apply(get_tax_group_cached)\n",
    "    \n",
    "    logging.info(\"Updated 'taxon_group' successfully.\")\n",
    "    \n",
    "    # ---------------------------\n",
    "    # Reorder columns after updating taxon_group\n",
    "    # ---------------------------\n",
    "    \n",
    "    # Reorder columns if desired\n",
    "    # Ensure that 'common_nr' is at the end\n",
    "    desired_order = [\n",
    "        \"sample\", \n",
    "        \"contig_name\", \n",
    "        \"contig_length\", \n",
    "        \"read_count\", \n",
    "        \"nr\", \n",
    "        \"nt_or_nr\",\n",
    "        \"taxid\", \n",
    "        \"bitscore\", \n",
    "        \"align_length\", \n",
    "        \"taxon_group\", \n",
    "        \"tax_id\",\n",
    "        \"identity\",\n",
    "        \"sci_name\",\n",
    "        \"com_names\",\n",
    "        \"phylum\",\n",
    "        \"family\",\n",
    "        \"lca_taxid\",\n",
    "        \"lca_name\",\n",
    "        \"common_nr\"  # Ensure the new column is included\n",
    "    ]\n",
    "    \n",
    "    # Adjust desired_order based on available columns\n",
    "    final_columns = [col for col in desired_order if col in combined.columns]\n",
    "    combined = combined[final_columns]\n",
    "    \n",
    "    logging.info(f\"Final selected columns after updating 'taxon_group': {combined.columns.tolist()}\")\n",
    "    \n",
    "    logging.info(f\"Total merged rows for {sample_name}: {len(combined)}\")\n",
    "    \n",
    "    # Append to complete_summary_data list\n",
    "    complete_summary_data.append(combined)\n",
    "\n",
    "# ===========================\n",
    "# Combine All Data and Save Outputs\n",
    "# ===========================\n",
    "\n",
    "# ---------------------------\n",
    "# Save contig_stats_all.tsv\n",
    "# ---------------------------\n",
    "\n",
    "if contig_stats_all_data:\n",
    "    # Concatenate all contig_stats DataFrames\n",
    "    contig_stats_all_df = pd.concat(contig_stats_all_data, ignore_index=True)\n",
    "    \n",
    "    # Optional: Sort the DataFrame for better readability\n",
    "    contig_stats_all_df.sort_values(by=['sample', 'contig_name'], inplace=True)\n",
    "    \n",
    "    # Optional: Reset index after sorting\n",
    "    contig_stats_all_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    # Save to TSV\n",
    "    contig_stats_all_df.to_csv(contig_stats_all_tsv, sep='\\t', index=False)\n",
    "    logging.info(f\"\\nSuccessfully saved combined contig stats to {contig_stats_all_tsv}\")\n",
    "else:\n",
    "    logging.warning(\"No contig stats data available to save.\")\n",
    "\n",
    "# ---------------------------\n",
    "# Save contig_stats_lca.tsv\n",
    "# ---------------------------\n",
    "\n",
    "if complete_summary_data:\n",
    "    # Concatenate all complete_summary DataFrames\n",
    "    complete_summary_df = pd.concat(complete_summary_data, ignore_index=True)\n",
    "    \n",
    "    # Optional: Sort the DataFrame for better readability\n",
    "    complete_summary_df.sort_values(by=['sample', 'contig_name'], inplace=True)\n",
    "    \n",
    "    # Optional: Reset index after sorting\n",
    "    complete_summary_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    # Save to TSV\n",
    "    complete_summary_df.to_csv(complete_summary_tsv, sep='\\t', index=False)\n",
    "    logging.info(f\"Successfully saved complete summary to {complete_summary_tsv}\")\n",
    "else:\n",
    "    logging.warning(\"No data available to save for contig_stats_lca.tsv.\")\n",
    "\n",
    "# ===========================\n",
    "# Display Previews\n",
    "# ===========================\n",
    "\n",
    "# Preview contig_stats_all.tsv\n",
    "if contig_stats_all_data:\n",
    "    logging.info(\"\\nPreview of contig_stats_all.tsv:\")\n",
    "    print(contig_stats_all_df.head())\n",
    "else:\n",
    "    logging.warning(\"No contig stats data available to preview.\")\n",
    "\n",
    "# Preview contig_stats_lca.tsv\n",
    "if complete_summary_data:\n",
    "    logging.info(\"\\nPreview of contig_stats_lca.tsv:\")\n",
    "    print(complete_summary_df.head())\n",
    "else:\n",
    "    logging.warning(\"No data available to preview for contig_stats_lca.tsv.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e40460f6-ed39-4093-b8f5-3f874ac330a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed AIAMA_GOT005_OS_S43_673459_contigs_stats.json\n",
      "Processed AIAMA_GOT006_RS_S44_673460_contigs_stats.json\n",
      "Processed AIAMA_GOT007_OS_S45_673461_contigs_stats.json\n",
      "Processed AIAMA_LIZ002_OS_S50_673462_contigs_stats.json\n",
      "Processed AIAMA_PIG003_RS_S51_673463_contigs_stats.json\n",
      "Processed AIAMA_SHP004_OS_S46_673464_contigs_stats.json\n",
      "Processed AIAMA_SHP005_RS_S47_673465_contigs_stats.json\n",
      "Processed AIAMA_SHP006_OS_S48_673466_contigs_stats.json\n",
      "Processed AIAMAAGR001_LIV_S46_662105_contigs_stats.json\n",
      "Processed AIAMACAT001_S69_657769_contigs_stats.json\n",
      "Processed AIAMADOG001_S40_657732_contigs_stats.json\n",
      "Processed AIAMADOG002_OS_S64_673467_contigs_stats.json\n",
      "Processed AIAMAGOAT001_S48_656109_contigs_stats.json\n",
      "Processed AIAMAGOAT002_S49_656110_contigs_stats.json\n",
      "Processed AIAMAGOAT003_S50_657720_contigs_stats.json\n",
      "Processed AIAMAGOAT004_S51_656112_contigs_stats.json\n",
      "Processed AIAMAPIG001_S63_657713_contigs_stats.json\n",
      "Processed AIAMAPIG002_S64_657714_contigs_stats.json\n",
      "Processed AIAMAR001_LIV_S21_662106_contigs_stats.json\n",
      "Processed AIAMAR002_LIV_S22_662107_contigs_stats.json\n",
      "Processed AIAMAR003_LIV_S23_662108_contigs_stats.json\n",
      "Processed AIAMAR004_LIV_S24_1_662161_contigs_stats.json\n",
      "Processed AIAMAR005_LIV_S25_1_662162_contigs_stats.json\n",
      "Processed AIAMAR006_LIV_S26_2_662165_contigs_stats.json\n",
      "Processed AIAMAR007_LIV_S27_2_662166_contigs_stats.json\n",
      "Processed AIAMAR008_LIV_S28_1_662167_contigs_stats.json\n",
      "Processed AIAMAR009_LIV_S29_1_662168_contigs_stats.json\n",
      "Processed AIAMAR010_LIV_S30_663038_contigs_stats.json\n",
      "Processed AIAMAR011_LIV_S31_663039_contigs_stats.json\n",
      "Processed AIAMAR012_LIV_S32_663040_contigs_stats.json\n",
      "Processed AIAMAR013_LIV_S33_663041_contigs_stats.json\n",
      "Processed AIAMAR014_LIV_S34_663042_contigs_stats.json\n",
      "Processed AIAMAR015_LIV_S35_663046_contigs_stats.json\n",
      "Processed AIAMAR016_LIV_S36_663047_contigs_stats.json\n",
      "Processed AIAMAR017_LIV_S37_663048_contigs_stats.json\n",
      "Processed AIAMAR018_LIV_S38_663049_contigs_stats.json\n",
      "Processed AIAMAR019_LIV_S39_663050_contigs_stats.json\n",
      "Processed AIAMAR020_LIV_S40_663051_contigs_stats.json\n",
      "Processed AIAMASHEEP001_S53_656078_contigs_stats.json\n",
      "Processed AIAMASHEEP002_S54_656079_contigs_stats.json\n",
      "Processed AIMASHEEP003_S55_656080_contigs_stats.json\n",
      "Processed AIMGBR021_LIV_S41_663052_contigs_stats.json\n",
      "Processed AIMGBR022_LIV_S42_663077_contigs_stats.json\n",
      "Processed AIMGBR023_LIV_S43_662128_contigs_stats.json\n",
      "Processed AIMGBR024_LIV_S44_662129_contigs_stats.json\n",
      "Processed AIMGL_GOT003_OS_S65_673468_contigs_stats.json\n",
      "Processed AIMGL_GOT004_RS_S66_673469_contigs_stats.json\n",
      "Processed AIMGLDOG001_S59_657733_contigs_stats.json\n",
      "Processed AIMGLGOAT001_S60_656113_contigs_stats.json\n",
      "Processed AIMGLGOAT002_S61_656114_contigs_stats.json\n",
      "Processed AIMUZSHEEP001_S52_656081_contigs_stats.json\n",
      "Processed AINWZ_CHK003_CS_S49_673471_contigs_stats.json\n",
      "Processed AINWZ_DOG004_RS_S40_673472_contigs_stats.json\n",
      "Processed AINWZ_DOG005_OS_S41_673473_contigs_stats.json\n",
      "Processed AINWZ_DOG006_RS_S42_674096_contigs_stats.json\n",
      "Processed AINWZ_GOT008_OS_S67_673475_contigs_stats.json\n",
      "Processed AINWZ_GOT009_RS_S68_673476_contigs_stats.json\n",
      "Processed AINWZ_GOT010_OS_S69_673481_contigs_stats.json\n",
      "Processed AINWZ_GOT011_RS_S37_673482_contigs_stats.json\n",
      "Processed AINWZ_GOT012_RS_S38_673479_contigs_stats.json\n",
      "Processed AINWZ_GOT013_OS_S39_673480_contigs_stats.json\n",
      "Processed AINWZAGR001_LIV_S45_662130_contigs_stats.json\n",
      "Processed AINWZCHICKEN001_S56_655747_contigs_stats.json\n",
      "Processed AINWZCHICKEN002_S57_655748_contigs_stats.json\n",
      "Processed AINWZDOG001_S37_657734_contigs_stats.json\n",
      "Processed AINWZDOG002_S38_657735_contigs_stats.json\n",
      "Processed AINWZDOG003_S39_657736_contigs_stats.json\n",
      "Processed AINWZGOAT001_S41_656115_contigs_stats.json\n",
      "Processed AINWZGOAT002_S42_656116_contigs_stats.json\n",
      "Processed AINWZGOAT003_S43_1_656182_contigs_stats.json\n",
      "Processed AINWZGOAT004_S44_1_656183_contigs_stats.json\n",
      "Processed AINWZGOAT005_S45_1_656184_contigs_stats.json\n",
      "Processed AINWZGOAT006_S46_1_656185_contigs_stats.json\n",
      "Processed AINWZGOAT007_S47_2_656739_contigs_stats.json\n",
      "Processed AINWZLIZARD001_S58_657718_contigs_stats.json\n",
      "Processed AINWZPIG001_S62_657715_contigs_stats.json\n",
      "Processed AINWZR001_LIV_S1_662131_contigs_stats.json\n",
      "Processed AINWZR002_LIV_S2_662132_contigs_stats.json\n",
      "Processed AINWZR003_LIV_S3_662133_contigs_stats.json\n",
      "Processed AINWZR004_LIV_S4_662134_contigs_stats.json\n",
      "Processed AINWZR005_LIV_S5_663064_contigs_stats.json\n",
      "Processed AINWZR006_LIV_S6_663065_contigs_stats.json\n",
      "Processed AINWZR007_LIV_S7_663066_contigs_stats.json\n",
      "Processed AINWZR008_LIV_S8_663067_contigs_stats.json\n",
      "Processed AINWZR009_LIV_S9_663068_contigs_stats.json\n",
      "Processed AINWZR010_LIV_S10_663069_contigs_stats.json\n",
      "Processed AINWZR011_LIV_S11_663070_contigs_stats.json\n",
      "Processed AINWZR012_LIV_S12_663071_contigs_stats.json\n",
      "Processed AINWZR013_LIV_S13_663072_contigs_stats.json\n",
      "Processed AINWZR014_LIV_S14_663073_contigs_stats.json\n",
      "Processed AINWZR015_LIV_S15_663078_contigs_stats.json\n",
      "Processed AINWZR016_LIV_S16_663079_contigs_stats.json\n",
      "Processed AINWZR017_LIV_S17_663080_contigs_stats.json\n",
      "Processed AINWZR018_LIV_S18_663081_contigs_stats.json\n",
      "Processed AINWZR019_LIV_S19_663082_contigs_stats.json\n",
      "Processed AINWZR020_LIV_S20_663083_contigs_stats.json\n",
      "Processed NE_S70_670946_contigs_stats.json\n",
      "Processed NE_S82_663084_contigs_stats.json\n",
      "Processed NE_S82_673483_contigs_stats.json\n",
      "Processed NTC_S71_670947_contigs_stats.json\n",
      "Processed OD_ZAD_AGR001_LIV_S31_673484_contigs_stats.json\n",
      "Processed OD_ZAD_AGR002_LIV_S32_673485_contigs_stats.json\n",
      "Processed OD_ZAD_AGR003_LIV_S33_673487_contigs_stats.json\n",
      "Processed OD_ZAD_AGR004_LIV_S34_673490_contigs_stats.json\n",
      "Processed OD_ZAD_AGR005_LIV_S35_673488_contigs_stats.json\n",
      "Processed OD_ZAD_AGR006_LUNG_S36_673489_contigs_stats.json\n",
      "Processed OD_ZAD_AGR007_LUNG_S52_673492_contigs_stats.json\n",
      "Processed OD_ZAD_AGR008_LUNG_S53_673493_contigs_stats.json\n",
      "Processed OD_ZAD_AGR009_LUNG_S54_673491_contigs_stats.json\n",
      "Processed OD_ZAD_AGR010_LUNG_S55_674094_contigs_stats.json\n",
      "Processed OD_ZAD_BAT020_LIV_S21_673503_contigs_stats.json\n",
      "Processed OD_ZAD_BAT021_LIV_S22_674095_contigs_stats.json\n",
      "Processed OD_ZAD_BAT023_LIV_S24_673504_contigs_stats.json\n",
      "Processed OD_ZAD_BAT024_LIV_S25_673505_contigs_stats.json\n",
      "Processed OD_ZAD_BAT025_LUNG_S26_673506_contigs_stats.json\n",
      "Processed OD_ZAD_BAT026_LUNG_S27_673507_contigs_stats.json\n",
      "Processed OD_ZAD_BAT027_LUNG_S28_673508_contigs_stats.json\n",
      "Processed OD_ZAD_BAT028_LUNG_S29_673509_contigs_stats.json\n",
      "Processed OD_ZAD_BAT029_LUNG_S30_673510_contigs_stats.json\n",
      "Processed OD_ZAD_CEG031_LIV_S1_673511_contigs_stats.json\n",
      "Processed OD_ZAD_CEG032_LIV_S2_673512_contigs_stats.json\n",
      "Processed OD_ZAD_CEG033_LIV_S3_673513_contigs_stats.json\n",
      "Processed OD_ZAD_CEG034_LIV_S4_673514_contigs_stats.json\n",
      "Processed OD_ZAD_CEG035_LIV_S5_673515_contigs_stats.json\n",
      "Processed OD_ZAD_CEG036_LUNG_S6_673516_contigs_stats.json\n",
      "Processed OD_ZAD_CEG037_LUNG_S7_673517_contigs_stats.json\n",
      "Processed OD_ZAD_CEG038_LUNG_S8_673518_contigs_stats.json\n",
      "Processed OD_ZAD_CEG039_LUNG_S9_673615_contigs_stats.json\n",
      "Processed OD_ZAD_CEG040_LUNG_S10_673616_contigs_stats.json\n",
      "Processed OD_ZAD_SQL021_LIV_S11_673617_contigs_stats.json\n",
      "Processed OD_ZAD_SQL022_LIV_S12_673618_contigs_stats.json\n",
      "Processed OD_ZAD_SQL023_LIV_S13_673619_contigs_stats.json\n",
      "Processed OD_ZAD_SQL024_LIV_S14_673620_contigs_stats.json\n",
      "Processed OD_ZAD_SQL025_LIV_S15_673621_contigs_stats.json\n",
      "Processed OD_ZAD_SQL026_LUNG_S16_673622_contigs_stats.json\n",
      "Processed OD_ZAD_SQL027_LUNG_S17_673623_contigs_stats.json\n",
      "Processed OD_ZAD_SQL028_LUNG_S18_673624_contigs_stats.json\n",
      "Processed OD_ZAD_SQL029_LUNG_S19_673625_contigs_stats.json\n",
      "Processed OD_ZAD_SQL030_LUNG_S20_673626_contigs_stats.json\n",
      "Processed SENONOWGOT003_RS_S79_673627_contigs_stats.json\n",
      "Processed SENONOWGOT014_OS_S75_673628_contigs_stats.json\n",
      "Processed SENONOWIFGOT005_RS_S78_673657_contigs_stats.json\n",
      "Processed SENONOWISUGOT001_OS_S80_673658_contigs_stats.json\n",
      "Processed SENONOWISUGOT001_RS_S74_673659_contigs_stats.json\n",
      "Processed SENONOWISUGOT002_OS_S76_673660_contigs_stats.json\n",
      "Processed SENONOWISUGOT004_RS_S77_673661_contigs_stats.json\n",
      "Processed SENONOWISUGOT005_RS_S71_673662_contigs_stats.json\n",
      "Processed SENONOWISUGOT006_OS_S73_673663_contigs_stats.json\n",
      "Processed SENONOWOGGOT016_RS_S81_673664_contigs_stats.json\n",
      "Processed SENONOWOGGOT017_OS_S72_673665_contigs_stats.json\n",
      "Processed SEZ_ON_OWIS_R001_S79_687907_contigs_stats.json\n",
      "Processed SEZ_ON_OWIS_R002_S80_687908_contigs_stats.json\n",
      "Processed SEZ_ON_OWIS_R003_S81_687909_contigs_stats.json\n",
      "Processed SEZ_ON_OWIS_R004_S82_687910_contigs_stats.json\n",
      "Processed SEZ_ON_OWOG_R001_S75_687911_contigs_stats.json\n",
      "Processed SEZ_ON_OWOG_R002_S76_687912_contigs_stats.json\n",
      "Processed SEZ_ON_OWOG_R003_S77_687913_contigs_stats.json\n",
      "Processed SEZ_ON_OWOG_R004_S78_687914_contigs_stats.json\n",
      "Processed SEZONOWDOG009_S20_657738_contigs_stats.json\n",
      "Processed SEZONOWDOG012_S10_657739_contigs_stats.json\n",
      "Processed SEZONOWDOG01_S7_657737_contigs_stats.json\n",
      "Processed SEZONOWGOAT009_S67_2_656740_contigs_stats.json\n",
      "Processed SEZONOWGOAT010_S65_2_656741_contigs_stats.json\n",
      "Processed SEZONOWIFCAT001_S34_657717_contigs_stats.json\n",
      "Processed SEZONOWIFDOG009_S1_657740_contigs_stats.json\n",
      "Processed SEZONOWIFGOAT007_S68_657731_contigs_stats.json\n",
      "Processed SEZONOWIFGOAT011_S29_2_656747_contigs_stats.json\n",
      "Processed SEZONOWIFGOAT012_S30_657726_contigs_stats.json\n",
      "Processed SEZONOWIFGOT007_RS_S70_673666_contigs_stats.json\n",
      "Processed SEZONOWIFRR031_LIV_S71_663085_contigs_stats.json\n",
      "Processed SEZONOWIFRR032_LIV_S72_666406_contigs_stats.json\n",
      "Processed SEZONOWIFRR033_LIV_S73_666402_contigs_stats.json\n",
      "Processed SEZONOWIFRR034_LIV_S74_1_666404_contigs_stats.json\n",
      "Processed SEZONOWIFRR037_LIV_S77_666399_contigs_stats.json\n",
      "Processed SEZONOWIFRR038_LIV_S78_666400_contigs_stats.json\n",
      "Processed SEZONOWIFRR039_LIV_S79_666401_contigs_stats.json\n",
      "Processed SEZONOWIFRR040_LIV_S80_666396_contigs_stats.json\n",
      "Processed SEZONOWIFRR063_LUNG_S81_666397_contigs_stats.json\n",
      "Processed SEZONOWIFRR064_LUNG_S56_673522_contigs_stats.json\n",
      "Processed SEZONOWIFRR066_LUNG_S57_673523_contigs_stats.json\n",
      "Processed SEZONOWIFRR067_LUNG_S58_673524_contigs_stats.json\n",
      "Processed SEZONOWIFRR068_LUNG_S59_673525_contigs_stats.json\n",
      "Processed SEZONOWIFRR069_LUNG_S60_673526_contigs_stats.json\n",
      "Processed SEZONOWIFRR070_LUNG_S61_673529_contigs_stats.json\n",
      "Processed SEZONOWIFRR071_LUNG_S62_673527_contigs_stats.json\n",
      "Processed SEZONOWIFRR072_LUNG_S63_673528_contigs_stats.json\n",
      "Processed SEZONOWIFSHEEP001_S33_656082_contigs_stats.json\n",
      "Processed SEZONOWIFSHEEP002_S31_656083_contigs_stats.json\n",
      "Processed SEZONOWISDOG002_S4_657741_contigs_stats.json\n",
      "Processed SEZONOWISDOG008_S14_657742_contigs_stats.json\n",
      "Processed SEZONOWISDOG015_S13_657743_contigs_stats.json\n",
      "Processed SEZONOWISDOGOO1_S19_657744_contigs_stats.json\n",
      "Processed SEZONOWISGOAT008_S25_657722_contigs_stats.json\n",
      "Processed SEZONOWISGOAT013_S24_2_657729_contigs_stats.json\n",
      "Processed SEZONOWISLIZARD001_S36_657719_contigs_stats.json\n",
      "Processed SEZONOWISSHEEP001_S32_656084_contigs_stats.json\n",
      "Processed SEZONOWISUGOAT003_S26_657724_contigs_stats.json\n",
      "Processed SEZONOWOGDOG001_S6_657746_contigs_stats.json\n",
      "Processed SEZONOWOGDOG002_S5_657747_contigs_stats.json\n",
      "Processed SEZONOWOGDOG004_S12_657748_contigs_stats.json\n",
      "Processed SEZONOWOGDOG00_S2_657745_contigs_stats.json\n",
      "Processed SEZONOWOGDOG010_S3_657749_contigs_stats.json\n",
      "Processed SEZONOWOGDOG013_S18_657750_contigs_stats.json\n",
      "Processed SEZONOWOGDOG014_S21_657751_contigs_stats.json\n",
      "Processed SEZONOWOGDOG016_S23_657752_contigs_stats.json\n",
      "Processed SEZONOWOGDOG017_S22_657753_contigs_stats.json\n",
      "Processed SEZONOWOGGOAT015_S28_657725_contigs_stats.json\n",
      "Processed SEZONOWOGGOAT018_S66_1_657730_contigs_stats.json\n",
      "Processed SEZONOWOGR001_LIV_S47_666398_contigs_stats.json\n",
      "Processed SEZONOWOGR002_LIV_S48_666373_contigs_stats.json\n",
      "Processed SEZONOWOGR003_LIV_S49_666374_contigs_stats.json\n",
      "Processed SEZONOWOGR004_LIV_S50_666375_contigs_stats.json\n",
      "Processed SEZONOWOGR005_LIV_S51_666376_contigs_stats.json\n",
      "Processed SEZONOWOGR006_LIV_S52_666377_contigs_stats.json\n",
      "Processed SEZONOWOGR007_LIV_S53_666370_contigs_stats.json\n",
      "Processed SEZONOWOGR008_LIV_S54_666371_contigs_stats.json\n",
      "Processed SEZONOWOGR009_LIV_S55_666372_contigs_stats.json\n",
      "Processed SEZONOWOGR010_LIV_S56_666407_contigs_stats.json\n",
      "Processed SEZONOWOGR081_LUNG_S57_666409_contigs_stats.json\n",
      "Processed SEZONOWOGR082_LUNG_S58_666410_contigs_stats.json\n",
      "Processed SEZONOWOGR083_LUNG_S59_3_666413_contigs_stats.json\n",
      "Processed SEZONOWOGR085_LUNG_S61_666414_contigs_stats.json\n",
      "Processed SEZONOWOWISR001_LIV_S62_666415_contigs_stats.json\n",
      "Processed SEZONOWOWISR002_LIV_S63_666416_contigs_stats.json\n",
      "Processed SEZONOWOWISR003_LIV_S64_666417_contigs_stats.json\n",
      "Processed SEZONOWOWISR004_LIV_S65_666418_contigs_stats.json\n",
      "Processed SEZONOWOWISR005_LIV_S66_666419_contigs_stats.json\n",
      "Processed SEZONOWOWISR086_LIV_S67_666422_contigs_stats.json\n",
      "Processed SEZONOWOWISR087_LUNG_S68_666423_contigs_stats.json\n",
      "Processed SEZONOWOWISR088_LIV_S69_666424_contigs_stats.json\n",
      "Processed SEZONOWOWISR089_LUNG_S70_666425_contigs_stats.json\n",
      "Processed SEZOOWISDOG003_S11_657754_contigs_stats.json\n",
      "Processed SEZOOWISUGOAT002_S27_657721_contigs_stats.json\n",
      "Processed SEZOOWOGCHICKEN001_S35_655749_contigs_stats.json\n",
      "Processed SEZOOWOGDOG003_S17_657755_contigs_stats.json\n",
      "Processed SEZOOWOGDOG005_S15_657756_contigs_stats.json\n",
      "Processed SEZOOWOGDOG006_S16_1_657760_contigs_stats.json\n",
      "Processed SEZOOWOGDOG011_S8_1_657761_contigs_stats.json\n",
      "Processed SEZOOWOGDOG019_S9_1_657762_contigs_stats.json\n",
      "Processed SSP_ERCC_S72_670948_contigs_stats.json\n",
      "Processed SSP_ERCC_S83_666426_contigs_stats.json\n",
      "Processed SSP_ERCC_S83_673530_contigs_stats.json\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import os\n",
    "\n",
    "def remove_substring_from_keys(obj, substring):\n",
    "    if isinstance(obj, dict):\n",
    "        new_obj = {}\n",
    "        for key, value in obj.items():\n",
    "            new_key = key.replace(substring, \"\")\n",
    "            new_obj[new_key] = remove_substring_from_keys(value, substring)\n",
    "        return new_obj\n",
    "    elif isinstance(obj, list):\n",
    "        return [remove_substring_from_keys(item, substring) for item in obj]\n",
    "    else:\n",
    "        return obj\n",
    "\n",
    "directory = \"/Volumes/aine_store/SENZOR_project/SPlited_SENSOR_porject/IDseq/Combined_Contig_quality\"\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith(\".json\"):\n",
    "        filepath = os.path.join(directory, filename)\n",
    "        with open(filepath, 'r') as f:\n",
    "            data = json.load(f)\n",
    "        \n",
    "        updated_data = remove_substring_from_keys(data, \"extracted_csv\")\n",
    "        \n",
    "        with open(filepath, 'w') as f:\n",
    "            json.dump(updated_data, f, indent=4)\n",
    "        \n",
    "        print(f\"Processed {filename}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
